{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-08-31T12:04:52.894217Z",
     "start_time": "2025-08-31T12:04:52.194835Z"
    }
   },
   "source": [
    "from scripts.data_genertion.consts import *\n",
    "from scripts.features.feature_extraction import load_all_features\n",
    "\n",
    "main_df = load_all_features()\n",
    "main_df[SEQUENCE] = main_df[SEQUENCE].astype(str)"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-31T12:04:53.013660Z",
     "start_time": "2025-08-31T12:04:52.898333Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from asodesigner.read_human_genome import get_locus_to_data_dict\n",
    "import pickle\n",
    "from asodesigner.consts import CACHE_DIR\n",
    "\n",
    "genes_u = ['HIF1A', 'APOL1', 'YAP1', 'SOD1', 'SNCA', 'IRF4', 'KRAS', 'KLKB1', 'SNHG14', 'DGAT2', 'IRF5', 'HTRA1',\n",
    "           'MYH7', 'MALAT1', 'HSD17B13']\n",
    "cache_path = CACHE_DIR / 'gene_to_data_simple_cache.pickle'\n",
    "if not cache_path.exists():\n",
    "    gene_to_data = get_locus_to_data_dict(include_introns=True, gene_subset=genes_u)\n",
    "    with open(cache_path, 'wb') as f:\n",
    "        pickle.dump(gene_to_data, f)\n",
    "else:\n",
    "    with open(cache_path, 'rb') as f:\n",
    "        gene_to_data = pickle.load(f)"
   ],
   "id": "50d64118f2c22bfd",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-31T12:05:08.423086Z",
     "start_time": "2025-08-31T12:04:53.095752Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from scripts.data_genertion.data_handling import get_populated_df_with_structure_features\n",
    "\n",
    "main_df = get_populated_df_with_structure_features(main_df, genes_u, gene_to_data)"
   ],
   "id": "48b0dbba50a89a98",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-31T12:05:08.437086Z",
     "start_time": "2025-08-31T12:05:08.430633Z"
    }
   },
   "cell_type": "code",
   "source": "main_df[[SENSE_TYPE, 'sense_intron', 'sense_exon', 'sense_utr', SENSE_START, CANONICAL_GENE]]",
   "id": "803511a1c7fe474b",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      sense_type  sense_intron  sense_exon  sense_utr  sense_start  \\\n",
       "0         intron             1           0          0        41212   \n",
       "1         intron             1           0          0        23686   \n",
       "2         intron             1           0          0        43363   \n",
       "3         intron             1           0          0        23680   \n",
       "4         intron             1           0          0        41168   \n",
       "...          ...           ...         ...        ...          ...   \n",
       "34760     intron             1           0          0         7827   \n",
       "34761     intron             1           0          0         8250   \n",
       "34762     intron             1           0          0         8335   \n",
       "34763        utr             0           1          1        13848   \n",
       "34764     intron             1           0          0         5361   \n",
       "\n",
       "      Canonical Gene Name  \n",
       "0                    KRAS  \n",
       "1                    KRAS  \n",
       "2                    KRAS  \n",
       "3                    KRAS  \n",
       "4                    KRAS  \n",
       "...                   ...  \n",
       "34760               APOL1  \n",
       "34761               APOL1  \n",
       "34762               APOL1  \n",
       "34763               APOL1  \n",
       "34764               APOL1  \n",
       "\n",
       "[29987 rows x 6 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sense_type</th>\n",
       "      <th>sense_intron</th>\n",
       "      <th>sense_exon</th>\n",
       "      <th>sense_utr</th>\n",
       "      <th>sense_start</th>\n",
       "      <th>Canonical Gene Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41212</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23686</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>43363</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23680</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41168</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34760</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7827</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34761</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8250</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34762</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8335</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34763</th>\n",
       "      <td>utr</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13848</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34764</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5361</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29987 rows × 6 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-31T12:05:08.502512Z",
     "start_time": "2025-08-31T12:05:08.496630Z"
    }
   },
   "cell_type": "code",
   "source": [
    "main_df.loc[:, 'mrna_length'] = [len(gene_to_data[gene].full_mrna) for gene in main_df[CANONICAL_GENE]]\n",
    "main_df['normalized_sense_start_from_end'] = main_df['sense_start_from_end'] / main_df['mrna_length']"
   ],
   "id": "603adfe4a05886bb",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-31T12:05:08.555677Z",
     "start_time": "2025-08-31T12:05:08.551260Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from scripts.data_genertion.consts import *\n",
    "\n",
    "\n",
    "class Features:\n",
    "    def __init__(self, features):\n",
    "        self.__all_features = features\n",
    "        self.__organize()\n",
    "\n",
    "    def __organize(self):\n",
    "        self.hybridization = []\n",
    "        self.sequence = []\n",
    "        self.fold = []\n",
    "        self.one_hot = []\n",
    "        for feature in self.__all_features:\n",
    "            if \"dsm\" in feature:\n",
    "                self.hybridization.append(feature)\n",
    "            if feature in ['md_ps_hybr_norm', 'md_ps_hybr', 'exp_ps_hybr', 'exp_ps_hybr_norm', 'melting_temperature_1',\n",
    "                           'dsm_su95_rev_wGU_pos1382t37Falseon_target_energy_max600'\n",
    "                           ]:\n",
    "                self.hybridization.append(feature)\n",
    "            # if feature in [SENSE_TYPE]:\n",
    "            #     self.one_hot.append(feature)\n",
    "            if feature in ['at_skew', 'gc_content', 'entropy', 'gc_skew',\n",
    "                           'gc_content_3_prime_5', 'hairpin_score', 'homooligo_count', 'internal_fold',\n",
    "                           'nucleotide_diversity', 'stop_codon_count', 'self_energy', 'at_rich_region_score',\n",
    "                           'poly_pyrimidine_stretch', 'normalized_start', 'sense_exon', 'sense_intron', 'sense_utr']:\n",
    "                self.sequence.append(feature)\n",
    "            if 'on_target_fold' in feature or feature in ['mfe_edges_45', 'mfe_window_45']:\n",
    "                self.fold.append(feature)\n",
    "\n",
    "    def get_all(self):\n",
    "        return self.hybridization + self.sequence + self.fold\n",
    "\n",
    "\n",
    "main_df.columns"
   ],
   "id": "97b1bb12534d5f05",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['index', 'ASO_volume(nM)', 'Canonical Gene Name', 'Cell line organism',\n",
       "       'Cell_line', 'Chemical_Pattern', 'Density(cells_per_well)', 'ISIS',\n",
       "       'Inhibition(%)', 'Linkage', 'Linkage_Location', 'Location',\n",
       "       'Location_div_by_length', 'Location_in_sequence', 'Modification',\n",
       "       'Primer_probe_set', 'Sequence', 'Target_gene', 'Transcript',\n",
       "       'Transfection', 'Treatment_Period(hours)', 'at_rich_region_score',\n",
       "       'at_skew', 'cell_line_uniform',\n",
       "       'dsm_su95_rev_wGU_pos1382t37Falseon_target_energy_max600',\n",
       "       'dsm_su95_rev_wGU_pos1382t37Trueon_target_energy_max600',\n",
       "       'dsm_su95_rev_wGU_pos1384t37Falseon_target_energy_max600',\n",
       "       'dsm_su95_rev_wGU_pos1384t37Trueon_target_energy_max600',\n",
       "       'dsm_su95_rev_wGU_pos1386t37Falseon_target_energy_max600',\n",
       "       'dsm_su95_rev_wGU_pos1386t37Trueon_target_energy_max600',\n",
       "       'dsm_su95_rev_woGU_pos1382t37Falseon_target_energy_max600',\n",
       "       'dsm_su95_rev_woGU_pos1382t37Trueon_target_energy_max600',\n",
       "       'dsm_su95_rev_woGU_pos1384t37Falseon_target_energy_max600',\n",
       "       'dsm_su95_rev_woGU_pos1384t37Trueon_target_energy_max600',\n",
       "       'dsm_su95_rev_woGU_pos1386t37Falseon_target_energy_max600',\n",
       "       'dsm_su95_rev_woGU_pos1386t37Trueon_target_energy_max600', 'entropy',\n",
       "       'exp_ps_hybr', 'exp_ps_hybr_norm', 'first_nucleotide', 'gc_content',\n",
       "       'gc_content_3_prime_5', 'gc_content_5_prime_5', 'gc_skew',\n",
       "       'hairpin_score', 'homooligo_count', 'index.1', 'internal_fold',\n",
       "       'log_inhibition', 'md_moe_hybr', 'md_moe_hybr_norm', 'md_ps_hybr',\n",
       "       'md_ps_hybr_norm', 'melting_temperature_1', 'mfe_edges_45',\n",
       "       'mfe_window_45', 'min_mfe_45', 'mod_scan', 'moe_hybr', 'moe_hybr_norm',\n",
       "       'mrna_length', 'normalized_start', 'nucleotide_diversity',\n",
       "       'off_target_score_TPM_x', 'off_target_score_log_x',\n",
       "       'off_target_score_TPM_y', 'off_target_score_log_y',\n",
       "       'off_target_score_TPM', 'off_target_score_log',\n",
       "       'on_target_fold_openness40_15',\n",
       "       'on_target_fold_openness_normalized40_15', 'poly_pyrimidine_stretch',\n",
       "       'second_nucleotide', 'self_energy', 'sense_avg_accessibility',\n",
       "       'sense_length', 'sense_start', 'sense_start_from_end', 'sense_type',\n",
       "       'seq_length', 'stop_codon_count', 'true_length_of_seq', 'sense_exon',\n",
       "       'sense_intron', 'sense_utr', 'normalized_sense_start_from_end'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-31T12:05:08.590739Z",
     "start_time": "2025-08-31T12:05:08.588825Z"
    }
   },
   "cell_type": "code",
   "source": "# populate_features(main_df, ['at_rich_region_score', 'poly_pyrimidine_stretch'])",
   "id": "f7b6ee43bffdc2f9",
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-31T12:05:08.643027Z",
     "start_time": "2025-08-31T12:05:08.641540Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# from scripts.features.feature_extraction import save_feature\n",
    "#\n",
    "# save_feature(main_df, 'at_rich_region_score')\n",
    "# save_feature(main_df, 'sense_start_from_end')"
   ],
   "id": "ab9d4952ed98d60b",
   "outputs": [],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-31T12:05:08.719945Z",
     "start_time": "2025-08-31T12:05:08.690755Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "first_filtered = main_df.copy()\n",
    "\n",
    "first_filtered = first_filtered[first_filtered['Cell line organism'] == 'human']\n",
    "# filtered = filtered[filtered[CANONICAL_GENE] == 'MALAT1']\n",
    "first_filtered = first_filtered.dropna(subset=[INHIBITION]).copy()\n",
    "# filtered = filtered.dropna(subset=[DENSITY_UPDATED]).copy()\n",
    "log_correction = 1.01\n",
    "first_filtered.loc[:, 'log_inhibition'] = -np.log(log_correction - first_filtered[INHIBITION] / 100)\n",
    "first_filtered = first_filtered[~first_filtered[CELL_LINE].isin([\n",
    "    'Hela',  # scanning modifications\n",
    "    'Human Neuronal Cell',  # scanning modifications;\n",
    "    'CC-2580',  # scanning modifications\n",
    "    'SH-SY5Y'  # non pure PS based\n",
    "])]\n",
    "# first_filtered = first_filtered[~((first_filtered[CELL_LINE] == 'A431')\n",
    "#                                   & (first_filtered[CANONICAL_GENE] == 'SOD1'))] # (Kind of) Scanning modifications\n",
    "\n",
    "mean_metric = 'log_inhibition'\n",
    "# mean_metric = INHIBITION\n",
    "\n",
    "# Ensure you remove only the columns you don't want to group by\n",
    "# cols_except_inhibition = [c for c in first_filtered.columns\n",
    "#                           if c not in [INHIBITION, mean_metric, 'index']]\n",
    "\n",
    "# Group by all remaining columns and average log_inhibition\n",
    "collapsed = (\n",
    "    first_filtered\n",
    "    .groupby(['ISIS', VOLUME, TREATMENT_PERIOD, DENSITY_UPDATED, CHEMICAL_PATTERN], as_index=False)[mean_metric]\n",
    "    .mean()\n",
    ")\n",
    "\n",
    "# first_filtered['mean_inhibition'] = first_filtered.groupby('ISIS')[mean_metric].transform('mean')\n",
    "# first_filtered = first_filtered.drop_duplicates(subset='ISIS').assign(**{mean_metric: first_filtered['mean_inhibition']}).drop(\n",
    "#     columns='mean_inhibition')"
   ],
   "id": "909c3b214ad4bc5",
   "outputs": [],
   "execution_count": 9
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-31T12:05:08.745938Z",
     "start_time": "2025-08-31T12:05:08.740395Z"
    }
   },
   "cell_type": "code",
   "source": "collapsed",
   "id": "6b022ed716a0b126",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "             ISIS  ASO_volume(nM)  Treatment_Period(hours)  \\\n",
       "0         76577.0          4000.0                     48.0   \n",
       "1        146143.0           187.5                     16.0   \n",
       "2        146143.0           750.0                     16.0   \n",
       "3        146143.0          3000.0                     16.0   \n",
       "4        146143.0          4000.0                     24.0   \n",
       "...           ...             ...                      ...   \n",
       "23189  70001072.0          5000.0                     96.0   \n",
       "23190  70001081.0          5000.0                     96.0   \n",
       "23191  70001101.0          5000.0                     96.0   \n",
       "23192  70001102.0          5000.0                     96.0   \n",
       "23193  70001103.0          5000.0                     96.0   \n",
       "\n",
       "       Density(cells_per_well)      Chemical_Pattern  log_inhibition  \n",
       "0                      10000.0      CCCddddddddddCCC        0.462035  \n",
       "1                      20000.0  MMMMMddddddddddMMMMM        0.051293  \n",
       "2                      20000.0  MMMMMddddddddddMMMMM        0.116534  \n",
       "3                      20000.0  MMMMMddddddddddMMMMM        0.693147  \n",
       "4                      20000.0  MMMMMddddddddddMMMMM        0.713350  \n",
       "...                        ...                   ...             ...  \n",
       "23189                  15000.0      LLLdddddddddLLLL        1.347074  \n",
       "23190                   2000.0     LLLLdddddddddLLLL        1.108663  \n",
       "23191                  15000.0      LLLddddddddddLLL        1.386294  \n",
       "23192                  15000.0      LLLdddddddddLLLL        1.272966  \n",
       "23193                  15000.0      LLLLddddddddLLLL        1.560648  \n",
       "\n",
       "[23194 rows x 6 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ISIS</th>\n",
       "      <th>ASO_volume(nM)</th>\n",
       "      <th>Treatment_Period(hours)</th>\n",
       "      <th>Density(cells_per_well)</th>\n",
       "      <th>Chemical_Pattern</th>\n",
       "      <th>log_inhibition</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>76577.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>10000.0</td>\n",
       "      <td>CCCddddddddddCCC</td>\n",
       "      <td>0.462035</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>146143.0</td>\n",
       "      <td>187.5</td>\n",
       "      <td>16.0</td>\n",
       "      <td>20000.0</td>\n",
       "      <td>MMMMMddddddddddMMMMM</td>\n",
       "      <td>0.051293</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>146143.0</td>\n",
       "      <td>750.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>20000.0</td>\n",
       "      <td>MMMMMddddddddddMMMMM</td>\n",
       "      <td>0.116534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>146143.0</td>\n",
       "      <td>3000.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>20000.0</td>\n",
       "      <td>MMMMMddddddddddMMMMM</td>\n",
       "      <td>0.693147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>146143.0</td>\n",
       "      <td>4000.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>20000.0</td>\n",
       "      <td>MMMMMddddddddddMMMMM</td>\n",
       "      <td>0.713350</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23189</th>\n",
       "      <td>70001072.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>15000.0</td>\n",
       "      <td>LLLdddddddddLLLL</td>\n",
       "      <td>1.347074</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23190</th>\n",
       "      <td>70001081.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>LLLLdddddddddLLLL</td>\n",
       "      <td>1.108663</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23191</th>\n",
       "      <td>70001101.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>15000.0</td>\n",
       "      <td>LLLddddddddddLLL</td>\n",
       "      <td>1.386294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23192</th>\n",
       "      <td>70001102.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>15000.0</td>\n",
       "      <td>LLLdddddddddLLLL</td>\n",
       "      <td>1.272966</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23193</th>\n",
       "      <td>70001103.0</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>96.0</td>\n",
       "      <td>15000.0</td>\n",
       "      <td>LLLLddddddddLLLL</td>\n",
       "      <td>1.560648</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>23194 rows × 6 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 10
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-31T12:05:08.812803Z",
     "start_time": "2025-08-31T12:05:08.806815Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# filtered = filtered[filtered['Linkage'] == 'phosphorothioate']\n",
    "\n",
    "moe_pattern = 'MMMMMddddddddddMMMMM'\n",
    "# filtered = filtered[filtered['Chemical_Pattern'] == moe_pattern].copy()\n",
    "# filtered = filtered[filtered[VOLUME] == 2500].copy()\n",
    "filtered = first_filtered.copy()\n",
    "\n",
    "filtered['log_volume'] = np.log(filtered[VOLUME])\n",
    "filtered['log_treatment'] = np.log(filtered[TREATMENT_PERIOD])\n",
    "li = filtered['log_inhibition']\n",
    "\n",
    "# filtered['correct_log_inhibition'] = filtered['log_inhibition'] / (filtered['log_volume'])\n",
    "filtered['correct_log_inhibition'] = li / (filtered[VOLUME] / (filtered[VOLUME] + 10))\n",
    "filtered['correct_log_inhibition2'] = li / (filtered[VOLUME] * 0.0000601 + 0.537)\n",
    "# filtered['correct_log_inhibition'] = li\n",
    "\n",
    "\n",
    "features_obj = Features(filtered.columns)\n",
    "features_obj.hybridization = ['exp_ps_hybr']\n",
    "features_obj.fold = ['on_target_fold_openness_normalized40_15']\n",
    "# features_obj.fold = []\n",
    "features = features_obj.get_all()\n",
    "# features = features_obj.hybridization + features_obj.fold\n",
    "epi_features = [\n",
    "    TREATMENT_PERIOD,\n",
    "    # 'log_volume',\n",
    "    # DENSITY_UPDATED\n",
    "]\n",
    "print(features_obj.one_hot)\n",
    "for feature in features_obj.one_hot:\n",
    "    filtered = pd.get_dummies(filtered, columns=[feature]).copy()\n",
    "\n",
    "one_hot_features = []\n",
    "for one_hot_feature in features_obj.one_hot:\n",
    "    one_hot_features.append(\n",
    "        [feature for feature in filtered.columns if one_hot_feature in feature and one_hot_feature != feature])\n",
    "flat_one_hot_encoded_features = [feature for sublist in one_hot_features for feature in sublist]\n",
    "# features = epi_features\n",
    "features = epi_features + features + flat_one_hot_encoded_features\n"
   ],
   "id": "a5930b883db04e3",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[]\n"
     ]
    }
   ],
   "execution_count": 11
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-31T12:05:08.858692Z",
     "start_time": "2025-08-31T12:05:08.851616Z"
    }
   },
   "cell_type": "code",
   "source": [
    "filtered_original = filtered\n",
    "\n",
    "ALL_CELL_LINES = filtered_original[CELL_LINE].unique()\n",
    "only_cancer = filtered_original[\n",
    "    # (~((filtered_original[CELL_LINE] == 'HepG2') & (filtered_original[TRANSFECTION].str.contains('Lipo')))) &\n",
    "    (filtered_original[CELL_LINE] != 'HepG2')  # experiment settings too different\n",
    "    & (filtered_original[CELL_LINE] != 'HepaRG')  # not similar to cancer\n",
    "    & (filtered_original[CELL_LINE] != 'A-431')\n",
    "    ]"
   ],
   "id": "d4220aceccb06fd5",
   "outputs": [],
   "execution_count": 12
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-01T09:32:27.423662Z",
     "start_time": "2025-09-01T09:32:27.311526Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import pearsonr, spearmanr\n",
    "\n",
    "x = only_cancer['sense_intron']\n",
    "y = only_cancer[INHIBITION]\n",
    "\n",
    "plt.clf()\n",
    "plt.scatter(x, y, alpha=0.02)\n",
    "plt.xlabel('Sense intron')\n",
    "plt.ylabel('Inhibition (%)')\n",
    "\n",
    "# === toggle fit type ===\n",
    "use_quadratic = False   # set False for linear\n",
    "\n",
    "xs = np.linspace(np.nanmin(x), np.nanmax(x), 500)\n",
    "\n",
    "if use_quadratic:\n",
    "    # Quadratic fit\n",
    "    a, b, c = np.polyfit(x, y, 2)\n",
    "    ys = a*xs**2 + b*xs + c\n",
    "    y_pred = a*x**2 + b*x + c\n",
    "    fit_label = f\"y = {a:.3f}x² + {b:.3f}x + {c:.3f}\"\n",
    "else:\n",
    "    # Linear fit\n",
    "    m, b = np.polyfit(x, y, 1)\n",
    "    ys = m*xs + b\n",
    "    y_pred = m*x + b\n",
    "    fit_label = f\"y = {m:.3f}x + {b:.3f}\"\n",
    "\n",
    "plt.plot(xs, ys, color=\"red\", linewidth=2)\n",
    "\n",
    "# Correlations\n",
    "r_p, p_p = pearsonr(x, y)\n",
    "r_s, p_s = spearmanr(x, y)\n",
    "\n",
    "print(f\"Pearson r = {r_p:.3f}, p = {p_p:.2e}\")\n",
    "print(f\"Spearman r = {r_s:.3f}, p = {p_s:.2e}\")\n",
    "print(\"Fit:\", fit_label)\n",
    "\n",
    "# R²\n",
    "ss_res = np.sum((y - y_pred)**2)\n",
    "ss_tot = np.sum((y - np.mean(y))**2)\n",
    "r2 = 1 - ss_res/ss_tot\n",
    "\n",
    "plt.text(0.05, 0.95,\n",
    "         # f\"R² = {r2:.3f}\\n\"\n",
    "         f\"Spearman r = {r_s:.3f}, p = {p_s:.2e}\",\n",
    "         transform=plt.gca().transAxes,\n",
    "         ha=\"left\", va=\"top\",\n",
    "         fontsize=10,\n",
    "         bbox=dict(facecolor=\"white\", alpha=0.5, edgecolor=\"none\"))\n",
    "\n",
    "plt.show()\n"
   ],
   "id": "20bf7442a255c547",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pearson r = -0.057, p = 1.62e-12\n",
      "Spearman r = -0.063, p = 6.96e-15\n",
      "Fit: y = -3.601x + 40.856\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 640x480 with 1 Axes>"
      ],
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAkcAAAGwCAYAAACjPMHLAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjkuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/TGe4hAAAACXBIWXMAAA9hAAAPYQGoP6dpAABKzUlEQVR4nO3df5yM9f7/8ec1Mzt21661P/ymH8qSn0tEEUU/nfSDdHREh37gpEJHFqmwQqUtlpBI0umHTqpzfFKdvs45pONHiISU3xK7WL+tnbm+f4ydrmHXzrB7ze563G83NzvXvK/res9rdnee+77e13UZpmmaAgAAgCTJEe4OAAAAlCSEIwAAAAvCEQAAgAXhCAAAwIJwBAAAYEE4AgAAsCAcAQAAWBCOAAAALAhHAAAAFq5wd6A0y8o6rKK8vrhhSImJsUW+XQSizvagzvah1vagzvYozjrnbbswhKMLYJoqlh+Q4touAlFne1Bn+1Bre1Bne4SzzhxWAwAAsCAcAQAAWBCOAAAALAhHAAAAFoQjAAAAC8IRAACABeEIAADAgnAEAABgQTgCAACw4ArZJcjRo0fldTl0/OhRRUeXD3d3AACw3apVq7QjLla1sg8rJaVpWPpAOCoBzAhDR47maNu+Q/JmH5Ejx6tLq7kVU96t8hER4e5emWMYvv+jo93cAqAYUefQHDuWE+4uAGH1/qpV+vzrw1onSTosSWr4r//otvax+mNTe0MS4SjM9h49qp37D2nrfum4JMkrSdp0MFOXJUhNa1cnIAEAyrT3V63Sy18fPmv5Oknrvj4saZWtAYk5R2H2t69X6kd/MPrdcUk/7pfWbdkdjm4BAGCb/IJRKM8XNcJRGB09elTvbDp3m++zpFOnTtnTIQAAbLZq1aoibVcUCEdh9M2aNUG127FnTzH3BACA8Ogf5KhQsO2KAuEojP69Mjeodlt/9RZzTwAACI9gT0Ww85QFwlEYLQzynf7ZU7z9AAAAv+NstRLs2OGDWv7ZO9q+fqVmHMlWhQoVVKdOXfXp009NmjQJd/dgcejQIb344jj95z+LJElt296gIUNSFRtbocB1TNPUtGmv6+9//0iHDx9Sw4aNlJo6TFdccWVAuzVr1mjy5Ilat26tXK4I1a1bV5MmTVFkZKQkacCAx7Vp00bt379fFSpU0DXXtNKTTw5QpUqVi+31FqX//vc/euONqfrpp58UFRWlpk2v1oQJ6QW2z8rK0sSJ6Vq6dKmOHDmspk2baciQobrkkksD2hVWt/OxcuUKvf32W/rxxx+VmblPEya8qhtvbB/Q5rnnntFnn30asKxhw0Z6++25571fAPYiHJVgC98YK6/Ho/Y9B+rRm6/W/v1ZWrbsfzp0KDvcXdOpU6cUUYIvMWCapjwej1wue77Fhw0bor17f9OkSa9LksaMGalnnhmm117LKHCd2bNnae7cOXr++dG69NJLNWPGG+rXr48+/vhTlS/vuwjomjVr9Pjj/dSr10MaMmSoIiIitGnTRjkcvw/6Nm9+jXr3flhJSZW0b99epadP0ODBT+mtt+YU74suAv/615caPXqk+vd/Qi1aXCPTNLV5808FtjdNU4MGPSmXy6X09NdUvnx5vfPOHPXt+6g++uhjRUVFSwqubufjxInjSk6uqzvvvFuDBw8qsN1117XW88+P9j8uyT8rAM5GOCqhTh47oj0/r9ddA15Q9TqNVL16dVWvXl0NGzYKaNesWWMNHTpc//73Iq1cuUKJiYl68slBuvnmW/xt9u79TRMmvKxvv/1GDodDKSlNNXjwEFWvXkOS9MMP65SRMVEbN25Qbm6ukpPr6qmnBuuqq+oH7GfYsGe0ZMli/e9/36pHjwdlGIYWLfp/6tbtT5o+/XVlZ2frD3+4Q0OGDNOcObM1d+4ceb1e3X9/dz388KP+bb3zztv69NP52rlzp+Li4tS2bTs9+eQgRUf7Ptg+/fQTvfzyixo37kW9/PKL+u23PUpJaarnnx+tSpUq5VuvFSuW69FHH1JGxuuaPHmSfvppkyZPnqoWLa4psvekIL/88ou++WaJZs9+R40aNZYkPfPMc/rzn3to69Ytuuyyy89axzRNvfvuO3rooUfUocNNkqRRo9J000036v/+b4HuvberJGnChBfVrduf1KvXQ/51zxwheeCBHv6vq1evrl69emvQoAEhBdipU6do0aL/p3vvvU9vvjld2dnZatPmeo0Y8dw5R78uRG5url56abwGDBiku+/u7F+eX73ybN++TWvXfq8PP/y7f4Rt6NDhuummG/T55/+ne+7pIim4uhX2c5Gf1q2vV+vW1xf62txut5KSkgptB6BkYs5RCRVRLkoR5aK0Zc238hRyKv/rr09Whw436b33PlTHjndo2LAh+uWXXyRJx48f16OPPqzo6GjNmDFLb745W9HR0erfv5//EgFHjx7VHXfcqRkz3tJbb72jSy65RE888ZiOHj0asJ+pU6eoXbsb9cEHf9ddd90jSdq5c4e++WaxMjJe1wsvjNcnn8zXE088pr179+qNN2bqiScGasqUDH3//e9n5hmGocGDU/Xhh3/XyJFpWr58mV577ZWAfZ04cVxz5sxWWtoLmjFjlvbs2aNXX51QaN1eey1djz/+pD766BPVqZOcb5suXe5R06ZNdd11LdW69dn/7r33nkL3Y/X992sUExPrD0aS1LhxE8XExGpNAWck7tq1S5mZmWrV6lr/Mrfbrauvvlrff79akrR/f5bWrVurhIQE/fnPPXTTTTfo4Yd7adWq7wrsS3Z2thYsWKAmTVJCHq3YsWO7vvxyoV59dZImTZqijRs3aty4F865Tn71y/t33XUt9fDDDxe47oYNP2rv3r0yDIfuv/8+3XJLe/Xv308//7y5wHVycnwT9dzucv5lTqdTERERWr3ad5pvMHUL5ufiQqxYsUIdOrTT3Xd30ujRz2v//qwL3iYA+zByVEI5nE7d2ONJ/fvdDP2w+HP9r359NWt2tW699XYlJwd+6N90083+v5j/8pf++vbbpXr//Xc1dOgzWrjwczkchp599nkZp+/n8Pzzo9WuXWutWLFc1157na65pmXA9oYPf1Y33NBGK1euUNu27fzLb7uto+6+OzA4eL1ePffcKJUvX161a1+h5s1baNu2rZo0aYocDocuu+xyzZ49UytXrlDjxr55Ut27/z7SUaNGTfXr119jx6Zp6NBn/Mtzc3M1bNgI1apVS5L0xz920xtvTCu0bv36PRYQOPIzadJkRUQ4dPx4Tr63tQj1UFxWVqYSEhLOWp6QkKCsrMwC15GkxMTEM9ZJ1K+//ipJ2rlzpyRp2rTXNWDAU6pbt67+8Y/P1LfvI/rww78HjIS89lq63n//bzpx4oQaNWp8zsN5BcnJydGoUWmqUqWqJOnpp1P15JP9NXDgXwscBfnb3z4scHuGIcXHFzzqtGvX76/vqaf+qmrVauidd2brkUd66+OPP1NcXNxZ61x22eWqVq26MjJe0/DhzyoqKkrvvPO2MjMztW+fr6bB1C2Yn4vzdd11bXTTTbeoWrVq2rVrl15/fbL69HlYc+e+L7fbfd7bBWAfwlEJdkXT1rq0YQv9uvkH1czdo2++WaK3335LI0Y8rzvvvMvfLi90WB9v2rRRkvTjj+u1Y8cOtWnTKqDNyZMntXPnDkm+v7Rff32Kli9fpv37s+TxeHTixAnt2fNrwDr16zc4q4/Vq9fwz4+RfB/2TqczYG5HQkKi9u/f73+8fPkyzZw5Q7/88ouOHj0ij8ejkydP6vjxY/45I5GRkf5gJElJSZUCtlGQ+vXrF9qmevXqio4up2PHToZ8z68xY0ZrwYJ/+B8vWfI/Sb/fR8zKNE3/B2/Bznze9G/LPN25zp3v1V133S1JqlfvKi1b9j998sl8Pf74k/61evb8s+6++x79+uuvmj59qp59drheey0jiP3/rmrVqv5gJPm+j7xer7Zt21pgOLrkkksKfmWG/HXOj9fru0SF79DizZJ8AeW2227Wl19+4T+0aBUREaGXXnpFo0Y9pxtuaCOn06lrrmmp1q3b+NsEU7fCfi6++26lHn/8L/7lw4c/q44d/1Dga7W69dbb/F9feWUd1a/fQH/4w63673//4z+ECqBkIxyVcK4It2pd1VSPXHupHn20r0aNek5Tp04JCEfnYppeXXXVVUpLG3fWc/Hx8ZKk554boQMHDuivf31a1apVU0SEW3/+c4+zDi9ERUWd3b8zRlkMw8h3Wd4H4e7du/XEE4+pS5eu6tfvMcXFxWnVqlUaNeo55ebmWrYbcdY2zCCSTH59PFOXLvdoz55fC9xetWrVNW/ex/k+16/fX9Sjx4MByxITk5SVdXZwO3DggBISEs9anreO5BtBss6j2r9/v380KS+Q1K59RcC6l19e+6zgGh8fr/j4eF166WW6/PLLdfvtt+j777+/oLMagwlWrVu3POfzzZs318SJk/N9LinJ97pr167tX+Z2u1WzZo2zXp9V/fr19d57H+rw4cPKzT2l+PgE9ez5J111VYPT2y28boX9XERERASMip05wheKSpUqqVq16tqxY/t5bwOAvQhHpUzt2ldo0aL/F7Bs7drvdccddwY8rlevniTfX8xffLFQCQkJiomJyXebq1Z9p9TU4WrTxjfRdM+ePTp48ECx9P/HH3+Qx+PRoEF/9Y8uffnlF8Wyr4JcyGG1hITEswJP48ZNdOTIYa1bt9Y/YX7t2u915MjhAsNJjRo1lJSUpG+/Xap69a6S5DsDcOXKlXriiQGSfKNylSpV1rZtWwPW3b59m667rnWBfcx7TadOhXbJtD179mjfvr3+SwB8//0aORwOXXrpZQWucyGH1a66qr7cbre2bduqpk2bne7zKe3evVvVqlUvtL+xsbGSfPVYv369+vXrLym4ugXzc3GuUbFQHDx4UL/9tocJ2kApQjgqoU4cOaSFb47XVdfepMQal2nXLqfWr1+v2bNnqV27GwPafvnll7rqqgZq2rSpFiz4p374YZ2ee26kJOn22/+gt9+erUGDnlTfvn9RlSpVtGfPHn399Vfq2fPPqlKlqmrVukQLFvxD9es30NGjR/Tqq69c0LVgzqVmzVrKzc3Ve++9q7Ztb9Dq1as0b94HxbKvglzIYbX81K5dW9dd11qjR4/U8OEjJElpaaN0/fVtA8686tz5TvXv/6Tat+8gwzD0pz89oJkz39Qll1yqSy65RDNnzlBkZKRuv72jJN/ITc+eD2ratNeVnJys5OR6+sc/PtXWrVv04ou+yenr1q3VunXr1LRpU8XGVtCuXTv1+uuTVbNmrbMOtxbG7Xbr2Wef0cCBT+nIkSN66aVxuvnmW875oX4hh9ViYmLUpUtXTZ06RVWqVFW1atX09ttvSVLA2ZbWukm+MB0fH6+qVatp8+af9NJL43XDDTf65wkFU7dgfi7yc+zYsYARoF27dmnjxg2qUCFO1apV07FjxzRt2hS1b3+zKlVK0u7du5WRMVEVK1bUjTd2KLBWAEoWwlEJFVEuSlUuS9aarz/Rocw9+vQVj6pWrap77umi3r0DzwDq27efvvjic40bN0aJiUkaM2as/5BCVFSUZsyYpYkT0/XXvw7SsWNHVblyZbVo0VLly/v+Yn7uuZFKSxulP/3pPlWtWk39+z+u9PRXzupTUahbt54GDRqst96apYyMiWratJn6939Szz47vFj2Z5cxY8bpxRfH6bHH+kryXQQyNXVoQJutW7fqyJEj/scPPthLJ06c0LhxY3TokO8ikFOmTA2Yw9W9ew/l5ORowoSXlJ2dreTkupoyZZp/Pla5cuX09ddfadq0KTp+/LiSkpJ03XWtNW7ciwGTf//wh9vUqdOd6tv393k0Z6pV6xK1b99Bjz/+mA4dylbr1m0CJskXhwEDBsnlcmrEiGE6efKkGjZspGnTZqhChd9HnM6sW2bmPr3yykvKyspSUlIl3XFHJz3ySJ+A7RZWt2B+LvKzfv0PevTR3y8P8MorL0mSOnW6UyNHpsnhcOinnzbrH//4TIcPH1ZSUiW1aNFC48a9FPC+AijZDDOYiRzIV2bm4QsaeWgx4T+SpAE31Sm07SPXXprv8mbNGud7lV4UzDqicTF89584cUI33ni9Jk6cXOB1n/Kuc/TeewUfJgvVxVbnC3Xs2PnfOcowpKSk2Av+nYRzo87FI++zMBjLn2p7QfvKew8Lw3WOgDJu5coVatHiGlsuiAkAZQGH1YAyrnXrNgGnugMAzo1wVMp999334e4CyoC+ff9yzvlIAHAx4bAaAACABeEIAADAgnAEAABgwZyjksSUXN7c018ap2+7dfoWDh5P4A288r4O4d5ZAACgcISjEuDVr35S5KkTev/doWqy56cL2pZpGIHByfrvzGV5j2UUvJ7ky2f5Pa/fl5nn2k9+61m2e+5957MdBT42891XPn22rhfhVESu9/f9n6vPOvd2z/naz+zvGW0C962C1yugfwXuW2cuU/77LvC9Ce57x8zbdn41NiSVj5R5PCff7Qa89rP6W/j7aZ6rz4W8hsB9n73toOseVL2U73bNM+oVEcJrMM+ol+GQFB8jV/Yx3/V3zrEd8zzqZX0N5rm+Lwur+zn3Xfj7cCH79j8GgkA4KiGSM7dfcDCSJMM0dT5XJ7sYf21EFN4ERYDrQtunYrg7UEoU+gdNfiEr749Ih6GEMwNoAQHYGvzz/SOsoH351/t9uyH/4WsJkgXv+9zrWdcp9I9Y/3r51yL/P2J9X0//2XfjbvN0O1OGVldP1vRrOss0wjP7h3BUQnxftY7SW/9JLXeskyFThmn+HlhMU81qVPg99OQFINOU79vpzGWnFwc8Nn8PTv7wdOY6Zj7bMX9f1//Ysm6++z77sVFof818+xyw7xD6m/f493UBwOdC/4hksm7RuiWfZbdv+kZrqiXr20sa294fiXBUchiGXmvzpwKfvtBLpl/0rIFJppISY5S579DpvBZEyAoqmCn/cBdskCx03wp6nfPf99m1CrW/eSHYME3FVYhSdvYxS5sgXoMlQBvntW/LuqHWvNB9q5B9m8Hv+8z9Wt6DUF+3IVNRkRE6fuyk8u4IddYfQ/n1v4D3N9/+n6NWZ247/31bHqvg7ea/b/2+rKDv6cL2nd8fZvl9P561jcD1XA5DuR5vgbUJfO8srzXY9zLf74kz11G+2zAKep2F7Lsk/hG5s0IlbUy6NGz7Jxzh4nDmULLLJUVE/P7LsRAX8quj5P3asYdhSEqK1SnuQ1XsDEOKSorVUWpdrAzDd1+ug2W5zsEG+nweG0EFs7P3c9vUpTJOB0nj9Hb2lY+X1+G06UWfjXAEAAB8LmDy+vnmxczy8ee5ZvHh0CkAAIAF4QgAAMCCcAQAAGBBOAIAALAgHAEAAFgQjgAAACwIRwAAABYXbTiaO3eu2rdvr0aNGqlz585asWJFuLsEAABKgIsyHC1YsEBjx45Vv379NH/+fF199dV65JFHtHv37nB3DQAAhNlFGY5mzZqlLl26qGvXrrriiis0fPhwVa1aVX/729/C3TUAABBmF93tQ3JycvTDDz/o0UcfDVjeunVrrVq1KqRtnecV1s+Lnfsq6/y3WKOmxYo624da24M6h9+F1j7Y9S+6cHTgwAF5PB4lJiYGLE9KStK+fftC2lZiYmxRdu2ckpLs29fFws7372JGne1Dre1BncPHrs/Ciy4c5THOiI+maZ61rDBZWfbdmTkz87A9O7oIGIbvl5ud79/FiDrbh1rbgzqH34V+Fua9h4W56MJRfHy8nE6nMjMzA5ZnZWUpKSkppG2Zpi7oB6SZpO+CbMcPYtG70PcPwaHO9qHW9qDO4WNX3S+6Cdlut1sNGjTQkiVLApZ/8803atq0qa19qRlTtO0AAMCFu+hGjiSpV69eevrpp9WwYUM1bdpU77//vn799Vd169bN1n5UqyTpSJDtAACALS7KcNSxY0cdOHBAU6ZM0d69e5WcnKzp06erRo0atvajRmWHtMUbXDsAAMogh6TCPwntPdR1UYYjSerevbu6d+8e1j44oqMVzNCRrx0AAGXPFZJ+CrKdXRiSCKMoV3DZNNh2AACUNsEes7Hz2A7hKIwSY4KbaR1sOwAASps6NYu2XVEgHIVRVFSUKhbSpuLpdgAAlEWVagQXRYJtVxQIR2EU6XbrkgjfxK+8N8J5+n+HfMsvifC1AwCgLKoQ5NGRYNsVBSazhFE5p1PxiVKtPb7HByVFSTou+UeU4hN97QAAKIuSIiMlHQqynT0IR+HkdKpKhQidyD0lOaTobCk2QTq8X4qJk+SVqlSIkAhHAIAyyh0bK2lvkO3sQTgKI6dhKL5CjHK9h5VruBRb7oSiY2MV4zqs6MhIucxcxVeIkZNbQAMAyqhoh0OXStp2jjaXnm5nF8JRGLldLiVER0teyeFw67A7QlHlo3XckGKjouX15ighOlpuTuUHAJRRUZGRio+Rso/4QkmWfp9ikigpV1J8jK+dXfjUDSOn06nKcVHyGKYMr1PREaZi4mN0xGnK5XLLdDhVuUKUnBxWAwCUUYbDoaSK0rFcqZxbijskxSRJRzKl6ArSyRwpqaKvnV0IR2HkdDhUqUK0crxenTjh1SmHW+XdEfK43YpwORUZGaFKFaLltPEbAgAAOzkkJVUsL4/3qHK8UkSEFFneIafpVVSE5I7xPc/tQy4STodDFaIilOOJkjfa0PETJxUdW05RHo+iIsvJ4TRVISqCcAQAKLNcERGqFBUtV6JDLodbR48eUrmYWJ2MOKzy5Sso15uj+MhIuSIi7OuTbXtCvqIjXIqPcsvjlWLKGapYMVoHTa9cTpecDt/zAACUVeWcTlWOi5YcpgxFyG2UV0xirI7Iq8hybpkyVDk22tbL2vDJG0amJLfLqQpRbp085ZXHI8WUc+uU2yWn06VyEQ5FOJ0yJXG+GgCgLIqIiFCthPI6ZXiVe9wjw4xQlMMhjztC5SMMuaKiVCu+vCIYObo4GJIchqFyLqcinE7l5hqKcruUUy5CLpdLDsPXimAEACirHIah+AqRis/JUY7hVflyDlVIiNYh7ym5nG65ox2KrxAph42XtSEchZFhGHI4DOV6TEmSw+GQy+WQ4/QcI68puZyGDK5zBAAoowzDUHREhKrERMsTaerQCZfKR0bIERutCpHl5HT5nrfzs5BwFGaGIXm8Xnm9UoRTinA55DCkU7leORy+xwAAlFVer1cypfioCOV6pKgIQ7EVo3XY61W5iAi5nJJMXzuHTScoEY7CzDR9Z605HZJpmvJ4TZkyFOEy/M8DAFBWeU//73L6UpDTcCk20i2P2yXD4ZDLafjb2TVcQDgKI9M05fWacjnyDp2Zcjsdcrsckgz/86bD5NAaAKBMcki+zzrTK6fDIa8cchiGHA6HHA5DHo9XhuGw9TpHHLMJo7xBobzgkzcHyfrY2g4AgLLGMAyZhpTj8crrNeV0GHK5HHI6DHm9pnI8XpmGbB0kIByFUd7bbBZw7CxvOWNGAICyypTkdPgGBzym6f/sM01THtOUw2HI6TBsHSjgsFoY5Y0Ueb2mnPkkIK+pgJEkAADKHNOUQ4aiIpzK9Uq5Hq9O5XqV65VcTod85yUZvkm4Nn0eEo7CzOkw/BOxnafH8XyPfd8DTgfBCABQhuVNKZFTbpch0/SqvNulU5EuGYbj9IlKsi0YSRxWCzuHYcjl9E0685qSx2v6R4xcToetF70CAMBuhvIGAnwDBYbhm3NkGIY8XlOSbx6SnZ+GjByVAA7DkMN59tlqAACUdcbpQYLc04+9p4+m5A0USL7nuQjkRcp6thrXNwIAXCx8U0x8n30Oh+R2OhThNOQ1jbBMMSEcAQCAsMqbYuLxmjJN3xlspgz/mWp2TzEhHAEAgLDLm2Jiml65DEMRTkOGEZ6p0YQjAAAQdnlzjUzTVK5p6pTHlGF4wzJyxNlqAAAgrLymqdzTV8h2nJ5j5DB8txXJ9XjltXkiLuEIAACEVd5cI+cZt9DyTdTW6VP67UM4AgAAYZN3k/WCTkjLG0Eq6FZbxYFwBAAAwubMm7CfKRw3YSccAQCAsCmJN2EnHAEAgLDx34S9gKGhcNyEnXAEAADCyjcRW/5T+SX5b8rOFbIBAMBFx3qF7DPvrcYVsgEAwEWpJN2EncNqAACgxLDehD1cCEcAAAAWhCMAAAALwhEAAIAF4QgAAMCCcAQAAGBBOAIAALAgHAEAAFgQjgAAACwIRwAAoMTwer3KzfXK6/WGrQ/cPgQAAIRdrternFyvPB6v3Dm5OnIiV87TtxFxOewdy2HkCAAAhFWu16sTObnKzfXKaZhyOQw5DVO5uaeX2zyKRDgCAABhlZPrVa7HlGFIuV4px+NVrle+xx5TObmEIwAAcJHwer3KOeWRTK+8XlMOQ3I6DDkMyes1JdP3vJ1zkAhHAAAgbLySPF5ThuGQ02H4lpmmJF9IMgyHPF5Tdo4dMSEbAACEjWGa8npNeeWVaTrkNX2H0XJyvXIYhkzTK6/XkHE6MNmBkSMAABA2hsMhp9PQSY8pzxmH1TxeUyc9ppxOQ4aNZ6wxcgQAAMLGkC8MyTTlMU05Tg8QmaYpjynJNH2H12zsE+EIAACEldPhUKTbJa8p5Xq8OpXrO1vN5XTI4XLIafN1jghHAAAgbEzlHUZzyjQlw2WovNulnHJOmXLIMCTDMGRKto0eEY4AAEDYGJIchiEZkteUTNMhw2FIhkMOwzf3SOKwGgAAuEgYhiGHw5DXK0U4DUmm3KdvGyIZvknaDkOGYV884mw1AAAQVr7rGfnOTpMkx+nrHfmufyT/9Y/swsgRAAAIK4dhyOU8fbFH0zz9vy8k+eYjEY4AAMBFxmEYcuRzWC0sfQnLXgEAAPKRNwfJzjlGZyIcAQAAWBCOAAAALEpFONq5c6eGDRum9u3bq3Hjxrrppps0ceJE5eTkBLTbvXu3+vbtq5SUFLVs2VJpaWlntdm4caMeeOABNW7cWNdff70yMjJk2ngzOwAAULKVignZv/zyi0zT1KhRo3TppZdq06ZNGjFihI4fP64hQ4ZIkjwej/r06aP4+Hi9++67OnjwoIYMGSLTNDVixAhJ0pEjR9S7d2+1bNlS8+bN09atW5Wamqro6Gj17t07nC8RAACUEKUiHLVt21Zt27b1P65Vq5a2bNmiv/3tb/5wtHjxYm3evFmLFi1SlSpVJEmpqalKTU3VwIEDFRMTo08//VQnT57UuHHj5Ha7lZycrK1bt2rWrFnq1atXWCd/AQCAkqFUhKP8HD58WHFxcf7Hq1evVp06dfzBSJLatGmjnJwcrVu3Tq1atdLq1avVokULud3ugDYTJkzQzp07VatWrZD6UPRZypTXa0oyCWrFKK+0lLh4UWf7UGt7UGd7FGedg91mqQxH27dv1zvvvKPU1FT/sszMTCUlJQW0i4uLU0REhDIzM/1tatSoEdAmMTHR/1yo4SgxMfZ8un8Wr9eUx+uVx5RyPF7FxJWX0/Ddpdhh81VBLyZF9f7h3Kizfai1PaizPcJZ57CGo0mTJikjI+OcbebNm6dGjRr5H//22296+OGHddttt6lr164BbQsabbEuD6ZNsLKyDutC53J7TVO5Hq9MU3I4pKSkCsrMPCSv15dwXU6H7VcGLesMw/dDVxTvHwpGne1Dre1Bne1RnHXO23ZhwhqOunfvro4dO56zTc2aNf1f//bbb+rZs6dSUlI0evTogHZJSUlas2ZNwLLs7GydOnXKPzqUlJSkffv2BbTJysqS9PsIUihMUxf8xuV6THm9p+8rc3qZId9diD1eU7kyT9+ID0WtKN4/FI4624da24M62yOcdQ5rOEpISFBCQkJQbfOCUYMGDTR27Fg5HIFXIUhJSdHUqVO1d+9eVa5cWZK0ZMkSud1uNWzY0N8mPT1dOTk5/nlHixcvVuXKlQNCmF1M0zfHqKAjZw7Dd8jNdDAHCQAAu5SK6xz99ttv6tGjh6pWraohQ4Zo//792rdvX8AoUJs2bXTllVfq6aef1vr167V06VKNHz9e9913n2JiYiRJnTp1ktvt1tChQ7Vp0yZ9+eWXmjZtWtjOVMsLxIUd6uMPFAAA7FMqJmQvWbJE27Zt07Zt2wJO6Zd8F3WUJKfTqWnTpmnkyJG6//77FRkZqTvuuMN/qr8kxcbGaubMmRo1apS6dOmiuLg49erVS7169bL19eTJi0Smmf/IUN7FKRkzAgDAPobJ5aHPW2bmhU8WO+Xxyus1fXOODCkxKVZZp7fr8ZpyOAxFOEvFAF+pYRhSUlJskbx/KBh1tg+1tgd1tkdx1jlv24XhUzfM8kKRx2v6R4pM05THa8owfM8DAAD7lIrDamWZwzDkcjrk8Zryng5FXlNyOAw5HQan8QMAYDPCUQngMAw5nIYkU26nQ26XQ8w0AgAgPEIKR7/88ov++c9/asWKFdq1a5dOnDih+Ph41a9fX23atNGtt94acGsOhMYwDDkchgzD4Hg2AABhElQ4Wr9+vV566SWtWLFCTZs2VePGjXXTTTcpMjJSBw8e1E8//aT09HSlpaXpoYce0p///GdCEgAAKJWCCkePPfaYHnroIaWnp6tixYoFtlu1apXeeustzZw5U3379i2qPgIAANgmqHC0cOHCoEaCmjZtqqZNmyonJ+eCOwYAABAOQZ3KH+ohMg6pAQCA0uq8r3O0d+9ePfHEE2rVqpWuueYa9e3bVzt27CjKvgEAANjuvMPRsGHDVKdOHb3zzjuaPXu2EhMT9dRTTxVl3wAAAGwXdDhKS0vTsWPH/I+3b9+uRx55RFdeeaWuuuoq9ezZU1u2bCmWTgIAANgl6OscValSRZ07d9bgwYPVoUMH3X777eratavatWun3NxcffHFF+rUqVNx9hUAAKDYhXTj2R07dmjkyJGKjIzUM888o/Xr12vZsmXyeDxq1qyZbrvttnzvLl9WFf1N8UwlJsYqK+uwuEJ28eHmkfagzvah1vagzvYoCTeeDekK2bVq1dKMGTP0ySefqEePHurZs6eGDBlyUQWi4pB3TzXTNJXj8Son1yvD4N5qAACEQ8gTsg8cOKC77rpL8+bN0/r16/XHP/5RGzZsKI6+XRS8pqlcj1derymHodOBSPJ6Ty/nzxMAAGwVdDj69ttv1bp1a1177bVq27atfvnlF40dO1YDBw7UoEGD9OKLL+rEiRPF2dcyyTdi5AtFeSNweaNGpul7HgAA2CfocDRy5Eg99NBDWrNmjZ599lm98MILkqRrr71W8+fPl8vl0l133VVsHS2LTNP0jxjlJ28EKYRpYQAA4AIFHY727t2rdu3aqVy5crr++uu1f/9+/3Nut1uDBg1SRkZGsXSyrMqLPAXN2cpbTjQCAMA+QU/Ibt++vZ588km1b99eK1euVLt27c5qU6dOnSLtXFmXF4lM08w3IOWNGDElGwAA+wQdjsaMGaP3339fv/zyi+6880516dKlOPt1UTAMQw6HIa/XlDOfBOQ1JYdlLhIAACh+QYcjt9utHj16FGdfLkq+ide+U/mdpw9y+h77rsfgLGhCEgAAKBZBzTlatWpV0Bs8duyYfvrpp/Pu0MXGYRhyOR2+EaTTZ6fljRi5nA6ucwQAgM2CCkdDhgxRr169tGDBAh09ejTfNps3b9Yrr7yiW265RT/88EORdrKscxiGIpwOuV0OuU//H0EwAgAgLII6rPbPf/5T77//viZOnKjBgwfrsssuU+XKlVWuXDllZ2frl19+0fHjx3XzzTdr5syZSk5OLu5+l0l5c5AMw+DS9AAAhElI91aTpB9++EErVqzQrl27dOLECcXHx6t+/fpq2bKlKlasWEzdLJmK+r4v3LfHHtTZHtTZPtTaHtTZHqXu3mqS1KBBAzVo0OC8OgUAAFDShXxvNQAAgLKMcAQAAGBBOAIAALAgHAEAAFgQjgAAACxCPltNkpYuXaqlS5cqKytLXq834LmxY8cWSccAAADCIeRwlJGRocmTJ6thw4aqVKkSN0UFAABlSsjh6L333tPYsWN19913F0N3AAAAwivkOUenTp1Ss2bNiqMvAAAAYRdyOLr33nv12WefFUdfAAAAwi7kw2onT57UBx98oKVLl6pu3bpyuQI3MXTo0CLrHAAAgN1CDkcbN25UvXr1JEmbNm0KeI7J2QAAoLQLORzNmTOnOPoBAABQIlzQRSD37Nmj3377raj6AgAAEHYhjxx5vV5NmTJFs2bN0rFjxyRJ5cuXV69evdSvXz85HFx0GwAAlF4hh6P09HTNmzdPTz31lP+U/pUrVyojI0M5OTkaOHBgkXcSAADALiGHo48//lhpaWnq0KGDf1m9evVUpUoVjRw5knAEAABKtZCPgWVnZ6t27dpnLa9du7ays7OLpFMAAADhEnI4qlevnubOnXvW8rlz5/pP8QcAACitQj6sNnjwYPXp00fffPONUlJSZBiGVq1apV9//VVvvPFGcfQRAADANiGPHF1zzTX6/PPPdfPNN+vw4cPKzs7WzTffrM8//1zNmzcvjj4CAADYJuSRI0mqUqUKE68BAECZFFQ42rBhg5KTk+VwOLRhw4ZztmXeEQAAKM2CCkd33323lixZosTERN19990yDEOmaZ7VzjAM/fjjj0XeSQAAALsEFY7+9a9/KSEhwf81AABAWRVUOKpRo4b/6927d6tp06ZyuQJXzc3N1apVqwLaAgAAlDYhn63Ws2fPfC/2ePjwYfXs2bNIOgUAABAuIYcj0zRlGMZZyw8ePKioqKgi6RQAAEC4BH0qf//+/SX5Jl2npqbK7Xb7n/N4PNq4caOaNm1a9D0EAACwUdDhKDY2VpJv5Kh8+fKKjIz0PxcREaGUlBR17dq16HsIAABgo6DD0dixYyX5Jmf37t1b0dHRxdYpAACAcAn5Ctl5h9cAAADKoqDC0T333KO33npLcXFx/otAFuTjjz8uss4BAADYLahw1KFDB/8E7JtuuqlYOwQAABBOQYUj66E0DqsBAICyLOQ5R3nWrl2rn3/+WYZh6IorrlDDhg2Lsl8AAABhEXI42rNnjwYNGqTvvvtOFSpUkCQdOnRITZs21SuvvKJq1aoVeScBAADsEvIVsocNG6bc3FwtWLBAy5Yt07Jly7RgwQKZpqnhw4cXRx8BAABsE/LI0YoVK/Tee++pdu3a/mW1a9fWiBEjdP/99xdp5wAAAOwW8shRtWrVlJube9Zyj8ejKlWqFEmnziUnJ0d33XWX6tatqx9//DHgud27d6tv375KSUlRy5YtlZaWppycnIA2Gzdu1AMPPKDGjRvr+uuvV0ZGhkzTLPZ+AwCA0iHkcDR48GCNHj1aa9eu9YeKtWvXasyYMRoyZEiRd/BML774oipXrnzWco/Hoz59+ujYsWN69913lZ6eroULF2r8+PH+NkeOHFHv3r1VuXJlzZs3TyNGjNDMmTM1a9asYu83AAAoHYI6rNaiRYuACz8eO3ZM9913n5xOpyRfMHE6nRo2bFixXgfp3//+t5YsWaJJkybpP//5T8Bzixcv1ubNm7Vo0SL/CFZqaqpSU1M1cOBAxcTE6NNPP9XJkyc1btw4ud1uJScna+vWrZo1a5Z69ep1zotbAgCAi0NQ4WjYsGHF3Y9CZWZmasSIEZo8eXLATW/zrF69WnXq1Ak4tNemTRvl5ORo3bp1atWqlVavXq0WLVr4L2iZ12bChAnauXOnatWqFVKfijpL5W2PjFa8qLM9qLN9qLU9qLM9irPOwW4z6NuHhJNpmkpNTVW3bt3UqFEj7dy586w2mZmZSkpKClgWFxeniIgIZWZm+tvUqFEjoE1iYqL/uVDDUWJibEjtw71dBKLO9qDO9qHW9qDO9ghnnYMKR0eOHFFMTIz/63PJaxeMSZMmKSMj45xt5s2bp1WrVunIkSPq06fPOdsWdFjMujyYNsHKyjqsopzLbRi+b4ai3i4CUWd7UGf7UGt7UGd7FGed87ZdmKDnHC1evFiJiYlq3rx5vkHCNE0ZhnHWGWTn0r17d3Xs2PGcbWrWrKnXX39da9asUaNGjQKe69Klizp16qTx48crKSlJa9asCXg+Oztbp06d8o8OJSUlad++fQFtsrKyJP0+ghQK01Sx/IAU13YRiDrbgzrbh1rbgzrbI5x1DioczZ49W3FxcZKkt99+u8h2npCQoISEhELbPfPMMxowYID/8d69e/XQQw8pPT1dTZo0kSSlpKRo6tSp2rt3r/9stiVLlsjtdvtvbZKSkqL09HTl5OT45x0tXrxYlStXVs2aNYvsdQEAgNIrqHB0zTXX5Pu1XapXrx7wODo6WpJ0ySWXqGrVqpJ8E6uvvPJKPf3003r66aeVnZ2t8ePH67777vMf6uvUqZMmT56soUOHqk+fPtq2bZumTZumxx57jDPVAACApPO88eyhQ4f0/fffKysr66wLKN59991F0a+QOZ1OTZs2TSNHjtT999+vyMhI3XHHHQHXXoqNjdXMmTM1atQodenSRXFxcerVq5d69eoVlj4DAICSxzBDvDz0119/rb/+9a86fvy4ypcvf9Zk52XLlhV5J0uqzMyin5CdlBRb5NtFIOpsD+psH2ptD+psj+Ksc962CxPyyNH48ePVpUsXDRo0SFFRUefVOQAAgJIq5NuH/Pbbb+rZsyfBCAAAlEkhh6M2bdpo7dq1xdEXAACAsAvqsNq//vUv/9ft2rXTSy+9pJ9//lnJyclyuQI30aFDh6LtIQAAgI2CCkePPfbYWcsmT5581rJQLwIJAABQ0gQVjjZs2FDc/QAAACgRQp5zBAAAUJad10Ugly5dqqVLlyorK0terzfgubFjxxZJxwAAAMIh5HCUkZGhyZMnq2HDhqpUqRK33QAAAGVKyOHovffe09ixY8N2mxAAAIDiFPKco1OnTqlZs2bF0RcAAICwCzkc3Xvvvfrss8+Koy8AAABhF/JhtZMnT+qDDz7Q0qVLVbdu3bMuAjl06NAi6xwAAIDdQg5HGzduVL169SRJmzZtCniOydkAAKC0CzkczZkzpzj6AQAAUCJwEUgAAACLoEeO+vfvH1S7jIyM8+4MAABAuAUdjmJjY4uzHwAAACVC0OGI24IAAICLAXOOAAAALAhHAAAAFoQjAAAAC8IRAACABeEIAADAgnAEAABgQTgCAACwIBwBAABYEI4AAAAsCEcAAAAWhCMAAAALwhEAAIAF4QgAAMCCcAQAAGBBOAIAALAgHAEAAFgQjgAAACwIRwAAABaEIwAAAAvCEQAAgAXhCAAAwIJwBAAAYEE4AgAAsCAcAQAAWBCOAAAALAhHAAAAFoQjAAAAC8IRAACABeEIAADAgnAEAABgQTgCAACwIBwBAABYEI4AAAAsCEcAAAAWhCMAAAALwhEAAIAF4QgAAMCCcAQAAGBBOAIAALAgHAEAAFgQjgAAACwIRwAAABaEIwAAAAvCEQAAgAXhCAAAwIJwBAAAYFGqwtGiRYvUtWtXNW7cWC1btlT//v0Dnt+9e7f69u2rlJQUtWzZUmlpacrJyQlos3HjRj3wwANq3Lixrr/+emVkZMg0TTtfBgAAKMFc4e5AsBYuXKgRI0Zo4MCBatWqlUzT1KZNm/zPezwe9enTR/Hx8Xr33Xd18OBBDRkyRKZpasSIEZKkI0eOqHfv3mrZsqXmzZunrVu3KjU1VdHR0erdu3e4XhoAAChBSkU4ys3N1ZgxYzR48GB17drVv7x27dr+rxcvXqzNmzdr0aJFqlKliiQpNTVVqampGjhwoGJiYvTpp5/q5MmTGjdunNxut5KTk7V161bNmjVLvXr1kmEYIfUrxOZBb6+ot4tA1Nke1Nk+1Noe1NkexVnnYLdZKsLR+vXr9dtvv8nhcOjuu+9WZmam6tWrpyFDhqhOnTqSpNWrV6tOnTr+YCRJbdq0UU5OjtatW6dWrVpp9erVatGihdxud0CbCRMmaOfOnapVq1ZI/UpMjC2aF2jTdhGIOtuDOtuHWtuDOtsjnHUuFeFox44dkqSMjAylpqaqRo0amjVrlh544AEtXLhQFStWVGZmppKSkgLWi4uLU0REhDIzMyVJmZmZqlGjRkCbxMRE/3OhhqOsrMMqyulKhuH7Zijq7SIQdbYHdbYPtbYHdbZHcdY5b9uFCWs4mjRpkjIyMs7ZZt68efJ6vZKkvn376tZbb5UkjR07Vm3bttXnn3+ubt26SVKBh8Wsy4NpEyzTVLH8gBTXdhGIOtuDOtuHWtuDOtsjnHUOazjq3r27OnbseM42NWvW1NGjRyVJV1xxhX+52+1WrVq19Ouvv0qSkpKStGbNmoB1s7OzderUKf/oUFJSkvbt2xfQJisrS9LvI0gAAODiFtZwlJCQoISEhELbNWzYUG63W1u2bFHz5s0lSadOndKuXbtUvXp1SVJKSoqmTp2qvXv3qnLlypKkJUuWyO12q2HDhv426enpysnJ8c87Wrx4sSpXrqyaNWsWx0sEAAClTKm4zlFMTIy6deumSZMmafHixfrll1/0/PPPS5Juu+02Sb6J1VdeeaWefvpprV+/XkuXLtX48eN13333KSYmRpLUqVMnud1uDR06VJs2bdKXX36padOmndeZagAAoGwqFROyJenpp5+Wy+XS008/rRMnTqhJkyaaPXu24uLiJElOp1PTpk3TyJEjdf/99ysyMlJ33HGHhgwZ4t9GbGysZs6cqVGjRqlLly6Ki4tTr1691KtXr3C9LAAAUMIYJpeHPm+ZmUV/tlpSUmyRbxeBqLM9qLN9qLU9qLM9irPOedsuTKk4rAYAAGAXwhEAAIAF4QgAAMCCcAQAAGBBOAIAALAgHAEAAFgQjgAAACwIRwAAABaEIwAAAAvCEQAAgAXhCAAAwIJwBAAAYEE4AgAAsCAcAQAAWBCOAAAALAhHAAAAFoQjAAAAC8IRAACABeEIAADAgnAEAABgQTgCAACwIBwBAABYEI4AAAAsCEcAAAAWhCMAAAALwhEAAIAF4QgAAMCCcAQAAGBBOAIAALAgHAEAAFgQjgAAACwIRwAAABaEIwAAAAvCEQAAgAXhCAAAwIJwBAAAYEE4AgAAsCAcAQAAWBCOAAAALAhHAAAAFoQjAAAAC8IRAACABeEIAADAgnAEAABgQTgCAACwIBwBAABYEI4AAAAsCEcAAAAWhCMAAAALwhEAAIAF4QgAAMCCcAQAAGBBOAIAALAgHAEAAFgQjgAAACwIRwAAABaEIwAAAAvCEQAAgAXhCAAAlBimacrrNWWaZtj64ArbngEAAE7zmqY8p0NRjsernFyvDMOQ02HIYRi29oWRIwAAEFZe01Suxyuv15TD0OlAJHm9p5fbPIpEOAIAAGHlGzHyhSLj9ChR3qiRafqetxPhCAAAhE3eHCNHAUfO8kaQ7JyDRDgCAABhkxd5jALmFeUtt3PsiHAEAADCJi8SFTQylLfczinZpSYcbdmyRf369VPLli3VrFkzdevWTd9++21Am927d6tv375KSUlRy5YtlZaWppycnIA2Gzdu1AMPPKDGjRvr+uuvV0ZGRlhPFwQA4GJmGIYcDkMFTSvympLDMhfJDqUmHPXp00cej0ezZ8/W3//+d1111VXq27ev9u3bJ0nyeDzq06ePjh07pnfffVfp6elauHChxo8f79/GkSNH1Lt3b1WuXFnz5s3TiBEjNHPmTM2aNStcLwsAgIuebyK2/KfyS74RI4/XlHH67DU7lYrrHO3fv1/btm3TCy+8oHr16kmSnnrqKb377rvavHmzKlWqpMWLF2vz5s1atGiRqlSpIklKTU1VamqqBg4cqJiYGH366ac6efKkxo0bJ7fbreTkZG3dulWzZs1Sr169Qk6lRR1i87Zn8+UcLjrU2R7U2T7U2h7Uufg4DUOG4fCHI4/XlCnJ6Sza6xwFu5lSEY7i4+N1xRVXaP78+apfv77cbrfef/99JSUlqUGDBpKk1atXq06dOv5gJElt2rRRTk6O1q1bp1atWmn16tVq0aKF3G53QJsJEyZo586dqlWrVkj9SkyMLZoXaNN2EYg624M624da24M6Fy/T9J3WX61KnK2H0qxKRTgyDEOzZs1Sv3791KxZMzkcDiUmJmrGjBmqUKGCJCkzM1NJSUkB68XFxSkiIkKZmZn+NjVq1Ahok5iY6H8u1HCUlXVYRTldyTB8P3RFvV0Eos72oM72odb2oM72KM465227MGENR5MmTVJGRsY528ybN08NGzbU888/r8TERM2dO1eRkZH68MMP1adPH82bN0+VK1eWVPhpgMG2CZZpqlh+QIpruwhEne1Bne1Dre1Bne0RzjqHNRx1795dHTt2PGebmjVr6ttvv9WiRYu0fPlyxcTESJIaNGigb775RvPnz9ejjz6qpKQkrVmzJmDd7OxsnTp1yj86lJSU5J/AnScrK0vS7yNIAADg4hbWcJSQkKCEhIRC2x0/flzS2aM7hmHI6/VKklJSUjR16lTt3bvXP5K0ZMkSud1uNWzY0N8mPT1dOTk5/nlHixcvVuXKlVWzZs0ie10AAKD0KhWn8qekpKhChQpKTU3Vhg0btGXLFo0fP167du3SDTfcIMk3sfrKK6/U008/rfXr12vp0qUaP3687rvvPv9oU6dOneR2uzV06FBt2rRJX375paZNm3ZeZ6oBAICyyTBLyRUQ165dq1dffVXr1q3TqVOnVKdOHf3lL39Ru3bt/G12796tkSNH6ttvv1VkZKTuuOMODRkyJODstI0bN2rUqFH6/vvvFRcXp27duumxxx47r3CUmVn0E7KTkmKLfLsIRJ3tQZ3tQ63tQZ3tUZx1ztt2oe1KSzgqiQhHpRN1tgd1tg+1tgd1tkdJCEel4rAaAACAXQhHAAAAFoQjAAAAi1JxheySinurlU7U2R7U2T7U2h7U2R7FWedgt8mEbAAAAAsOqwEAAFgQjgAAACwIRwAAABaEIwAAAAvCEQAAgAXhCAAAwIJwBAAAYEE4AgAAsCAcAQAAWBCOAAAALAhHNps7d67at2+vRo0aqXPnzlqxYsU52y9btkydO3dWo0aN1KFDB/3tb3+zqaelWyh1/uKLL9SrVy+1atVKzZo10x//+Ef997//tbG3pVeo3895Vq5cqfr16+uuu+4q5h6WDaHWOScnR+np6brxxhvVsGFD3XTTTZo3b55NvS3dQq31p59+qjvvvFNNmjRRmzZtNHToUB04cMCm3pY+y5cvV9++fdWmTRvVrVtXX331VaHrhOVz0IRt/vnPf5oNGjQwP/jgA3Pz5s1mWlqamZKSYu7atSvf9tu3bzebNGlipqWlmZs3bzY/+OADs0GDBubnn39uc89Ll1DrnJaWZk6fPt1cs2aNuWXLFnPChAlmgwYNzB9++MHmnpcuodY5z6FDh8wOHTqYvXv3Nu+8806belt6nU+d+/bta3bt2tVcsmSJuWPHDnPNmjXmypUrbex16RRqrZcvX27Wq1fPnD17trl9+3Zz+fLl5h/+8AfzL3/5i809Lz0WLVpkvvLKK+bChQvN5ORk88svvzxn+3B9DhKObHTvvfeazz77bMCy2267zXz55Zfzbf/iiy+at912W8CyESNGmPfdd1+x9bEsCLXO+enYsaM5adKkou5amXK+dR4wYICZnp5uTpw4kXAUhFDr/O9//9u8+uqrzQMHDtjQu7Il1FrPmDHD7NChQ8Cyt99+22zbtm2x9bEsCSYchetzkMNqNsnJydEPP/ygNm3aBCxv3bq1Vq1ale86q1evVuvWrQOWXX/99Vq3bp1OnTpVbH0tzc6nzmfyer06evSoKlasWAw9LBvOt84fffSRtm/frv79+xd3F8uE86nz119/rYYNG2rGjBm6/vrrdeutt2r8+PE6ceKEHV0utc6n1k2bNtWePXv073//W6ZpKjMzUwsXLlS7du3s6PJFIVyfg65i2zICHDhwQB6PR4mJiQHLk5KStG/fvnzXyczMVFJSUsCyxMRE5ebm6sCBA6pcuXKx9be0Op86n2nmzJk6fvy4br/99uLoYplwPnXeunWrJkyYoLlz58rl4ldPMM6nzjt27NDKlStVrlw5TZ48WQcOHNDIkSN18OBBjR071o5ul0rnU+tmzZrp5Zdf1oABA5STk6Pc3Fy1b99eI0aMsKPLF4VwfQ4ycmQzwzACHpumedaywtrntxyBQq1znn/84x/KyMhQenr6Wb8kcbZg6+zxePTUU0/p8ccf1+WXX25X98qMUL6f8557+eWX1bhxY7Vr106pqan6+OOPGT0KQii13rx5s9LS0vTYY4/po48+0owZM7Rz504999xzdnT1ohGOz0H+fLNJfHy8nE6nMjMzA5ZnZWWdlYrz5PcXy/79++VyuTjkU4DzqXOeBQsWaPjw4Xrttdd03XXXFWc3S71Q63z06FGtW7dOP/74o0aPHi3Jd/jSNE3Vr19fb775pq699lpb+l6anM/3c6VKlVSlShXFxsb6l11xxRUyTVN79uzRZZddVpxdLrXOp9bTpk1Ts2bN9PDDD0uS6tWrp6ioKHXv3l0DBgxgdL8IhOtzkJEjm7jdbjVo0EBLliwJWP7NN9+oadOm+a6TkpKib775JmDZ4sWL1bBhQ0VERBRbX0uz86mz5BsxSk1N1YQJE3TDDTcUcy9Lv1DrHBMTo88++0zz58/3/+vWrZsuv/xyzZ8/X02aNLGr66XK+Xw/N2vWTHv37tXRo0f9y7Zs2SKHw6GqVasWa39Ls/Op9YkTJ+RwBH6MOp1OSb+PbuDChO1zsFineyNA3mmiH374obl582ZzzJgxZkpKirlz507TNE3z5ZdfNgcPHuxvn3cK4wsvvGBu3rzZ/PDDDzmVPwih1vmzzz4z69evb77zzjvm3r17/f8OHToUrpdQKoRa5zNxtlpwQq3zkSNHzLZt25qPP/64+dNPP5nLli0zb7nlFnP48OHhegmlRqi1/uijj8z69eubc+fONbdv326uWLHC7Ny5s3nvvfeG6yWUeEeOHDHXr19vrl+/3kxOTjZnzZplrl+/3n+5hJLyOchhNRt17NhRBw4c0JQpU7R3714lJydr+vTpqlGjhiRp3759+vXXX/3ta9WqpenTp2vs2LGaO3euKleurOHDh+vWW28N10soFUKt8/vvv6/c3FyNGjVKo0aN8i+/5557NG7cONv7X1qEWmecn1DrXL58ec2cOVNpaWnq0qWLKlasqNtvv10DBgwI0ysoPUKtdefOnXX06FHNnTtX48ePV2xsrFq1aqXBgweH6yWUeOvWrVPPnj39j/NOEsj7fVtSPgcN02TsDwAAIA9zjgAAACwIRwAAABaEIwAAAAvCEQAAgAXhCAAAwIJwBAAAYEE4AgAAsCAcAQAAWBCOAOAcevTooTFjxoS7GwBsxBWyAdgiKytLr732mv7zn/8oMzNTcXFxqlevnvr373/OmwKH28GDB+VyuRQTExNU+507d6pDhw6aP3++rrrqqmLuHYDiwL3VANji8ccfV25ursaNG6datWopKytLS5cuVXZ2dri7dk4VK1Yslu3m5OTI7XYXy7YBXBhGjgAUu0OHDqlFixaaM2eOrrnmmgLbHT58WC+++KK++uornTx5Ug0bNtSwYcNUr149SdKkSZP01VdfqVevXpo4caKys7PVtm1bjR492j+y8/nnn2vy5Mnatm2boqKidNVVV2nKlCmKjo6WJH300UeaMWOGdu7cqRo1aqhHjx7q3r17gX3q0aOH6tWrp+HDh0uS2rdvr/vuu0/btm3T559/rri4OPXr109//OMfJUl169YNWP+aa67RnDlzlJqaqkOHDqlJkyaaM2eO3G63vv76a23cuFFjxozR6tWrFRUVpVtuuUWpqakqX768JPnXu/rqqzVr1iydOnVKHTt21LBhwxQREXGe7wiAc2HOEYBiFx0drejoaH311VfKycnJt41pmnr00Ue1b98+TZ8+XX//+9/VoEEDPfjggzp48KC/3fbt2/Wvf/1LU6dO1bRp07R8+XK98cYbkqS9e/fqqaeeUpcuXbRgwQK9/fbbuvnmm5X3N+AHH3yg9PR0DRw4UAsWLNCgQYM0ceJEffzxxyG9nlmzZqlhw4aaP3++/vSnP+n555/Xzz//LEn68MMPJUlvvfWWFi9erEmTJvnXW7p0qX7++WfNmjVLU6dO1fHjx/Xwww8rLi5O8+bN06uvvqpvvvlGo0ePDtjf//73P23fvl2zZ8/WuHHj9PHHH4fcZwDBIxwBKHYul0vjxo3T/Pnz1bx5c3Xr1k2vvPKKNmzY4G/z7bffatOmTZo4caIaNWqkyy67TEOGDFGFChW0cOFCfzvTNDV27FglJyerefPmuvPOO7V06VJJ0r59+5Sbm6ubb75ZNWvWVN26ddW9e3f/KMyUKVOUmpqqW265RbVq1dItt9yiBx98UO+//35Ir6dt27bq3r27Lr30Uj3yyCOKj4/XsmXLJEkJCQmSfIfjKlWqFHBYLjo6WmlpaapTp46Sk5P12Wef6eTJkxo/frySk5N17bXX6tlnn9Unn3yizMxM/3pxcXF69tlndcUVV+jGG29Uu3bt/K8ZQNFjzhEAW9x666264YYbtGLFCq1atUqLFy/WjBkzlJaWps6dO+uHH37QsWPH1LJly4D1Tpw4oe3bt/sf16hRI2BydOXKlZWVlSVJqlevnq699lp16tRJbdq0UZs2bXTrrbcqLi5O+/fv16+//qrhw4drxIgR/vVzc3MVGxsb0muxHjozDENJSUn+PpxLcnJywDyjn3/+WXXr1vUf8pOkZs2ayev1asuWLUpKSpIkXXnllXI6nf42lSpV0qZNm0LqM4DgEY4A2KZcuXJq3bq1Wrdurf79+2v48OGaNGmSOnfuLK/Xq0qVKmnOnDlnrWcNLy7X2b+28g6bOZ1OzZo1S999952WLFmiOXPmKD09XR988IGioqIkSaNHj1aTJk0C1nc4QhtEP7MPhmEomOmbeX2w9tswjHzbWpef7/4AnB8OqwEImyuvvFLHjh2TJDVo0ECZmZlyOp269NJLA/7lHaoKhmEYuvrqq/XEE09o/vz5ioiI0FdffaWkpCRVqVJFO3bsOGv7tWrVKrLXlDdJ2uPxFNr2yiuv1IYNG/w1kKTvvvtODodDl112WZH1CUBoCEcAit2BAwfUs2dPffLJJ9qwYYN27Nih//u//9OMGTPUoUMHSdJ1112nlJQUPfbYY/rvf/+rnTt36rvvvlN6errWrl0b1H7WrFmjqVOnau3atdq9e7e++OIL7d+/X7Vr15bku5zA9OnTNXv2bG3ZskUbN27URx99pFmzZhXZa01MTFRkZKT++9//KjMzU4cPHy6wbadOneR2u5WamqpNmzbp22+/1ejRo3XXXXf5D6kBsB+H1QAUu/Lly6tJkyaaPXu2tm/frtzcXFWtWlVdu3ZV3759JflGfKZPn65XX31Vw4YN04EDB5SUlKTmzZsHHRRiYmK0fPlyzZ49W0eOHFH16tWVmpqqdu3aSZK6du2qyMhIvfnmm3rppZcUHR2t5ORkPfjgg0X2Wl0ul5555hlNnjxZEydOVPPmzfM9VCj5DrO9+eabGjNmjO69996AU/kBhA/XOQIAALDgsBoAAIAF4QgAAMCCcAQAAGBBOAIAALAgHAEAAFgQjgAAACwIRwAAABaEIwAAAAvCEQAAgAXhCAAAwIJwBAAAYPH/AbMIFp5cUTAAAAAAAElFTkSuQmCC"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "execution_count": 70
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-08-31T12:08:08.185292Z",
     "start_time": "2025-08-31T12:08:07.744635Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from scripts.Models.model_helpers import evaluate_top\n",
    "from sklearn.pipeline import make_pipeline\n",
    "from sklearn.preprocessing import PolynomialFeatures\n",
    "from sklearn.linear_model import LinearRegression, Ridge, Lasso, LassoCV\n",
    "from sklearn.neighbors import KNeighborsRegressor\n",
    "from sklearn.model_selection import train_test_split, cross_val_score\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from xgboost import XGBRanker\n",
    "from scipy.stats import pearsonr, spearmanr\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "cell_lines = ALL_CELL_LINES\n",
    "genes = only_cancer[CANONICAL_GENE].unique()\n",
    "\n",
    "seed = 42\n",
    "# model = RandomForestRegressor(n_jobs=-1)\n",
    "# model = XGBRanker(objective='rank:ndcg', ndcg_exp_gain=False, lambdarank_pair_method=\"topk\", seed=seed, n_jobs=-1)\n",
    "model = LinearRegression()\n",
    "# model = LassoCV(cv=5)  # alpha = regularization strength\n",
    "\n",
    "metric = 'correct_log_inhibition2'\n",
    "# metric = 'log_inhibition'\n",
    "# metric = INHIBITION\n",
    "\n",
    "df_copy = only_cancer.copy()\n",
    "gene = \"\"\n",
    "top = []\n",
    "spearman_dict = {}\n",
    "score_dict = {}\n",
    "for cell_line in cell_lines:\n",
    "    # for gene in genes:\n",
    "    if True:\n",
    "        # missing_cell_line = only_cancer[(only_cancer[CELL_LINE] == cell_line) & (only_cancer[CANONICAL_GENE] == gene)]\n",
    "        missing_cell_line = filtered_original[(filtered_original[CELL_LINE] == cell_line)]\n",
    "        if len(missing_cell_line) < 100:\n",
    "            continue\n",
    "\n",
    "        # filtered = only_cancer[(only_cancer[CELL_LINE] != cell_line) & (only_cancer[CELL_LINE] != gene)]\n",
    "        filtered = df_copy[(df_copy[CELL_LINE] != cell_line)].copy()\n",
    "\n",
    "\n",
    "        # mean_metric = INHIBITION\n",
    "\n",
    "        # keys = ['ISIS', VOLUME, TREATMENT_PERIOD, DENSITY_UPDATED, CHEMICAL_PATTERN]\n",
    "        #\n",
    "        # agg_spec = {mean_metric: 'mean'}\n",
    "        # agg_spec.update({c: 'first' for c in filtered.columns\n",
    "        #                  if c not in set(keys + [mean_metric])})\n",
    "        #\n",
    "        # filtered = filtered.groupby(keys, as_index=False).agg(agg_spec)\n",
    "\n",
    "\n",
    "        train, test = train_test_split(filtered, test_size=0.2, random_state=seed)\n",
    "        train['group_key'] = train[CELL_LINE].astype(str)\n",
    "        # train['group_key'] = train[CELL_LINE].astype(str) + '_' + train[CANONICAL_GENE].astype(str)\n",
    "        train = train.sort_values('group_key')\n",
    "        group_train = train.groupby('group_key').size().tolist()\n",
    "\n",
    "        X = train[features]\n",
    "        # y = train.groupby(CELL_LINE)[metric].rank(method='dense', ascending=True).astype(int)\n",
    "        y = train[metric]\n",
    "\n",
    "        # model = RandomForestRegressor(\n",
    "        #     n_estimators=600,      # more trees → stabler\n",
    "        #     max_depth=8,          # limit depth\n",
    "        #     min_samples_leaf=15,   # larger leaves\n",
    "        #     min_samples_split=20,  # stronger split guard\n",
    "        #     max_features=0.3,      # more randomness per split\n",
    "        #     bootstrap=True,\n",
    "        #     max_samples=0.6,       # subsample rows (bagging)\n",
    "        #     oob_score=True,        # quick gen-error check\n",
    "        #     n_jobs=-1,\n",
    "        #     random_state=seed\n",
    "        # )\n",
    "        # model = XGBRanker(objective='rank:pairwise')\n",
    "        # model.fit(X.to_numpy(), y.to_numpy(), group=group_train)\n",
    "        model.fit(X.to_numpy(), y.to_numpy())\n",
    "\n",
    "        predicted_unseen = model.predict(missing_cell_line[features].values)\n",
    "        true_unseen = missing_cell_line[metric]\n",
    "\n",
    "        predicted_seen = model.predict(filtered[features].values)\n",
    "        true_seen = filtered[metric]\n",
    "\n",
    "        corr, _ = pearsonr(predicted_unseen, true_unseen)\n",
    "        corrs, _ = spearmanr(predicted_unseen, true_unseen)\n",
    "        corr2, _ = pearsonr(predicted_seen, true_seen)\n",
    "        corrs2, _ = spearmanr(predicted_seen, true_seen)\n",
    "\n",
    "        print(\n",
    "            f\"Length {cell_line:>10}, {gene:>10}:  {len(missing_cell_line):<6d} Pearson unseen: {corr:.3f} seen: {corr2:.3f}\"\n",
    "            f\" Spearman unseen: {corrs:.3f} Spearman seen {corrs2:.3f}\")\n",
    "\n",
    "        score = model.score(missing_cell_line[features].to_numpy(), missing_cell_line[metric].to_numpy())\n",
    "        score_dict[cell_line] = score\n",
    "        spearman_dict[cell_line] = corrs\n",
    "        top.append(evaluate_top(model, missing_cell_line, metric=metric, log_correction=log_correction, top_k=50,\n",
    "                                features=features))\n",
    "        # plt.scatter(predicted_unseen, true_unseen)\n",
    "        # plt.show()\n",
    "\n",
    "print(top)\n",
    "train, test = train_test_split(only_cancer, test_size=0.2, random_state=seed)\n",
    "train['group_key'] = train[CELL_LINE].astype(str)\n",
    "train = train.sort_values('group_key')\n",
    "group_train = train.groupby('group_key').size().tolist()\n",
    "X = train[features]\n",
    "y = train[metric]\n",
    "# model.fit(X.to_numpy(), y.to_numpy(), group=group_train)\n",
    "model.fit(X.to_numpy(), y.to_numpy())\n",
    "\n",
    "predicted = model.predict(only_cancer[features].values)\n",
    "true = only_cancer[metric]\n",
    "\n",
    "corrp, _ = pearsonr(predicted, true)\n",
    "corrs, _ = spearmanr(predicted, true)\n",
    "\n",
    "print(f\"Total correlation: P:{corrp:.3f} S:{corrs:.3f}\")\n",
    "print(\"Spearman dict: \", spearman_dict)\n",
    "print(\"Score dict: \", score_dict)"
   ],
   "id": "155d139c6fd0b31",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Length       A431,           :  6541   Pearson unseen: 0.241 seen: 0.486 Spearman unseen: 0.284 Spearman seen 0.492\n",
      "Length KARPAS-229,           :  972    Pearson unseen: 0.379 seen: 0.424 Spearman unseen: 0.422 Spearman seen 0.451\n",
      "Length      A-431,           :  126    Pearson unseen: -0.398 seen: 0.420 Spearman unseen: -0.366 Spearman seen 0.443\n",
      "Length      MM.1R,           :  3394   Pearson unseen: 0.385 seen: 0.466 Spearman unseen: 0.471 Spearman seen 0.468\n",
      "Length  SK-MEL-28,           :  2836   Pearson unseen: 0.349 seen: 0.423 Spearman unseen: 0.345 Spearman seen 0.451\n",
      "Length    SNU-449,           :  1119   Pearson unseen: 0.137 seen: 0.420 Spearman unseen: 0.181 Spearman seen 0.425\n",
      "Length     HepaRG,           :  6608   Pearson unseen: 0.049 seen: 0.420 Spearman unseen: 0.064 Spearman seen 0.443\n",
      "Length       U251,           :  133    Pearson unseen: 0.249 seen: 0.428 Spearman unseen: 0.278 Spearman seen 0.445\n",
      "Length      HepG2,           :  4129   Pearson unseen: 0.048 seen: 0.420 Spearman unseen: 0.023 Spearman seen 0.443\n",
      "[(53.78, 99.2, 50.06), (52.86, 81.6, 28.84), (23.54, 72.22, 39.7), (48.74, 92.82, 24.6), (58.84, 88.86, 32.9), (54.72, 94.34, 50.52), (49.36, 95.64, 47.1), (77.24, 91.632, 75.2), (53.96, 96.7, 59.92)]\n",
      "Total correlation: P:0.420 S:0.443\n",
      "Spearman dict:  {'A431': 0.2836989478271723, 'KARPAS-229': 0.4222986589230671, 'A-431': -0.36603819760335377, 'MM.1R': 0.4706590710914889, 'SK-MEL-28': 0.3449163217204732, 'SNU-449': 0.1806786308005665, 'HepaRG': 0.06391643765913771, 'U251': 0.27762427139424406, 'HepG2': 0.023496466785738288}\n",
      "Score dict:  {'A431': -0.10410008595726428, 'KARPAS-229': -1.2109533513688189, 'A-431': -0.7978527831123214, 'MM.1R': -0.5872555199978389, 'SK-MEL-28': -0.3464765174423057, 'SNU-449': -0.04918305614394791, 'HepaRG': -0.21288601169461363, 'U251': -1.5078999732594562, 'HepG2': -1.05252765961645}\n"
     ]
    }
   ],
   "execution_count": 15
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "filtered_original = filtered\n",
    "\n",
    "print(filtered_original[CELL_LINE].unique())\n",
    "# only_cancer = filtered_original\n",
    "only_cancer = filtered_original[filtered_original[CELL_LINE] != 'A-431']\n",
    "\n",
    "# only_cancer = filtered[(filtered_original[CELL_LINE] == 'HepG2')\n",
    "#                        | (filtered_original[CELL_LINE] == 'HepaRG')\n",
    "#                           ]"
   ],
   "id": "e8f8103037d0f36f",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from sklearn.pipeline import make_pipeline\n",
    "from sklearn.preprocessing import PolynomialFeatures\n",
    "from sklearn.linear_model import LinearRegression, Ridge\n",
    "from sklearn.neighbors import KNeighborsRegressor\n",
    "from scripts.Models.model_helpers import evaluate_top\n",
    "from sklearn.model_selection import train_test_split, cross_val_score\n",
    "from sklearn.ensemble import RandomForestRegressor\n",
    "from xgboost import XGBRanker\n",
    "from scipy.stats import pearsonr, spearmanr\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "cell_lines = only_cancer[CELL_LINE].unique()\n",
    "genes = only_cancer[CANONICAL_GENE].unique()\n",
    "\n",
    "seed = 42\n",
    "# model = RandomForestRegressor(\n",
    "#     n_estimators=600,      # more trees → stabler\n",
    "#     max_depth=8,          # limit depth\n",
    "#     min_samples_leaf=15,   # larger leaves\n",
    "#     min_samples_split=20,  # stronger split guard\n",
    "#     max_features=0.3,      # more randomness per split\n",
    "#     bootstrap=True,\n",
    "#     max_samples=0.6,       # subsample rows (bagging)\n",
    "#     oob_score=True,        # quick gen-error check\n",
    "#     n_jobs=-1,\n",
    "#     random_state=seed\n",
    "# )\n",
    "# model = XGBRanker(objective='rank:ndcg', ndcg_exp_gain=False, lambdarank_pair_method=\"topk\", seed=seed, n_jobs=-1)\n",
    "model = LinearRegression()\n",
    "\n",
    "metric = 'correct_log_inhibition'\n",
    "\n",
    "gene = ''\n",
    "top = []\n",
    "for cell_line in cell_lines:\n",
    "    # for gene in genes:\n",
    "    if True:\n",
    "        # missing_cell_line = only_cancer[(only_cancer[CELL_LINE] == cell_line) & (only_cancer[CANONICAL_GENE] == gene)]\n",
    "        missing_cell_line = only_cancer[(only_cancer[CELL_LINE] == cell_line)]\n",
    "        if len(missing_cell_line) < 100:\n",
    "            continue\n",
    "\n",
    "        # filtered = only_cancer[(only_cancer[CELL_LINE] != cell_line) & (only_cancer[CELL_LINE] != gene)]\n",
    "        filtered = only_cancer[(only_cancer[CELL_LINE] != cell_line)]\n",
    "        train, test = train_test_split(filtered, test_size=0.2, random_state=seed)\n",
    "        train['group_key'] = train[CELL_LINE].astype(str)\n",
    "        # train['group_key'] = train[CELL_LINE].astype(str) + '_' + train[CANONICAL_GENE].astype(str)\n",
    "        train = train.sort_values('group_key')\n",
    "        group_train = train.groupby('group_key').size().tolist()\n",
    "\n",
    "        # metric = 'log_inhibition'\n",
    "        # metric = INHIBITION\n",
    "\n",
    "        X = train[features]\n",
    "        # y = train.groupby(CELL_LINE)[metric].rank(method='dense', ascending=True).astype(int)\n",
    "        y = train[metric]\n",
    "\n",
    "        # model = RandomForestRegressor(\n",
    "        #     n_estimators=600,      # more trees → stabler\n",
    "        #     max_depth=8,          # limit depth\n",
    "        #     min_samples_leaf=15,   # larger leaves\n",
    "        #     min_samples_split=20,  # stronger split guard\n",
    "        #     max_features=0.3,      # more randomness per split\n",
    "        #     bootstrap=True,\n",
    "        #     max_samples=0.6,       # subsample rows (bagging)\n",
    "        #     oob_score=True,        # quick gen-error check\n",
    "        #     n_jobs=-1,\n",
    "        #     random_state=seed\n",
    "        # )\n",
    "        # model = XGBRanker(objective='rank:pairwise')\n",
    "        # model.fit(X.to_numpy(), y.to_numpy(), group=group_train)\n",
    "        model.fit(X.to_numpy(), y.to_numpy())\n",
    "\n",
    "        predicted_unseen = model.predict(missing_cell_line[features].values)\n",
    "        true_unseen = missing_cell_line[metric]\n",
    "\n",
    "        predicted_seen = model.predict(filtered[features].values)\n",
    "        true_seen = filtered[metric]\n",
    "\n",
    "        top.append(evaluate_top(model, missing_cell_line, metric=metric, log_correction=log_correction, top_k=50,\n",
    "                                features=features))\n",
    "\n",
    "        corr, _ = pearsonr(predicted_unseen, true_unseen)\n",
    "        corrs, _ = spearmanr(predicted_unseen, true_unseen)\n",
    "        corr2, _ = pearsonr(predicted_seen, true_seen)\n",
    "        corrs2, _ = spearmanr(predicted_seen, true_seen)\n",
    "\n",
    "        print(\n",
    "            f\"Length {cell_line:>10}, {gene:>10}:  {len(missing_cell_line):<6d} Pearson unseen: {corr:.3f} seen: {corr2:.3f}\"\n",
    "            f\" Spearman unseen: {corrs:.3f} Spearman seen {corrs2:.3f}\")\n",
    "\n",
    "        # plt.scatter(predicted_unseen, true_unseen)\n",
    "        # plt.show()\n",
    "\n",
    "print(top)\n",
    "train, test = train_test_split(only_cancer, test_size=0.2, random_state=seed)\n",
    "train['group_key'] = train[CELL_LINE].astype(str)\n",
    "train = train.sort_values('group_key')\n",
    "group_train = train.groupby('group_key').size().tolist()\n",
    "metric = 'correct_log_inhibition'\n",
    "X = train[features]\n",
    "y = train[metric]\n",
    "# model.fit(X.to_numpy(), y.to_numpy(), group=group_train)\n",
    "model.fit(X.to_numpy(), y.to_numpy())\n",
    "\n",
    "predicted = model.predict(only_cancer[features].values)\n",
    "true = only_cancer[metric]\n",
    "\n",
    "corrp, _ = pearsonr(predicted, true)\n",
    "corrs, _ = spearmanr(predicted, true)\n",
    "\n",
    "print(f\"Total correlation: P:{corrp:.3f} S:{corrs:.3f}\")\n"
   ],
   "id": "ed0ed987ebd1effc",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from matplotlib import pyplot as plt\n",
    "from scipy.stats import pearsonr, spearmanr\n",
    "\n",
    "train_filtered = train.copy()\n",
    "test_filtered = test.copy()\n",
    "\n",
    "# train_filtered = train.nlargest(200, metric)\n",
    "# test_filtered = test.nlargest(200, metric)\n",
    "\n",
    "train_pred = model.predict(train_filtered[features].to_numpy())\n",
    "test_pred = model.predict(test_filtered[features].to_numpy())\n",
    "\n",
    "y_test = test_filtered[metric].to_numpy()\n",
    "y_train = train_filtered[metric].to_numpy()\n",
    "\n",
    "cell_lines = filtered[CELL_LINE].unique()\n",
    "# for cell_line in cell_lines:\n",
    "for cell_line in cell_lines:\n",
    "    cell_line_mask = filtered[CELL_LINE] == cell_line\n",
    "    print(f\"cell-line:{cell_line}, length:{cell_line_mask.sum()} \", cell_line)\n",
    "    print(metric)\n",
    "\n",
    "    for gene in filtered[cell_line_mask][CANONICAL_GENE].unique():\n",
    "        hepa_mask = cell_line_mask & (filtered[CANONICAL_GENE] == gene)\n",
    "        print(f\"gene-{gene} length: {hepa_mask.sum()}\")\n",
    "        # for f in ['sense_exon', 'sense_intron', 'sense_utr']:\n",
    "        for f in ['NA']:\n",
    "            # hep_pred = filtered[hepa_mask][f].to_numpy()\n",
    "            # print(f\"{f}: {filtered[hepa_mask][f].sum()} / {hepa_mask.sum()}\")\n",
    "            hep_pred = model.predict(filtered[hepa_mask][features].to_numpy())\n",
    "            hepa_truth = filtered[hepa_mask][metric]\n",
    "            p_hepa, pp_value = pearsonr(hepa_truth, hep_pred)\n",
    "            s_hepa, sp_value = spearmanr(hepa_truth, hep_pred)\n",
    "            print(f\"P:{p_hepa:.2}, p-value: {pp_value:.2}\")\n",
    "            print(f\"S:{s_hepa:.2} , p-value: {sp_value:.2}\")\n",
    "\n",
    "for feature in features:\n",
    "    # EXAMPLE = 'GGCATATGCAGATAATGTTC'\n",
    "\n",
    "    # malat_mask = filtered[CELL_LINE] == 'U251'\n",
    "\n",
    "    # malat_mask = filtered[CANONICAL_GENE] == \"HSD17B13\"\n",
    "    malat_mask = filtered[CANONICAL_GENE] == \"MALAT1\"\n",
    "    # example_mask = (filtered[CANONICAL_GENE] == \"MALAT1\") & (filtered[SEQUENCE] == EXAMPLE)\n",
    "    malat_pred = model.predict(filtered[malat_mask][features].to_numpy())\n",
    "    print(metric)\n",
    "    malat_truth = filtered[malat_mask][metric]\n",
    "    p_hepa, pp_value = pearsonr(malat_truth, filtered[malat_mask][feature])\n",
    "    s_hepa, sp_value = spearmanr(malat_truth, filtered[malat_mask][feature])\n",
    "\n",
    "    plt.scatter(filtered[malat_mask][feature], malat_truth)\n",
    "    # example_df = filtered.loc[example_mask, :].copy()\n",
    "    # plt.scatter(example_df[feature], malat_truth[example_mask], color='orange')\n",
    "    plt.scatter(filtered[malat_mask][feature], malat_truth, color='blue', alpha=0.05)\n",
    "    plt.xlabel(feature)\n",
    "    plt.ylabel(metric)\n",
    "    print(\"Feature: \", feature)\n",
    "    print(f\"P:{p_hepa:.2} p-value {pp_value:.2}\")\n",
    "    print(f\"S:{s_hepa:.2} p-value {sp_value:.2}\")\n",
    "    plt.show()\n",
    "\n",
    "p_hepa, _ = pearsonr(malat_truth, malat_pred)\n",
    "s_hepa, _ = spearmanr(malat_truth, malat_pred)\n",
    "print(\"P: \", p_hepa)\n",
    "print(\"S: \", s_hepa)\n",
    "\n",
    "p_train, _ = pearsonr(y_train, train_pred)\n",
    "s_train, _ = spearmanr(y_train, train_pred)\n",
    "p_test, p_value_ptest = pearsonr(y_test, test_pred)\n",
    "s_test, p_value_stest = spearmanr(y_test, test_pred)\n",
    "\n",
    "print(\"PTRAIN: \", p_train)\n",
    "print(\"PTEST: \", p_test)\n",
    "print(\"STRAIN: \", s_train)\n",
    "print(\"STEST: \", s_test)\n",
    "print(\"PvaluePtest: \", p_value_ptest)\n",
    "print(\"PValueStest: \", p_value_stest)\n",
    "\n",
    "# scatter plot\n",
    "# plt.scatter(test_pred, y_test, alpha=0.2)\n",
    "plt.scatter(malat_pred, malat_truth, alpha=0.2)\n",
    "\n",
    "# least‑squares regression line\n",
    "# slope, intercept = np.polyfit(test_pred, y_test, 1)\n",
    "# x = np.linspace(test_pred.min(), test_pred.max(), 100)\n",
    "# plt.plot(x, slope * x + intercept, color='red')\n",
    "\n",
    "# labels and stats\n",
    "plt.xlabel('Predicted ranking')\n",
    "plt.ylabel(metric)\n",
    "txt = (\n",
    "    f\"Train Pearson:  {p_train:.2f}\\n\"\n",
    "    f\"Train Spearman: {s_train:.2f}\\n\"\n",
    "    f\"Test Pearson:   {p_test:.2f}\\n\"\n",
    "    f\"Test Spearman:  {s_test:.2f}\"\n",
    ")\n",
    "plt.text(\n",
    "    0.05, 0.95, txt,\n",
    "    transform=plt.gca().transAxes,\n",
    "    va='top',\n",
    "    bbox=dict(boxstyle='round', facecolor='white', alpha=0.5)\n",
    ")\n",
    "\n",
    "plt.show()\n"
   ],
   "id": "c2e596a979ec84f9",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "def geo_mean_overflow(iterable):\n",
    "    return np.exp(np.log(iterable).mean())"
   ],
   "id": "798cd2e3736032ea",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from scripts.Models.model_helpers import log_inhibition_to_regular\n",
    "\n",
    "\n",
    "def get_threshold_for_percentile(percentile, df):\n",
    "    k = max(1, int(np.ceil((percentile / 100) * len(df))))  # desired count\n",
    "    threshold = np.partition(df, -k)[-k]  # value at 99th percentile\n",
    "    return threshold\n",
    "\n",
    "\n",
    "# def correction(df):\n",
    "#     return (df['log_volume'])\n",
    "\n",
    "# def correction(df):\n",
    "#     return (df[VOLUME]) / df[VOLUME]\n",
    "\n",
    "def correction(df):\n",
    "    return df[VOLUME] / (df[VOLUME] + 10)\n",
    "\n",
    "for percentile in [1, 5, 10, 50]:\n",
    "    threshold = get_threshold_for_percentile(percentile, log_inhibition_to_regular(filtered['log_inhibition']))\n",
    "    # print(correction(filtered))\n",
    "    # print(log_inhibition_to_regular(filtered['log_inhibition']))\n",
    "    print(f\"Threshold for top_{percentile}: {threshold}\")"
   ],
   "id": "b499ac04ac8095b9",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "evaluate_top(model, train, test)",
   "id": "41cb7eca0683bbf9",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "MALAT_scores = model.predict(filtered[filtered[CANONICAL_GENE] == 'MALAT1'][features])\n",
    "MALAT_scores.sort()\n",
    "print(MALAT_scores)\n",
    "get_threshold_for_percentile(1, MALAT_scores)"
   ],
   "id": "8315e8a3909ab118",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# ---------------------------------------------------------------------------\n",
    "# 5-fold “stable” permutation-importance for an XGBRanker (rank:pairwise)\n",
    "# – one query = all rows that share the same CELL_LINE ----------------------\n",
    "# ---------------------------------------------------------------------------\n",
    "\n",
    "from xgboost import XGBRanker\n",
    "from sklearn.model_selection import GroupKFold\n",
    "from sklearn.inspection import permutation_importance\n",
    "from sklearn.base import clone\n",
    "from scipy import stats\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "metric = \"correct_log_inhibition\"  # target column\n",
    "groups = filtered[CELL_LINE]  # one “qid” per row\n",
    "gkf = GroupKFold(n_splits=6)\n",
    "\n",
    "\n",
    "def spearman_scorer(est, X, y):  # (estimator, X, y) signature\n",
    "    return stats.spearmanr(y, est.predict(X)).correlation ** 2\n",
    "\n",
    "\n",
    "all_imps = []\n",
    "\n",
    "for tr_idx, te_idx in gkf.split(filtered, filtered[metric], groups):\n",
    "    # ---- build train / test sets -----------------------------------------\n",
    "    train_df = filtered.iloc[tr_idx].sort_values(CELL_LINE)  # keep queries contiguous\n",
    "    test_df = filtered.iloc[te_idx]\n",
    "\n",
    "    X_tr, y_tr = train_df[features], train_df[metric]\n",
    "    X_te, y_te = test_df[features], test_df[metric]\n",
    "\n",
    "    group_train = train_df.groupby(CELL_LINE).size().tolist()  # sizes per query\n",
    "\n",
    "    # ---- fit a fresh ranker on this fold ---------------------------------\n",
    "    ranker = XGBRanker(objective=\"rank:pairwise\", random_state=42)\n",
    "    ranker.fit(X_tr.to_numpy(), y_tr.to_numpy(), group=group_train)\n",
    "\n",
    "    # ---- permutation importance on the hold-out --------------------------\n",
    "    r = permutation_importance(\n",
    "        ranker, X_te.to_numpy(), y_te.to_numpy(),\n",
    "        n_repeats=30, scoring=spearman_scorer,\n",
    "        random_state=42, n_jobs=1\n",
    "    )\n",
    "    all_imps.append(r.importances_mean)\n",
    "\n",
    "# --------------------------------------------------------------------------\n",
    "# aggregate across folds\n",
    "imp_mat = np.vstack(all_imps)  # (folds × features)\n",
    "imp_df = pd.DataFrame(\n",
    "    imp_mat.T,\n",
    "    index=features,\n",
    "    columns=[f\"fold_{i + 1}\" for i in range(imp_mat.shape[0])]\n",
    ")\n",
    "\n",
    "summary = (imp_df\n",
    "           .agg([\"mean\", \"std\"], axis=1)\n",
    "           .rename(columns={\"mean\": \"mean_imp\", \"std\": \"std_imp\"})\n",
    "           .sort_values(\"mean_imp\", ascending=False))\n"
   ],
   "id": "c305878e3529fa5a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "print(summary.head(15))\n",
   "id": "7b5f488d9830dd16",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "\n",
    "GFP_IN_YEAST = (\n",
    "    'ATGGTtAGtAAaGGaGAaGAGTTgTTCACaGGaGTGGTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAGGGCGAGGGCGATGCCACCTACGGCAAGCTGACCCTGAAGTTCATCTGCACCACCGGCAAGCTGCCCGTGCCCTGGCCCACCCTCGTGACCACCCTGACCTACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGCACGACTTCTTCAAGTCCGCCATGCCCGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGCAACTACAAGACCCGCGCCGAGGTGAAGTTCGAGGGCGACACCCTGGTGAACCGCATCGAGCTGAAGGGCATCGACTTCAAGGAGGACGGCAACATCCTGGGGCACAAGCTGGAGTACAACTACAACAGCCACAACGTCTATATCATGGCCGACAAGCAGAAGAACGGCATCAAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCGACGGCCCCGTGCTGCTGCCCGACAACCACTACCTGAGCACCCAGTCCGCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACAAGGGTGCTGGGGCAggtacCCCTAAAGATCCAGCCAAACCTCCGGCCAcGGCACAAGTTGTGGGATGGCCACCGGTGAGATCATACCGGAAGAACGTGATGGTTTCCTGCCAAAAATCAAGCGGTGGCCCGGAGGCGGCGGCGTTCGTGAAGTAA'\n",
    "    .upper())\n",
    "GFP_IN_HUMAN = (\n",
    "    \"atggtgagcaagggcgaggagctgttcaccggggtggtgcccatcctggtcgagctggacggcgacgtaaacggccacaagttcagcgtgtccggcgagggcgagggcgatgccacctacggcaagctgaccctgaagttcatctgcaccaccggcaagctgcccgtgccctggcccaccctcgtgaccaccctgacctacggcgtgcagtgcttcagccgctaccccgaccacatgaagcagcacgacttcttcaagtccgccatgcccgaaggctacgtccaggagcgcaccatcttcttcaaggacgacggcaactacaagacccgcgccgaggtgaagttcgagggcgacaccctggtgaaccgcatcgagctgaagggcatcgacttcaaggaggacggcaacatcctggggcacaagctggagtacaactacaacagccacaacgtctatatcatggccgacaagcagaagaacggcatcaaggtgaacttcaagatccgccacaacatcgaggacggcagcgtgcagctcgccgaccactaccagcagaacacccccatcggcgacggccccgtgctgctgcccgacaaccactacctgagcacccagtccgccctgagcaaagaccccaacgagaagcgcgatcacatggtcctgctggagttcgtgaccgccgccgggatcactctcggcatggacgagctgtacaagaagcttagccatggcttcccgccggaggtggaggagcaggatgatggcacgctgcccatgtcttgtgcccaggagagcgggatggaccgtcaccctgcagcctgtgcttctgctaggatcaatgtgaagcgacctgccgccacaaagaaggctggacaggctaagaagaagaaatgaggatcccgcgcgcgcatatgttaattaaccaactgcatggggatccacgcgttaagtcgacaatcaacctctggattacaaaatttgtgaaagattgactggtattcttaactatgttgctccttttacgctatgtggatacgctgctttaatgcctttgtatcatgctattgcttcccgtatggctttcattttctcctccttgtataaatcctggttgctgtctctttatgaggagttgtggcccgttgtcaggcaacgtggcgtggtgtgcactgtgtttgctgacgcaacccccactggttggggcattgccaccacctgtcagctcctttccgggactttcgctttccccctccctattgccacggcggaactcatcgccgcctgccttgcccgctgctggacaggggctcggctgttgggcactgacaattccgtggtgttgtcggggaaatcatcgtcctttccttggctgctcgcctgtgttgccacctggattctgcgcgggacgtccttctgctacgtcccttcggccctcaatccagcggaccttccttcccgcggcctgctgccggctctgcggcctcttccgcgtcttcgccttcgccctcagacgagtcggatctccctttgggccgcctccccgcgtcgactttaagaccaatgacttacaaggcagctgtagatcttagccactttttaaaagaaaaggggggactggaagggctaattcactcccaacgaagacaagatctgctttttgcttgtactgggtctctctggttagaccagatctgagcctgggagctctctggctaactagggaacccactgcttaagcctcaataaagcttgccttgagtgcttcaagtagtgtgtgcccgtctgttgtgtgactctggtaactagagatccctcagacccttttagtcagtgtggaaaatctctagcagtacgtatagtagttcatgtcatcttattattcagtatttataacttgcaaagaaatgaatatcagagagtgagagg\".upper()\n",
    ")\n",
    "# GFP + Degron x <= 842\n",
    "# NLS 843 <= x<= 893\n",
    "# 3UTR 1615 <= x <= 1848\n"
   ],
   "id": "facb96e0cf136933",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from scripts.data_genertion.data_handling import get_unique_human_genes\n",
    "\n",
    "from asodesigner.read_human_genome import get_locus_to_data_dict\n",
    "import pickle\n",
    "from asodesigner.consts import CACHE_DIR\n",
    "\n",
    "cache_path = CACHE_DIR / 'gene_to_data_simple_cache.pickle'\n",
    "if not cache_path.exists():\n",
    "    gene_to_data = get_locus_to_data_dict(include_introns=True, gene_subset=['MALAT1'])\n",
    "    with open(cache_path, 'wb') as f:\n",
    "        pickle.dump(gene_to_data, f)\n",
    "else:\n",
    "    with open(cache_path, 'rb') as f:\n",
    "        gene_to_data = pickle.load(f)\n",
    "\n",
    "MALAT = gene_to_data['MALAT1'].full_mrna"
   ],
   "id": "53091dfbffdc8fff",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from asodesigner.util import get_antisense\n",
    "import pandas as pd\n",
    "\n",
    "GFP_YEAST_END = len(GFP_IN_YEAST) - 3\n",
    "GFP_HUMAN_END = 891  # TGA stop codon in NLS\n",
    "malat_info = gene_to_data['MALAT1']\n",
    "MALAT_END = malat_info.exon_indices[-1][1] - malat_info.cds_start\n",
    "\n",
    "\n",
    "def get_init_df(target_mrna, end):\n",
    "    candidates = []\n",
    "    sense_starts = []\n",
    "    sense_lengths = []\n",
    "    sense_starts_from_end = []\n",
    "\n",
    "    for i in range(0, len(target_mrna) - 19):\n",
    "        target = target_mrna[i: i + 20]\n",
    "        candidates.append(get_antisense(str(target)))\n",
    "        sense_starts.append(i)\n",
    "        sense_lengths.append(20)\n",
    "        sense_starts_from_end.append(end - i)\n",
    "    df = pd.DataFrame(\n",
    "        {SEQUENCE: candidates, SENSE_START: sense_starts,\n",
    "         SENSE_LENGTH: sense_lengths, \"sense_start_from_end\": sense_starts_from_end})\n",
    "    return df\n",
    "\n",
    "\n",
    "SEQUENCES = [GFP_IN_YEAST, GFP_IN_HUMAN, MALAT]\n",
    "df_yeast = get_init_df(GFP_IN_YEAST, GFP_YEAST_END)\n",
    "df_human = get_init_df(GFP_IN_HUMAN, GFP_HUMAN_END)\n",
    "df_malat = get_init_df(MALAT, MALAT_END)"
   ],
   "id": "afd18cd78bfb0c6c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "df_yeast[CANONICAL_GENE] = 'YEAST_GFP'\n",
    "df_human[CANONICAL_GENE] = 'HUMAN_GFP'\n",
    "df_malat[CANONICAL_GENE] = 'MALAT1'\n",
    "dataframes = [df_yeast, df_human, df_malat]"
   ],
   "id": "7c3efb81430b0480",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "for df in dataframes:\n",
    "    df[CELL_LINE_ORGANISM] = 'human'\n",
    "    df[INHIBITION] = 0  # Just for the function, not important"
   ],
   "id": "2c430126bd9312ff",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "df_malat = get_populated_df_with_structure_features(df_malat, ['MALAT1'], gene_to_data)\n",
    "dataframes[2] = df_malat\n",
    "df_malat"
   ],
   "id": "c917e582b2bef67a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "df_yeast['sense_exon'] = 1\n",
    "df_human['sense_exon'] = 1\n",
    "df_yeast['sense_intron'] = 0\n",
    "df_human['sense_intron'] = 0\n",
    "df_yeast['sense_utr'] = 0\n",
    "df_human['sense_utr'] = [1 if sense_start > 842 else 0 for sense_start in df_human[SENSE_START]]"
   ],
   "id": "f52c4b88b6b3d785",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from scripts.data_genertion.data_handling import populate_features\n",
    "\n",
    "for i, df in enumerate(dataframes):\n",
    "    df[TREATMENT_PERIOD] = 24  # keep constant for all\n",
    "    df[VOLUME] = 1000  # keep constant for all\n",
    "    df['log_volume'] = np.log(df[VOLUME])\n",
    "    df['normalized_start'] = df[SENSE_START] / len(SEQUENCES[i])\n",
    "    df['normalized_sense_start_from_end'] = df['sense_start_from_end'] / len(SEQUENCES[i])\n",
    "    easy_to_populate = ['at_skew', 'entropy', 'gc_content', 'gc_content_3_prime_5', 'gc_skew', 'hairpin_score',\n",
    "                        'homooligo_count', 'internal_fold', 'nucleotide_diversity', 'self_energy', 'stop_codon_count',\n",
    "                        'at_rich_region_score', 'poly_pyrimidine_stretch']\n",
    "    populate_features(df, easy_to_populate)\n"
   ],
   "id": "61e6b59b4e7c6604",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "from asodesigner.process_utils import LocusInfo\n",
    "from scripts.data_genertion.data_handling import get_populate_fold\n",
    "\n",
    "genes_u = ['YEAST_GFP', 'HUMAN_GFP', 'MALAT1']\n",
    "gene_to_data = {'YEAST_GFP': LocusInfo(), 'HUMAN_GFP': LocusInfo(), 'MALAT1': LocusInfo()}\n",
    "gene_to_data['YEAST_GFP'].full_mrna = GFP_IN_YEAST\n",
    "gene_to_data['HUMAN_GFP'].full_mrna = GFP_IN_HUMAN\n",
    "gene_to_data['MALAT1'].full_mrna = MALAT\n"
   ],
   "id": "3a612d44e2d16a6a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "for i in range(len(dataframes)):\n",
    "    fold_variants = [(40, 15)]\n",
    "    dataframes[i] = get_populate_fold(dataframes[i], genes_u, gene_to_data, fold_variants=fold_variants)\n"
   ],
   "id": "775b8a3da67b2272",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from hybridization.hybridization_features import get_exp_psrna_hybridization\n",
    "\n",
    "for df in dataframes:\n",
    "    df.loc[:, 'exp_ps_hybr'] = [\n",
    "        get_exp_psrna_hybridization(antisense.replace('T', 'U'), temp=37) for\n",
    "        antisense in df[SEQUENCE]]\n",
    "    print(df.columns)"
   ],
   "id": "817aa6c27cafd4bb",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "scores_per_gfp = []\n",
    "for df in dataframes:\n",
    "    scores_per_gfp.append(model.predict(df[features]))"
   ],
   "id": "753930856d0a2c4e",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "index = 0\n",
    "scores = scores_per_gfp[index]\n",
    "k = 400  # how many top rows\n",
    "idx = np.argsort(-scores)[:k]\n",
    "# print(df.iloc[idx])\n",
    "# (optional) with scores:\n",
    "df_with_scores = dataframes[index].assign(score=scores)\n",
    "print(df_with_scores.iloc[idx])\n",
    "df_with_scores.iloc[idx].to_csv('yeast_gfp_asos.csv', index=False)"
   ],
   "id": "7541912e82292d5d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "index = 1\n",
    "scores = scores_per_gfp[index]\n",
    "k = 400  # how many top rows\n",
    "idx = np.argsort(-scores)[:k]\n",
    "# print(df.iloc[idx])\n",
    "# (optional) with scores:\n",
    "df_with_scores = dataframes[index].assign(score=scores)\n",
    "print(df_with_scores.iloc[idx])\n",
    "df_with_scores.iloc[idx].to_csv('human_gfp_asos.csv', index=False)\n"
   ],
   "id": "f46112bba2b35746",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "index = 2\n",
    "scores = scores_per_gfp[index]\n",
    "k = 400  # how many top rows\n",
    "idx = np.argsort(-scores)[:k]\n",
    "# print(df.iloc[idx])\n",
    "# (optional) with scores:\n",
    "df_with_scores = dataframes[index].assign(score=scores)\n",
    "print(df_with_scores.iloc[idx])\n",
    "df_with_scores.iloc[idx].to_csv('malat_asos.csv', index=False)"
   ],
   "id": "bcfb37cd0fdc0b3f",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "index = 0\n",
    "df_with_scores = dataframes[index].assign(score=scores_per_gfp[index])\n",
    "\n",
    "previous_experiment = [\n",
    "    'CCTAGGGACAACGGTCATCG',\n",
    "    'AAACACGAAATTGGCAGGGG',\n",
    "    'CAAGACATGGGCAGCGTGCC',\n",
    "    'GGACACGCTGAACTTGTGGC',\n",
    "    'TGAAGAAGATGGTGCGCTCC',\n",
    "    'GCGTGCCATCATCCTGCTCC',\n",
    "    'CGGGCACACACTACTTGAAG',\n",
    "    'TGATCGCGCTTCTCGTTGGG'\n",
    "]\n",
    "df_with_scores['relative_score'] = 1 - df_with_scores['score'].rank(method='max', pct=True)\n",
    "previous_asos = df_with_scores[df_with_scores[SEQUENCE].isin(previous_experiment)]\n",
    "\n",
    "previous_asos[[SEQUENCE, 'relative_score', 'sense_start']]"
   ],
   "id": "2e5e0862a4e80c76",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "chinese_aso = df_with_scores[df_with_scores[SEQUENCE] == 'TTGCCGGTGGTGCAGATGAA']\n",
    "chinese_score = chinese_aso['score'].iloc[0]\n",
    "print(chinese_aso[[SEQUENCE, 'relative_score', 'sense_start']])"
   ],
   "id": "7de213728c6e97e8",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "plt.hist(df_with_scores['score'], bins=40, alpha=0.5, label='scores', log=True)\n",
    "plt.hist(chinese_aso['score'], bins=40, alpha=0.5, color='red', label='scores', log=True)\n",
    "\n",
    "plt.show()"
   ],
   "id": "79e05bfcc76b087c",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "index = 2\n",
    "df_with_scores = dataframes[index].assign(score=scores_per_gfp[index])\n",
    "df_with_scores['relative_score'] = 1 - df_with_scores['score'].rank(method='max', pct=True)\n",
    "\n",
    "EXAMPLE = 'GGCATATGCAGATAATGTTC'\n",
    "\n",
    "plt.hist(df_with_scores['score'], bins=20, alpha=0.5, label='scores', log=True)\n",
    "plt.hist(df_with_scores[df_with_scores[SEQUENCE] == EXAMPLE]['score'], bins=20, alpha=0.5, color='red', label='scores',\n",
    "         log=True)\n",
    "\n",
    "plt.show()\n",
    "score = df_with_scores[df_with_scores[SEQUENCE] == EXAMPLE]['score'].iloc[0]\n",
    "len(df_with_scores[df_with_scores['score'] > score]) / len(df_with_scores)\n",
    "print(score)\n",
    "print(df_with_scores[df_with_scores[SEQUENCE] == EXAMPLE]['relative_score'])"
   ],
   "id": "3d7cd1f6adc18c74",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "k = 10  # how many bot rows\n",
    "idx = np.argsort(scores)[:k]\n",
    "print(df_with_scores.iloc[idx])\n",
    "# print(df_with_scores[df_with_scores[SEQUENCE] == 'TTGCCGGTGGTGCAGATGAA'])"
   ],
   "id": "e5ac5a75f47d8cb1",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "index = 2\n",
    "df_with_scores = dataframes[index].assign(score=scores_per_gfp[index])\n",
    "df_with_scores['relative_score'] = 1 - df_with_scores['score'].rank(method='max', pct=True)\n",
    "\n",
    "from scipy.stats import pearsonr, spearmanr\n",
    "from sklearn.feature_selection import mutual_info_regression\n",
    "\n",
    "metric = 'correct_log_inhibition'\n",
    "\n",
    "\n",
    "def print_correlations(df, name1, name2, p_value_threshold=None):\n",
    "    if p_value_threshold is None:\n",
    "        p_value_threshold = 1\n",
    "    corr, p_value = pearsonr(df[name1], df[name2])\n",
    "    if p_value < p_value_threshold:\n",
    "        print(f\"Feature: {name1:<35}, Pearson: {corr:<5.2f}, p-value: {p_value:<10.2} Target: {name2:<35}\")\n",
    "    corr, p_value = spearmanr(df[name1], df[name2])\n",
    "    if p_value < p_value_threshold:\n",
    "        print(f\"Feature: {name1:<35}, Spearman: {corr:<5.2f}, p-value: {p_value:<10.2} Target: {name2:<35}\")\n",
    "\n",
    "        # MIC (via Mutual Information)\n",
    "    try:\n",
    "        mic = mutual_info_regression(df[[name1]], df[name2], discrete_features='auto')[0]\n",
    "        print(f\"Feature: {name1:<35}, MIC:      {mic:<5.2f}                             Target: {name2:<35}\")\n",
    "    except Exception as e:\n",
    "        print(f\"Could not compute MIC for {name1} vs {name2}: {e}\")\n",
    "\n",
    "\n",
    "# for gene in genes_u:\n",
    "# print(\"gene, \", gene)\n",
    "feature = 'gc_content'\n",
    "current_df = df_with_scores\n",
    "current_df = df_with_scores[df_with_scores[CANONICAL_GENE] == 'MALAT1']\n",
    "plt.scatter(current_df[feature], current_df['score'], alpha=0.2)\n",
    "plt.scatter(current_df[current_df[SEQUENCE] == EXAMPLE][feature], current_df[current_df[SEQUENCE] == EXAMPLE]['score'])\n",
    "plt.xlabel('Place in gene')\n",
    "plt.ylabel(f'score')\n",
    "plt.show()\n",
    "\n",
    "print_correlations(current_df, feature, 'score')\n",
    "# print_correlations(current_df, 'on_target_fold_openness_normalized40_15', metric)\n",
    "# print_correlations(current_df, 'on_target_fold_openness_normalized40_15', 'score')"
   ],
   "id": "900103ff55872ffb",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "for feature in features:\n",
    "    plt.scatter(current_df[feature], current_df['score'])\n",
    "    # plt.scatter(current_df[feature], current_df['log_inhibition'])\n",
    "    special_aso = current_df[current_df[SEQUENCE] == 'GGCATATGCAGATAATGTTC']\n",
    "    plt.scatter(special_aso[feature], special_aso['score'])\n",
    "    plt.xlabel(f'Feature: {feature}')\n",
    "    plt.ylabel(f'Score: {score}')\n",
    "    plt.show()"
   ],
   "id": "e826a8f88e7a083a",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from fuzzysearch import find_near_matches\n",
    "\n",
    "m = find_near_matches('TCTGCATCTAGGCCATCATA', GFP_IN_YEAST, max_insertions=0, max_deletions=0, max_l_dist=2)\n",
    "print(m)\n",
    "m = find_near_matches('GATGGACATTGCCTCTTCAT', GFP_IN_YEAST, max_insertions=0, max_deletions=0, max_l_dist=2)\n",
    "print(m)\n",
    "m = find_near_matches('CTCCAGTCGTTTCACAATGC', GFP_IN_YEAST, max_insertions=0, max_deletions=0, max_l_dist=2)\n",
    "print(m)"
   ],
   "id": "10519cc56f0d95af",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Generate scrambled\n",
    "import random\n",
    "\n",
    "GFP_BEST = dataframes[0][:30][[SEQUENCE]]\n",
    "scrambled_df = GFP_BEST.copy()\n",
    "\n",
    "\n",
    "def shuffle_seq(seq, seed=45):\n",
    "    rng = random.Random(seed)\n",
    "    lst = list(seq)\n",
    "    rng.shuffle(lst)\n",
    "    return ''.join(lst)\n",
    "\n",
    "\n",
    "scrambled_df[SEQUENCE] = [shuffle_seq('CTGCTTGTCGGCCATGATAT') for seq in scrambled_df[SEQUENCE]]\n",
    "\n",
    "easy_to_populate = ['at_skew', 'entropy', 'gc_content', 'gc_content_3_prime_5', 'gc_skew', 'hairpin_score',\n",
    "                    'homooligo_count', 'internal_fold', 'nucleotide_diversity', 'self_energy', 'stop_codon_count',\n",
    "                    'at_rich_region_score', 'poly_pyrimidine_stretch']\n",
    "populate_features(scrambled_df, easy_to_populate)\n"
   ],
   "id": "553efe12c0ac9d9e",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# Check for on-target GFP\n",
    "for seq in scrambled_df[SEQUENCE]:\n",
    "    m = find_near_matches(seq, GFP_IN_YEAST, max_insertions=0, max_deletions=0, max_l_dist=3)\n",
    "    if m:\n",
    "        print(seq)"
   ],
   "id": "9e8bdea38f2874b3",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "scrambled_df[0:1]",
   "id": "b36c606c08797f",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from hybridization.hybridization_features import get_exp_psrna_hybridization\n",
    "\n",
    "a = get_exp_psrna_hybridization('ACGATGGTCCTTCTTGTGAC'.replace('T', 'U'), temp=37)\n",
    "scrambled_df['exp_ps_hybr'] = a"
   ],
   "id": "9e86ced8688359df",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "scrambled_df[0:1]",
   "id": "4c80883a23fc97c4",
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
