{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-09-17T21:06:47.624604Z",
     "start_time": "2025-09-17T21:06:38.997594Z"
    }
   },
   "source": [
    "from scripts.data_genertion.consts import *\n",
    "from scripts.features.feature_extraction import load_all_features\n",
    "\n",
    "main_df = load_all_features()\n",
    "main_df[SEQUENCE] = main_df[SEQUENCE].astype(str)"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:06:47.815133Z",
     "start_time": "2025-09-17T21:06:47.627811Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from asodesigner.read_human_genome import get_locus_to_data_dict\n",
    "import pickle\n",
    "from asodesigner.consts import CACHE_DIR\n",
    "\n",
    "genes_u = ['HIF1A', 'APOL1', 'YAP1', 'SOD1', 'SNCA', 'IRF4', 'KRAS', 'KLKB1', 'SNHG14', 'DGAT2', 'IRF5', 'HTRA1',\n",
    "           'MYH7', 'MALAT1', 'HSD17B13']\n",
    "cache_path = CACHE_DIR / 'gene_to_data_simple_cache.pickle'\n",
    "if not cache_path.exists():\n",
    "    gene_to_data = get_locus_to_data_dict(include_introns=True, gene_subset=genes_u)\n",
    "    with open(cache_path, 'wb') as f:\n",
    "        pickle.dump(gene_to_data, f)\n",
    "else:\n",
    "    with open(cache_path, 'rb') as f:\n",
    "        gene_to_data = pickle.load(f)"
   ],
   "id": "50d64118f2c22bfd",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:03.508033Z",
     "start_time": "2025-09-17T21:06:47.854533Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from scripts.data_genertion.data_handling import get_populated_df_with_structure_features\n",
    "\n",
    "main_df = get_populated_df_with_structure_features(main_df, genes_u, gene_to_data)"
   ],
   "id": "48b0dbba50a89a98",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:03.523875Z",
     "start_time": "2025-09-17T21:07:03.517243Z"
    }
   },
   "cell_type": "code",
   "source": "main_df[[SENSE_TYPE, 'sense_intron', 'sense_exon', 'sense_utr', SENSE_START, CANONICAL_GENE]]",
   "id": "803511a1c7fe474b",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      sense_type  sense_intron  sense_exon  sense_utr  sense_start  \\\n",
       "0         intron             1           0          0        41212   \n",
       "1         intron             1           0          0        23686   \n",
       "2         intron             1           0          0        43363   \n",
       "3         intron             1           0          0        23680   \n",
       "4         intron             1           0          0        41168   \n",
       "...          ...           ...         ...        ...          ...   \n",
       "34760     intron             1           0          0         7827   \n",
       "34761     intron             1           0          0         8250   \n",
       "34762     intron             1           0          0         8335   \n",
       "34763        utr             0           1          1        13848   \n",
       "34764     intron             1           0          0         5361   \n",
       "\n",
       "      Canonical Gene Name  \n",
       "0                    KRAS  \n",
       "1                    KRAS  \n",
       "2                    KRAS  \n",
       "3                    KRAS  \n",
       "4                    KRAS  \n",
       "...                   ...  \n",
       "34760               APOL1  \n",
       "34761               APOL1  \n",
       "34762               APOL1  \n",
       "34763               APOL1  \n",
       "34764               APOL1  \n",
       "\n",
       "[29987 rows x 6 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sense_type</th>\n",
       "      <th>sense_intron</th>\n",
       "      <th>sense_exon</th>\n",
       "      <th>sense_utr</th>\n",
       "      <th>sense_start</th>\n",
       "      <th>Canonical Gene Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41212</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23686</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>43363</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23680</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41168</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34760</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7827</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34761</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8250</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34762</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8335</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34763</th>\n",
       "      <td>utr</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13848</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34764</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5361</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29987 rows × 6 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:03.583530Z",
     "start_time": "2025-09-17T21:07:03.575502Z"
    }
   },
   "cell_type": "code",
   "source": [
    "main_df.loc[:, 'mrna_length'] = [len(gene_to_data[gene].full_mrna) for gene in main_df[CANONICAL_GENE]]\n",
    "main_df['normalized_sense_start_from_end'] = main_df['sense_start_from_end'] / main_df['mrna_length']"
   ],
   "id": "603adfe4a05886bb",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:03.749434Z",
     "start_time": "2025-09-17T21:07:03.641876Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "\n",
    "first_filtered = main_df.copy()\n",
    "first_filtered = first_filtered[first_filtered['Cell line organism'] == 'human']\n",
    "first_filtered = first_filtered.dropna(subset=[INHIBITION]).copy()\n",
    "log_correction = 1.01\n",
    "first_filtered.loc[:, 'log_inhibition'] = -np.log(log_correction - first_filtered[INHIBITION] / 100)\n",
    "first_filtered = first_filtered[~first_filtered[CELL_LINE].isin([\n",
    "    'Hela',  # scanning modifications\n",
    "    'Human Neuronal Cell',  # scanning modifications;\n",
    "    'CC-2580',  # scanning modifications\n",
    "    'SH-SY5Y'  # non pure PS based\n",
    "])]"
   ],
   "id": "425fb7c7e287abf8",
   "outputs": [],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:03.784472Z",
     "start_time": "2025-09-17T21:07:03.767368Z"
    }
   },
   "cell_type": "code",
   "source": [
    "\n",
    "ALL_CELL_LINES = first_filtered[CELL_LINE].unique()\n",
    "# non_hepa_cancer = filtered_original[\n",
    "#     # (~((filtered_original[CELL_LINE] == 'HepG2') & (filtered_original[TRANSFECTION].str.contains('Lipo')))) &\n",
    "#     (filtered_original[CELL_LINE] != 'HepG2')  # experiment settings too different\n",
    "#     & (filtered_original[CELL_LINE] != 'HepaRG')  # not similar to cancer\n",
    "#     & (filtered_original[CELL_LINE] != 'A-431')\n",
    "#     & (filtered_original[CELL_LINE] != 'SNU-449')\n",
    "#     # & (filtered_original[CELL_LINE] != 'KARPAS-229')\n",
    "#     # & (filtered_original[CELL_LINE] != 'MM.1R')\n",
    "# ]\n",
    "\n",
    "# ['A431', 'KARPAS-229', 'MM.1R', 'SK-MEL-28',\n",
    "non_hepa_cancer = first_filtered[\n",
    "    # (~((filtered_original[CELL_LINE] == 'HepG2') & (filtered_original[TRANSFECTION].str.contains('Lipo')))) &\n",
    "    (first_filtered[CELL_LINE] == 'A431')  # experiment settings too different\n",
    "    | (first_filtered[CELL_LINE] == 'KARPAS-229')  # not similar to cancer\n",
    "    | (first_filtered[CELL_LINE] == 'MM.1R')\n",
    "    | (first_filtered[CELL_LINE] == 'SK-MEL-28')\n",
    "    # & (filtered_original[CELL_LINE] != 'KARPAS-229')\n",
    "    # & (filtered_original[CELL_LINE] != 'MM.1R')\n",
    "\n",
    "    ].copy()"
   ],
   "id": "e95236aef72d3b78",
   "outputs": [],
   "execution_count": 7
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:53.756138Z",
     "start_time": "2025-09-17T21:07:53.753116Z"
    }
   },
   "cell_type": "code",
   "source": [
    "selected_features = [\n",
    "    TREATMENT_PERIOD,\n",
    "    'at_skew',\n",
    "    'CAI_score_global_CDS',\n",
    "    'stop_codon_count',\n",
    "    'sense_avg_accessibility',\n",
    "    'on_target_fold_openness_normalized40_15',\n",
    "    'sense_utr',\n",
    "    'nucleotide_diversity',\n",
    "    'internal_fold',\n",
    "    'normalized_start',\n",
    "    'RNaseH1_Krel_score_R7_krel', # renamed to best\n",
    "    'hairpin_score',\n",
    "    'Modification_min_distance_to_3prime',\n",
    "    'at_rich_region_score'\n",
    "]"
   ],
   "id": "e4a24b0fb313c16",
   "outputs": [],
   "execution_count": 15
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:54.051156Z",
     "start_time": "2025-09-17T21:07:54.046974Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import matplotlib.pyplot as plt\n",
    "from scipy.stats import spearmanr, pearsonr\n",
    "\n",
    "metric = 'correct_log_inhibition2'\n",
    "\n",
    "li = non_hepa_cancer['log_inhibition'].copy()\n",
    "non_hepa_cancer['correct_log_inhibition'] = li / (non_hepa_cancer[VOLUME] / (non_hepa_cancer[VOLUME] + 10))\n",
    "non_hepa_cancer['correct_log_inhibition2'] = li / (non_hepa_cancer[VOLUME] * 0.0000601 + 0.537)\n",
    "non_hepa_cancer['correct_log_inhibition3'] = non_hepa_cancer['correct_log_inhibition2'] / (-0.032*  non_hepa_cancer[TREATMENT_PERIOD] + 2.36)\n"
   ],
   "id": "75995680ab8af767",
   "outputs": [],
   "execution_count": 16
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:54.506368Z",
     "start_time": "2025-09-17T21:07:54.503423Z"
    }
   },
   "cell_type": "code",
   "source": [
    "min_val = non_hepa_cancer[metric].min()\n",
    "non_hepa_cancer[metric + 'ndcg'] = (\n",
    "    non_hepa_cancer[metric] - min_val\n",
    ") # for evaluation purposes, not training!"
   ],
   "id": "4020f7e13c9ff1c2",
   "outputs": [],
   "execution_count": 17
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:54.863707Z",
     "start_time": "2025-09-17T21:07:54.861343Z"
    }
   },
   "cell_type": "code",
   "source": [
    "cell_lines = ALL_CELL_LINES\n",
    "genes = non_hepa_cancer[CANONICAL_GENE].unique()"
   ],
   "id": "7c5d12a9a59c4a2",
   "outputs": [],
   "execution_count": 18
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:55.616925Z",
     "start_time": "2025-09-17T21:07:55.614946Z"
    }
   },
   "cell_type": "code",
   "source": "seed = 42\n",
   "id": "6e42db4dad1df7d2",
   "outputs": [],
   "execution_count": 19
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:56.107515Z",
     "start_time": "2025-09-17T21:07:56.105251Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from xgboost import XGBRanker\n",
    "\n",
    "model = XGBRanker(objective='rank:ndcg', ndcg_exp_gain=False, lambdarank_pair_method=\"topk\",\n",
    "                  lambdarank_num_pair_per_sample=200,\n",
    "                  seed=seed, n_jobs=-1\n",
    "                  )"
   ],
   "id": "1b7856fbebb56642",
   "outputs": [],
   "execution_count": 20
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:59.827729Z",
     "start_time": "2025-09-17T21:07:56.524352Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Real model, better to train on everything here\n",
    "df_copy = non_hepa_cancer.copy()\n",
    "df_copy['group_key'] = df_copy[CELL_LINE].astype(str)\n",
    "df_copy = df_copy.sort_values('group_key')\n",
    "group_train = df_copy.groupby('group_key').size().tolist()\n",
    "X = df_copy[selected_features]  # CHANGED: using selected_features\n",
    "y = df_copy[metric]\n",
    "model.fit(X.to_numpy(), y.to_numpy(), group=group_train)"
   ],
   "id": "b8f6499a215d0253",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "XGBRanker(base_score=None, booster=None, callbacks=None, colsample_bylevel=None,\n",
       "          colsample_bynode=None, colsample_bytree=None, device=None,\n",
       "          early_stopping_rounds=None, enable_categorical=False,\n",
       "          eval_metric=None, feature_types=None, feature_weights=None,\n",
       "          gamma=None, grow_policy=None, importance_type=None,\n",
       "          interaction_constraints=None, lambdarank_num_pair_per_sample=200,\n",
       "          lambdarank_pair_method='topk', learning_rate=None, max_bin=None,\n",
       "          max_cat_threshold=None, max_cat_to_onehot=None, max_delta_step=None,\n",
       "          max_depth=None, max_leaves=None, min_child_weight=None, missing=nan,\n",
       "          monotone_constraints=None, multi_strategy=None, n_estimators=None, ...)"
      ],
      "text/html": [
       "<style>#sk-container-id-1 {\n",
       "  /* Definition of color scheme common for light and dark mode */\n",
       "  --sklearn-color-text: #000;\n",
       "  --sklearn-color-text-muted: #666;\n",
       "  --sklearn-color-line: gray;\n",
       "  /* Definition of color scheme for unfitted estimators */\n",
       "  --sklearn-color-unfitted-level-0: #fff5e6;\n",
       "  --sklearn-color-unfitted-level-1: #f6e4d2;\n",
       "  --sklearn-color-unfitted-level-2: #ffe0b3;\n",
       "  --sklearn-color-unfitted-level-3: chocolate;\n",
       "  /* Definition of color scheme for fitted estimators */\n",
       "  --sklearn-color-fitted-level-0: #f0f8ff;\n",
       "  --sklearn-color-fitted-level-1: #d4ebff;\n",
       "  --sklearn-color-fitted-level-2: #b3dbfd;\n",
       "  --sklearn-color-fitted-level-3: cornflowerblue;\n",
       "\n",
       "  /* Specific color for light theme */\n",
       "  --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, white)));\n",
       "  --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, black)));\n",
       "  --sklearn-color-icon: #696969;\n",
       "\n",
       "  @media (prefers-color-scheme: dark) {\n",
       "    /* Redefinition of color scheme for dark theme */\n",
       "    --sklearn-color-text-on-default-background: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-background: var(--sg-background-color, var(--theme-background, var(--jp-layout-color0, #111)));\n",
       "    --sklearn-color-border-box: var(--sg-text-color, var(--theme-code-foreground, var(--jp-content-font-color1, white)));\n",
       "    --sklearn-color-icon: #878787;\n",
       "  }\n",
       "}\n",
       "\n",
       "#sk-container-id-1 {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 pre {\n",
       "  padding: 0;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-hidden--visually {\n",
       "  border: 0;\n",
       "  clip: rect(1px 1px 1px 1px);\n",
       "  clip: rect(1px, 1px, 1px, 1px);\n",
       "  height: 1px;\n",
       "  margin: -1px;\n",
       "  overflow: hidden;\n",
       "  padding: 0;\n",
       "  position: absolute;\n",
       "  width: 1px;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-dashed-wrapped {\n",
       "  border: 1px dashed var(--sklearn-color-line);\n",
       "  margin: 0 0.4em 0.5em 0.4em;\n",
       "  box-sizing: border-box;\n",
       "  padding-bottom: 0.4em;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-container {\n",
       "  /* jupyter's `normalize.less` sets `[hidden] { display: none; }`\n",
       "     but bootstrap.min.css set `[hidden] { display: none !important; }`\n",
       "     so we also need the `!important` here to be able to override the\n",
       "     default hidden behavior on the sphinx rendered scikit-learn.org.\n",
       "     See: https://github.com/scikit-learn/scikit-learn/issues/21755 */\n",
       "  display: inline-block !important;\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-text-repr-fallback {\n",
       "  display: none;\n",
       "}\n",
       "\n",
       "div.sk-parallel-item,\n",
       "div.sk-serial,\n",
       "div.sk-item {\n",
       "  /* draw centered vertical line to link estimators */\n",
       "  background-image: linear-gradient(var(--sklearn-color-text-on-default-background), var(--sklearn-color-text-on-default-background));\n",
       "  background-size: 2px 100%;\n",
       "  background-repeat: no-repeat;\n",
       "  background-position: center center;\n",
       "}\n",
       "\n",
       "/* Parallel-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item::after {\n",
       "  content: \"\";\n",
       "  width: 100%;\n",
       "  border-bottom: 2px solid var(--sklearn-color-text-on-default-background);\n",
       "  flex-grow: 1;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel {\n",
       "  display: flex;\n",
       "  align-items: stretch;\n",
       "  justify-content: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  position: relative;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:first-child::after {\n",
       "  align-self: flex-end;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:last-child::after {\n",
       "  align-self: flex-start;\n",
       "  width: 50%;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-parallel-item:only-child::after {\n",
       "  width: 0;\n",
       "}\n",
       "\n",
       "/* Serial-specific style estimator block */\n",
       "\n",
       "#sk-container-id-1 div.sk-serial {\n",
       "  display: flex;\n",
       "  flex-direction: column;\n",
       "  align-items: center;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  padding-right: 1em;\n",
       "  padding-left: 1em;\n",
       "}\n",
       "\n",
       "\n",
       "/* Toggleable style: style used for estimator/Pipeline/ColumnTransformer box that is\n",
       "clickable and can be expanded/collapsed.\n",
       "- Pipeline and ColumnTransformer use this feature and define the default style\n",
       "- Estimators will overwrite some part of the style using the `sk-estimator` class\n",
       "*/\n",
       "\n",
       "/* Pipeline and ColumnTransformer style (default) */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable {\n",
       "  /* Default theme specific background. It is overwritten whether we have a\n",
       "  specific estimator or a Pipeline/ColumnTransformer */\n",
       "  background-color: var(--sklearn-color-background);\n",
       "}\n",
       "\n",
       "/* Toggleable label */\n",
       "#sk-container-id-1 label.sk-toggleable__label {\n",
       "  cursor: pointer;\n",
       "  display: flex;\n",
       "  width: 100%;\n",
       "  margin-bottom: 0;\n",
       "  padding: 0.5em;\n",
       "  box-sizing: border-box;\n",
       "  text-align: center;\n",
       "  align-items: start;\n",
       "  justify-content: space-between;\n",
       "  gap: 0.5em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label .caption {\n",
       "  font-size: 0.6rem;\n",
       "  font-weight: lighter;\n",
       "  color: var(--sklearn-color-text-muted);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:before {\n",
       "  /* Arrow on the left of the label */\n",
       "  content: \"▸\";\n",
       "  float: left;\n",
       "  margin-right: 0.25em;\n",
       "  color: var(--sklearn-color-icon);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {\n",
       "  color: var(--sklearn-color-text);\n",
       "}\n",
       "\n",
       "/* Toggleable content - dropdown */\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content {\n",
       "  display: none;\n",
       "  text-align: left;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content pre {\n",
       "  margin: 0.2em;\n",
       "  border-radius: 0.25em;\n",
       "  color: var(--sklearn-color-text);\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-toggleable__content.fitted pre {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {\n",
       "  /* Expand drop-down */\n",
       "  display: block;\n",
       "  width: 100%;\n",
       "  overflow: visible;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {\n",
       "  content: \"▾\";\n",
       "}\n",
       "\n",
       "/* Pipeline/ColumnTransformer-specific style */\n",
       "\n",
       "#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator-specific style */\n",
       "\n",
       "/* Colorize estimator box */\n",
       "#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted input.sk-toggleable__control:checked~label.sk-toggleable__label {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label label.sk-toggleable__label,\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  /* The background is the default theme color */\n",
       "  color: var(--sklearn-color-text-on-default-background);\n",
       "}\n",
       "\n",
       "/* On hover, darken the color of the background */\n",
       "#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "/* Label box, darken color on hover, fitted */\n",
       "#sk-container-id-1 div.sk-label.fitted:hover label.sk-toggleable__label.fitted {\n",
       "  color: var(--sklearn-color-text);\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Estimator label */\n",
       "\n",
       "#sk-container-id-1 div.sk-label label {\n",
       "  font-family: monospace;\n",
       "  font-weight: bold;\n",
       "  display: inline-block;\n",
       "  line-height: 1.2em;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-label-container {\n",
       "  text-align: center;\n",
       "}\n",
       "\n",
       "/* Estimator-specific */\n",
       "#sk-container-id-1 div.sk-estimator {\n",
       "  font-family: monospace;\n",
       "  border: 1px dotted var(--sklearn-color-border-box);\n",
       "  border-radius: 0.25em;\n",
       "  box-sizing: border-box;\n",
       "  margin-bottom: 0.5em;\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-0);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-0);\n",
       "}\n",
       "\n",
       "/* on hover */\n",
       "#sk-container-id-1 div.sk-estimator:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-2);\n",
       "}\n",
       "\n",
       "#sk-container-id-1 div.sk-estimator.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-2);\n",
       "}\n",
       "\n",
       "/* Specification for estimator info (e.g. \"i\" and \"?\") */\n",
       "\n",
       "/* Common style for \"i\" and \"?\" */\n",
       "\n",
       ".sk-estimator-doc-link,\n",
       "a:link.sk-estimator-doc-link,\n",
       "a:visited.sk-estimator-doc-link {\n",
       "  float: right;\n",
       "  font-size: smaller;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1em;\n",
       "  height: 1em;\n",
       "  width: 1em;\n",
       "  text-decoration: none !important;\n",
       "  margin-left: 0.5em;\n",
       "  text-align: center;\n",
       "  /* unfitted */\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted,\n",
       "a:link.sk-estimator-doc-link.fitted,\n",
       "a:visited.sk-estimator-doc-link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "div.sk-estimator:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link:hover,\n",
       ".sk-estimator-doc-link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "div.sk-estimator.fitted:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover,\n",
       "div.sk-label-container:hover .sk-estimator-doc-link.fitted:hover,\n",
       ".sk-estimator-doc-link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "/* Span, style for the box shown on hovering the info icon */\n",
       ".sk-estimator-doc-link span {\n",
       "  display: none;\n",
       "  z-index: 9999;\n",
       "  position: relative;\n",
       "  font-weight: normal;\n",
       "  right: .2ex;\n",
       "  padding: .5ex;\n",
       "  margin: .5ex;\n",
       "  width: min-content;\n",
       "  min-width: 20ex;\n",
       "  max-width: 50ex;\n",
       "  color: var(--sklearn-color-text);\n",
       "  box-shadow: 2pt 2pt 4pt #999;\n",
       "  /* unfitted */\n",
       "  background: var(--sklearn-color-unfitted-level-0);\n",
       "  border: .5pt solid var(--sklearn-color-unfitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link.fitted span {\n",
       "  /* fitted */\n",
       "  background: var(--sklearn-color-fitted-level-0);\n",
       "  border: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".sk-estimator-doc-link:hover span {\n",
       "  display: block;\n",
       "}\n",
       "\n",
       "/* \"?\"-specific style due to the `<a>` HTML tag */\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link {\n",
       "  float: right;\n",
       "  font-size: 1rem;\n",
       "  line-height: 1em;\n",
       "  font-family: monospace;\n",
       "  background-color: var(--sklearn-color-background);\n",
       "  border-radius: 1rem;\n",
       "  height: 1rem;\n",
       "  width: 1rem;\n",
       "  text-decoration: none;\n",
       "  /* unfitted */\n",
       "  color: var(--sklearn-color-unfitted-level-1);\n",
       "  border: var(--sklearn-color-unfitted-level-1) 1pt solid;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted {\n",
       "  /* fitted */\n",
       "  border: var(--sklearn-color-fitted-level-1) 1pt solid;\n",
       "  color: var(--sklearn-color-fitted-level-1);\n",
       "}\n",
       "\n",
       "/* On hover */\n",
       "#sk-container-id-1 a.estimator_doc_link:hover {\n",
       "  /* unfitted */\n",
       "  background-color: var(--sklearn-color-unfitted-level-3);\n",
       "  color: var(--sklearn-color-background);\n",
       "  text-decoration: none;\n",
       "}\n",
       "\n",
       "#sk-container-id-1 a.estimator_doc_link.fitted:hover {\n",
       "  /* fitted */\n",
       "  background-color: var(--sklearn-color-fitted-level-3);\n",
       "}\n",
       "\n",
       ".estimator-table summary {\n",
       "    padding: .5rem;\n",
       "    font-family: monospace;\n",
       "    cursor: pointer;\n",
       "}\n",
       "\n",
       ".estimator-table details[open] {\n",
       "    padding-left: 0.1rem;\n",
       "    padding-right: 0.1rem;\n",
       "    padding-bottom: 0.3rem;\n",
       "}\n",
       "\n",
       ".estimator-table .parameters-table {\n",
       "    margin-left: auto !important;\n",
       "    margin-right: auto !important;\n",
       "}\n",
       "\n",
       ".estimator-table .parameters-table tr:nth-child(odd) {\n",
       "    background-color: #fff;\n",
       "}\n",
       "\n",
       ".estimator-table .parameters-table tr:nth-child(even) {\n",
       "    background-color: #f6f6f6;\n",
       "}\n",
       "\n",
       ".estimator-table .parameters-table tr:hover {\n",
       "    background-color: #e0e0e0;\n",
       "}\n",
       "\n",
       ".estimator-table table td {\n",
       "    border: 1px solid rgba(106, 105, 104, 0.232);\n",
       "}\n",
       "\n",
       ".user-set td {\n",
       "    color:rgb(255, 94, 0);\n",
       "    text-align: left;\n",
       "}\n",
       "\n",
       ".user-set td.value pre {\n",
       "    color:rgb(255, 94, 0) !important;\n",
       "    background-color: transparent !important;\n",
       "}\n",
       "\n",
       ".default td {\n",
       "    color: black;\n",
       "    text-align: left;\n",
       "}\n",
       "\n",
       ".user-set td i,\n",
       ".default td i {\n",
       "    color: black;\n",
       "}\n",
       "\n",
       ".copy-paste-icon {\n",
       "    background-image: url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZpZXdCb3g9IjAgMCA0NDggNTEyIj48IS0tIUZvbnQgQXdlc29tZSBGcmVlIDYuNy4yIGJ5IEBmb250YXdlc29tZSAtIGh0dHBzOi8vZm9udGF3ZXNvbWUuY29tIExpY2Vuc2UgLSBodHRwczovL2ZvbnRhd2Vzb21lLmNvbS9saWNlbnNlL2ZyZWUgQ29weXJpZ2h0IDIwMjUgRm9udGljb25zLCBJbmMuLS0+PHBhdGggZD0iTTIwOCAwTDMzMi4xIDBjMTIuNyAwIDI0LjkgNS4xIDMzLjkgMTQuMWw2Ny45IDY3LjljOSA5IDE0LjEgMjEuMiAxNC4xIDMzLjlMNDQ4IDMzNmMwIDI2LjUtMjEuNSA0OC00OCA0OGwtMTkyIDBjLTI2LjUgMC00OC0yMS41LTQ4LTQ4bDAtMjg4YzAtMjYuNSAyMS41LTQ4IDQ4LTQ4ek00OCAxMjhsODAgMCAwIDY0LTY0IDAgMCAyNTYgMTkyIDAgMC0zMiA2NCAwIDAgNDhjMCAyNi41LTIxLjUgNDgtNDggNDhMNDggNTEyYy0yNi41IDAtNDgtMjEuNS00OC00OEwwIDE3NmMwLTI2LjUgMjEuNS00OCA0OC00OHoiLz48L3N2Zz4=);\n",
       "    background-repeat: no-repeat;\n",
       "    background-size: 14px 14px;\n",
       "    background-position: 0;\n",
       "    display: inline-block;\n",
       "    width: 14px;\n",
       "    height: 14px;\n",
       "    cursor: pointer;\n",
       "}\n",
       "</style><body><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>XGBRanker(base_score=None, booster=None, callbacks=None, colsample_bylevel=None,\n",
       "          colsample_bynode=None, colsample_bytree=None, device=None,\n",
       "          early_stopping_rounds=None, enable_categorical=False,\n",
       "          eval_metric=None, feature_types=None, feature_weights=None,\n",
       "          gamma=None, grow_policy=None, importance_type=None,\n",
       "          interaction_constraints=None, lambdarank_num_pair_per_sample=200,\n",
       "          lambdarank_pair_method=&#x27;topk&#x27;, learning_rate=None, max_bin=None,\n",
       "          max_cat_threshold=None, max_cat_to_onehot=None, max_delta_step=None,\n",
       "          max_depth=None, max_leaves=None, min_child_weight=None, missing=nan,\n",
       "          monotone_constraints=None, multi_strategy=None, n_estimators=None, ...)</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator fitted sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label fitted sk-toggleable__label-arrow\"><div><div>XGBRanker</div></div><div><a class=\"sk-estimator-doc-link fitted\" rel=\"noreferrer\" target=\"_blank\" href=\"https://xgboost.readthedocs.io/en/release_3.0.0/python/python_api.html#xgboost.XGBRanker\">?<span>Documentation for XGBRanker</span></a><span class=\"sk-estimator-doc-link fitted\">i<span>Fitted</span></span></div></label><div class=\"sk-toggleable__content fitted\" data-param-prefix=\"\">\n",
       "        <div class=\"estimator-table\">\n",
       "            <details>\n",
       "                <summary>Parameters</summary>\n",
       "                <table class=\"parameters-table\">\n",
       "                  <tbody>\n",
       "                    \n",
       "        <tr class=\"default\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('objective',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">objective&nbsp;</td>\n",
       "            <td class=\"value\">&#x27;rank:ndcg&#x27;</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('base_score',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">base_score&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('booster',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">booster&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('callbacks',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">callbacks&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('colsample_bylevel',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">colsample_bylevel&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('colsample_bynode',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">colsample_bynode&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('colsample_bytree',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">colsample_bytree&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('device',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">device&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('early_stopping_rounds',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">early_stopping_rounds&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('enable_categorical',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">enable_categorical&nbsp;</td>\n",
       "            <td class=\"value\">False</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('eval_metric',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">eval_metric&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('feature_types',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">feature_types&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('feature_weights',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">feature_weights&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('gamma',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">gamma&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('grow_policy',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">grow_policy&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('importance_type',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">importance_type&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('interaction_constraints',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">interaction_constraints&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('learning_rate',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">learning_rate&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('max_bin',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">max_bin&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('max_cat_threshold',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">max_cat_threshold&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('max_cat_to_onehot',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">max_cat_to_onehot&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('max_delta_step',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">max_delta_step&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('max_depth',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">max_depth&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('max_leaves',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">max_leaves&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('min_child_weight',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">min_child_weight&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('missing',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">missing&nbsp;</td>\n",
       "            <td class=\"value\">nan</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('monotone_constraints',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">monotone_constraints&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('multi_strategy',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">multi_strategy&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('n_estimators',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">n_estimators&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('n_jobs',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">n_jobs&nbsp;</td>\n",
       "            <td class=\"value\">-1</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('num_parallel_tree',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">num_parallel_tree&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('random_state',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">random_state&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('reg_alpha',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">reg_alpha&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('reg_lambda',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">reg_lambda&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('sampling_method',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">sampling_method&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('scale_pos_weight',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">scale_pos_weight&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('subsample',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">subsample&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('tree_method',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">tree_method&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('validate_parameters',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">validate_parameters&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('verbosity',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">verbosity&nbsp;</td>\n",
       "            <td class=\"value\">None</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('ndcg_exp_gain',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">ndcg_exp_gain&nbsp;</td>\n",
       "            <td class=\"value\">False</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('lambdarank_pair_method',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">lambdarank_pair_method&nbsp;</td>\n",
       "            <td class=\"value\">&#x27;topk&#x27;</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('lambdarank_num_pair_per_sample',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">lambdarank_num_pair_per_sample&nbsp;</td>\n",
       "            <td class=\"value\">200</td>\n",
       "        </tr>\n",
       "    \n",
       "\n",
       "        <tr class=\"user-set\">\n",
       "            <td><i class=\"copy-paste-icon\"\n",
       "                 onclick=\"copyToClipboard('seed',\n",
       "                          this.parentElement.nextElementSibling)\"\n",
       "            ></i></td>\n",
       "            <td class=\"param\">seed&nbsp;</td>\n",
       "            <td class=\"value\">42</td>\n",
       "        </tr>\n",
       "    \n",
       "                  </tbody>\n",
       "                </table>\n",
       "            </details>\n",
       "        </div>\n",
       "    </div></div></div></div></div><script>function copyToClipboard(text, element) {\n",
       "    // Get the parameter prefix from the closest toggleable content\n",
       "    const toggleableContent = element.closest('.sk-toggleable__content');\n",
       "    const paramPrefix = toggleableContent ? toggleableContent.dataset.paramPrefix : '';\n",
       "    const fullParamName = paramPrefix ? `${paramPrefix}${text}` : text;\n",
       "\n",
       "    const originalStyle = element.style;\n",
       "    const computedStyle = window.getComputedStyle(element);\n",
       "    const originalWidth = computedStyle.width;\n",
       "    const originalHTML = element.innerHTML.replace('Copied!', '');\n",
       "\n",
       "    navigator.clipboard.writeText(fullParamName)\n",
       "        .then(() => {\n",
       "            element.style.width = originalWidth;\n",
       "            element.style.color = 'green';\n",
       "            element.innerHTML = \"Copied!\";\n",
       "\n",
       "            setTimeout(() => {\n",
       "                element.innerHTML = originalHTML;\n",
       "                element.style = originalStyle;\n",
       "            }, 2000);\n",
       "        })\n",
       "        .catch(err => {\n",
       "            console.error('Failed to copy:', err);\n",
       "            element.style.color = 'red';\n",
       "            element.innerHTML = \"Failed!\";\n",
       "            setTimeout(() => {\n",
       "                element.innerHTML = originalHTML;\n",
       "                element.style = originalStyle;\n",
       "            }, 2000);\n",
       "        });\n",
       "    return false;\n",
       "}\n",
       "\n",
       "document.querySelectorAll('.fa-regular.fa-copy').forEach(function(element) {\n",
       "    const toggleableContent = element.closest('.sk-toggleable__content');\n",
       "    const paramPrefix = toggleableContent ? toggleableContent.dataset.paramPrefix : '';\n",
       "    const paramName = element.parentElement.nextElementSibling.textContent.trim();\n",
       "    const fullParamName = paramPrefix ? `${paramPrefix}${paramName}` : paramName;\n",
       "\n",
       "    element.setAttribute('title', fullParamName);\n",
       "});\n",
       "</script></body>"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 21
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:59.854356Z",
     "start_time": "2025-09-17T21:07:59.838819Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Check the wellness of fit\n",
    "predicted = model.predict(X[selected_features].values)\n",
    "\n",
    "corr, _ = pearsonr(predicted, y)\n",
    "corrs, _ = spearmanr(predicted, y)\n",
    "\n",
    "print(\n",
    "    f\"Pearson {corr}, Spearman {corrs}\")\n"
   ],
   "id": "343b8a36c5a4445f",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Pearson 0.5207460562690249, Spearman 0.48690341402197834\n"
     ]
    }
   ],
   "execution_count": 22
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:59.903089Z",
     "start_time": "2025-09-17T21:07:59.893394Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from sklearn.metrics import ndcg_score\n",
    "\n",
    "y_true_corrected = df_copy[metric + 'ndcg'].to_numpy()\n",
    "\n",
    "# ---- New metrics ----\n",
    "# NDCG@50\n",
    "ndcg50 = ndcg_score(y_true_corrected.reshape(1, -1), predicted.reshape(1, -1), k=200)\n",
    "# NDCG (all, just drop k)\n",
    "ndcg_all = ndcg_score(y_true_corrected.reshape(1, -1), predicted.reshape(1, -1))\n",
    "# Precision@50 (manual overlap of top-50 by truth vs pred)\n",
    "precisions = []\n",
    "values = [50, 100]\n",
    "for K in values:\n",
    "    pred_top_idx = np.argpartition(predicted, -K)[-K:]\n",
    "    true_top_idx = np.argpartition(y_true_corrected, -K)[-K:]\n",
    "    precisions.append(len(set(pred_top_idx) & set(true_top_idx)) / K)\n",
    "\n",
    "print(f\"NDCG@50: {ndcg50:.4f}, NDCG(all): {ndcg_all:.4f}, Precision: {values, precisions}\")"
   ],
   "id": "e1d9d3ac13c39f10",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NDCG@50: 0.8041, NDCG(all): 0.9823, Precision: ([50, 100], [0.28, 0.36])\n"
     ]
    }
   ],
   "execution_count": 23
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:07:59.963828Z",
     "start_time": "2025-09-17T21:07:59.961795Z"
    }
   },
   "cell_type": "code",
   "source": "# Very impressive results above, but go figure",
   "id": "12da4dbe1607e96b",
   "outputs": [],
   "execution_count": 24
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:08:00.013461Z",
     "start_time": "2025-09-17T21:08:00.011389Z"
    }
   },
   "cell_type": "code",
   "source": [
    "TARGETS = ['MALAT1']\n",
    "SEQUENCES = {}\n",
    "for target in TARGETS:\n",
    "    SEQUENCES[target] = gene_to_data[target].full_mrna"
   ],
   "id": "859d6de7800098e1",
   "outputs": [],
   "execution_count": 25
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:08:00.090139Z",
     "start_time": "2025-09-17T21:08:00.061577Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from asodesigner.util import get_antisense\n",
    "import pandas as pd\n",
    "\n",
    "def get_init_df(target_mrna, end):\n",
    "    candidates = []\n",
    "    sense_starts = []\n",
    "    sense_lengths = []\n",
    "    sense_starts_from_end = []\n",
    "\n",
    "    for i in range(0, len(target_mrna) - 19):\n",
    "        target = target_mrna[i: i + 20]\n",
    "        candidates.append(get_antisense(str(target)))\n",
    "        sense_starts.append(i)\n",
    "        sense_lengths.append(20)\n",
    "        sense_starts_from_end.append(end - i)\n",
    "    df = pd.DataFrame(\n",
    "        {SEQUENCE: candidates, SENSE_START: sense_starts,\n",
    "         SENSE_LENGTH: sense_lengths, \"sense_start_from_end\": sense_starts_from_end})\n",
    "    return df\n",
    "\n",
    "dfs = {}\n",
    "for target in TARGETS:\n",
    "    gene_info = gene_to_data[target]\n",
    "    dfs[target] = get_init_df(gene_info.full_mrna, gene_info.exon_indices[-1][1] - gene_info.cds_start)"
   ],
   "id": "b1c63d66730ac4f3",
   "outputs": [],
   "execution_count": 26
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:08:00.143999Z",
     "start_time": "2025-09-17T21:08:00.141524Z"
    }
   },
   "cell_type": "code",
   "source": [
    "for gene, df in dfs.items():\n",
    "    df[CANONICAL_GENE] = gene"
   ],
   "id": "6bcfe4c25111f546",
   "outputs": [],
   "execution_count": 27
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:08:00.195297Z",
     "start_time": "2025-09-17T21:08:00.192541Z"
    }
   },
   "cell_type": "code",
   "source": [
    "for df in dfs.values():\n",
    "    df[CELL_LINE_ORGANISM] = 'human'\n",
    "    df[INHIBITION] = 0 # Just for the function, not important"
   ],
   "id": "4b0ab969f2950e76",
   "outputs": [],
   "execution_count": 28
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:08:05.038427Z",
     "start_time": "2025-09-17T21:08:00.273175Z"
    }
   },
   "cell_type": "code",
   "source": [
    "for gene, df in dfs.items():\n",
    "    df = get_populated_df_with_structure_features(df, TARGETS, gene_to_data)\n",
    "    dfs[gene] = df"
   ],
   "id": "c71dc0be7247deac",
   "outputs": [],
   "execution_count": 29
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:08:07.644803Z",
     "start_time": "2025-09-17T21:08:05.041240Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from scripts.data_genertion.data_handling import populate_features\n",
    "\n",
    "for gene, df in dfs.items():\n",
    "    df[TREATMENT_PERIOD] = 24  # keep constant for all\n",
    "    df[VOLUME] = 1000  # keep constant for all\n",
    "    df['log_volume'] = np.log(df[VOLUME])\n",
    "    df['normalized_start'] = df[SENSE_START] / len(SEQUENCES[gene])\n",
    "    df['normalized_sense_start_from_end'] = df['sense_start_from_end'] / len(SEQUENCES[gene])\n",
    "    easy_to_populate = ['at_skew', 'gc_content', 'gc_content_3_prime_5', 'gc_skew', 'hairpin_score',\n",
    "                        'homooligo_count', 'internal_fold', 'nucleotide_diversity', 'self_energy', 'stop_codon_count',\n",
    "                        'at_rich_region_score', 'poly_pyrimidine_stretch']\n",
    "    populate_features(df, easy_to_populate)"
   ],
   "id": "3d9fd2f39bd51fd4",
   "outputs": [],
   "execution_count": 30
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:08:10.245621Z",
     "start_time": "2025-09-17T21:08:07.670080Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from scripts.data_genertion.data_handling import get_populate_fold\n",
    "\n",
    "for gene, df in dfs.items():\n",
    "    fold_variants = [(40, 15)]\n",
    "    df = get_populate_fold(df, TARGETS, gene_to_data, fold_variants=fold_variants)\n",
    "    dfs[gene] = df"
   ],
   "id": "90909bb74cc573a5",
   "outputs": [],
   "execution_count": 31
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:08:10.288392Z",
     "start_time": "2025-09-17T21:08:10.267303Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from hybridization.hybridization_features import get_exp_psrna_hybridization\n",
    "\n",
    "for gene, df in dfs.items():\n",
    "    df.loc[:, 'exp_ps_hybr'] = [\n",
    "        get_exp_psrna_hybridization(antisense.replace('T', 'U'), temp=37) for\n",
    "        antisense in df[SEQUENCE]]"
   ],
   "id": "3640b43650283898",
   "outputs": [],
   "execution_count": 32
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:08:10.329476Z",
     "start_time": "2025-09-17T21:08:10.327039Z"
    }
   },
   "cell_type": "code",
   "source": "# Missing: CAI_score_global_CDS, 'sense_avg_accessibility', RNaseH1_Krel_score_R7_krel, Modification_min_distance_to_3prime",
   "id": "9d6f852e3a82470f",
   "outputs": [],
   "execution_count": 33
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:08:10.380713Z",
     "start_time": "2025-09-17T21:08:10.376141Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from features.mod_features import compute_mod_min_distance_to_3prime\n",
    "\n",
    "# generate MOE 20-mers\n",
    "for gene, df in dfs.items():\n",
    "    df.loc[:, 'Modification_min_distance_to_3prime'] = compute_mod_min_distance_to_3prime('MMMMMddddddddddMMMMM')"
   ],
   "id": "a74f497ab4fc00a1",
   "outputs": [],
   "execution_count": 34
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:40:09.936376Z",
     "start_time": "2025-09-17T21:15:50.326308Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from yehuda_code.Folding_Functions import get_sense_with_flanks\n",
    "from yehuda_code.access_calculator import AccessCalculator\n",
    "\n",
    "\n",
    "\n",
    "FLANK_SIZE = 120\n",
    "ACCESS_SIZE = 13\n",
    "SEED_SIZE = 13\n",
    "SEED_SIZES = [SEED_SIZE * m for m in range(1, 4)]\n",
    "ACCESS_WIN_SIZE = 80\n",
    "\n",
    "\n",
    "def compute_sense_accessibility(row, flank_size, access_win_size, seed_sizes, access_size, min_gc=0, max_gc=100, gc_ranges=1):\n",
    "\n",
    "\n",
    "    try:\n",
    "        # Skip invalid rows\n",
    "        if row['sense_start'] == -1 or pd.isna(row['sense_with_flank_120nt']) or row['sense_with_flank_120nt'] == \"\":\n",
    "            return None\n",
    "\n",
    "        seq = row[f'sense_with_flank_{flank_size}nt']\n",
    "        sense_start = row['sense_start']\n",
    "        sense_length = row['sense_length']\n",
    "\n",
    "        # Calculate accessibility\n",
    "        df_access = AccessCalculator.calc(\n",
    "            seq, access_size,\n",
    "            min_gc, max_gc, gc_ranges,\n",
    "            access_win_size, seed_sizes\n",
    "        )\n",
    "\n",
    "        flank_start = max(0, sense_start - flank_size)\n",
    "        sense_start_in_flank = sense_start - flank_start\n",
    "        sense_end_in_flank = sense_start_in_flank + sense_length\n",
    "\n",
    "        if 0 <= sense_start_in_flank < len(df_access) and sense_end_in_flank <= len(df_access):\n",
    "            values = df_access['avg_access'].iloc[sense_start_in_flank:sense_end_in_flank].dropna()\n",
    "            return values.mean() if not values.empty else None\n",
    "        else:\n",
    "            return None\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error at row {row.name} | seq start: {row['sense_start']} | error: {e}\")\n",
    "        return None\n",
    "\n",
    "\n",
    "for gene, df in dfs.items():\n",
    "    FLANKED_SENSE_COL = f'sense_with_flank_{FLANK_SIZE}nt'\n",
    "\n",
    "    val = gene_to_data[gene].full_mrna\n",
    "    df['pre_mrna_sequence'] = [val] * len(df)\n",
    "\n",
    "\n",
    "    # Create new column with flanked sequences\n",
    "    df[FLANKED_SENSE_COL] = df.apply(\n",
    "    lambda row: get_sense_with_flanks(\n",
    "        row['pre_mrna_sequence'],\n",
    "        row['sense_start'],\n",
    "        row['sense_length'],\n",
    "        flank_size=FLANK_SIZE\n",
    "    ) if row['sense_start'] != -1 else \"\",  # Handle cases where sense was not found\n",
    "    axis=1\n",
    "    )\n",
    "\n",
    "\n",
    "    batch_size = 500\n",
    "    for start_idx in range(0, len(df), batch_size):\n",
    "        end_idx = min(start_idx + batch_size, len(df))\n",
    "        batch = df.iloc[start_idx:end_idx].copy()\n",
    "\n",
    "        print(f\"Processing rows {start_idx} to {end_idx}...\")\n",
    "\n",
    "        batch['sense_avg_accessibility'] = batch.apply(\n",
    "            compute_sense_accessibility,\n",
    "            axis=1,\n",
    "            flank_size=FLANK_SIZE,\n",
    "            access_win_size=ACCESS_WIN_SIZE,\n",
    "            seed_sizes=SEED_SIZES,\n",
    "            access_size=ACCESS_SIZE,\n",
    "        )\n",
    "\n",
    "        # Save batch to the new folder\n",
    "        batch.to_csv(f\"out/batch_{start_idx}_{end_idx}.csv\", index=False)\n"
   ],
   "id": "82f108cada6ff835",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing rows 0 to 500...\n",
      "Processing rows 500 to 1000...\n",
      "Processing rows 1000 to 1500...\n",
      "Processing rows 1500 to 2000...\n",
      "Processing rows 2000 to 2500...\n",
      "Processing rows 2500 to 3000...\n",
      "Processing rows 3000 to 3500...\n",
      "Processing rows 3500 to 4000...\n",
      "Processing rows 4000 to 4500...\n",
      "Processing rows 4500 to 5000...\n",
      "Processing rows 5000 to 5500...\n",
      "Processing rows 5500 to 6000...\n",
      "Processing rows 6000 to 6500...\n",
      "Processing rows 6500 to 7000...\n",
      "Processing rows 7000 to 7500...\n",
      "Processing rows 7500 to 8000...\n",
      "Processing rows 8000 to 8500...\n",
      "Processing rows 8500 to 8810...\n"
     ]
    }
   ],
   "execution_count": 45
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:43:38.973052Z",
     "start_time": "2025-09-17T21:43:38.758442Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "\n",
    "# Load all batch files from the new output folder\n",
    "files = sorted(glob.glob(f\"out/batch_*.csv\"))\n",
    "df_all = pd.concat([pd.read_csv(f) for f in files], axis=0)\n"
   ],
   "id": "b78313c380856e88",
   "outputs": [],
   "execution_count": 46
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:43:42.378878Z",
     "start_time": "2025-09-17T21:43:42.366988Z"
    }
   },
   "cell_type": "code",
   "source": "df_all",
   "id": "fb21bc0fc19edeee",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "                 Sequence  sense_start  sense_length  sense_start_from_end  \\\n",
       "0    GCAACCGGTGGGGCTGCGTC            0            20                  8829   \n",
       "1    CGCAACCGGTGGGGCTGCGT            1            20                  8828   \n",
       "2    GCGCAACCGGTGGGGCTGCG            2            20                  8827   \n",
       "3    TGCGCAACCGGTGGGGCTGC            3            20                  8826   \n",
       "4    CTGCGCAACCGGTGGGGCTG            4            20                  8825   \n",
       "..                    ...          ...           ...                   ...   \n",
       "305  TTCCCCAATCAAGATTTTTT         8805            20                    24   \n",
       "306  TTTCCCCAATCAAGATTTTT         8806            20                    23   \n",
       "307  TTTTCCCCAATCAAGATTTT         8807            20                    22   \n",
       "308  TTTTTCCCCAATCAAGATTT         8808            20                    21   \n",
       "309  TTTTTTCCCCAATCAAGATT         8809            20                    20   \n",
       "\n",
       "    Canonical Gene Name Cell line organism  Inhibition(%)  sense_exon  \\\n",
       "0                MALAT1              human              0           1   \n",
       "1                MALAT1              human              0           1   \n",
       "2                MALAT1              human              0           1   \n",
       "3                MALAT1              human              0           1   \n",
       "4                MALAT1              human              0           1   \n",
       "..                  ...                ...            ...         ...   \n",
       "305              MALAT1              human              0           1   \n",
       "306              MALAT1              human              0           1   \n",
       "307              MALAT1              human              0           1   \n",
       "308              MALAT1              human              0           1   \n",
       "309              MALAT1              human              0           1   \n",
       "\n",
       "     sense_intron  sense_utr  ... on_target_fold_openness40_15  \\\n",
       "0               0          0  ...                   -10.062500   \n",
       "1               0          0  ...                    -9.875000   \n",
       "2               0          0  ...                    -9.687500   \n",
       "3               0          0  ...                    -9.500000   \n",
       "4               0          0  ...                    -9.312500   \n",
       "..            ...        ...  ...                          ...   \n",
       "305             0          0  ...                    -0.333333   \n",
       "306             0          0  ...                    -0.300000   \n",
       "307             0          0  ...                    -0.266667   \n",
       "308             0          0  ...                    -0.233333   \n",
       "309             0          0  ...                    -0.200000   \n",
       "\n",
       "     on_target_fold_openness_normalized40_15  exp_ps_hybr  \\\n",
       "0                                  -0.503125        -2923   \n",
       "1                                  -0.493750        -2941   \n",
       "2                                  -0.484375        -2956   \n",
       "3                                  -0.475000        -2907   \n",
       "4                                  -0.465625        -2867   \n",
       "..                                       ...          ...   \n",
       "305                                -0.016667        -2044   \n",
       "306                                -0.015000        -2044   \n",
       "307                                -0.013333        -2044   \n",
       "308                                -0.011667        -2044   \n",
       "309                                -0.010000        -2044   \n",
       "\n",
       "     Modification_min_distance_to_3prime  RNaseH1_Krel_score_R4a_krel_Krel  \\\n",
       "0                                    0.0                          0.978831   \n",
       "1                                    0.0                          0.949957   \n",
       "2                                    0.0                          0.980212   \n",
       "3                                    0.0                          0.968729   \n",
       "4                                    0.0                          0.958289   \n",
       "..                                   ...                               ...   \n",
       "305                                  0.0                          0.991729   \n",
       "306                                  0.0                          0.978869   \n",
       "307                                  0.0                          1.068860   \n",
       "308                                  0.0                          1.008057   \n",
       "309                                  0.0                          0.985967   \n",
       "\n",
       "     RNaseH1_Krel_score_R4b_krel_Krel  RNaseH1_Krel_score_R7_krel_Krel  \\\n",
       "0                            1.004691                         1.025778   \n",
       "1                            1.029123                         1.015559   \n",
       "2                            1.002733                         0.990129   \n",
       "3                            0.989912                         0.972045   \n",
       "4                            0.988861                         0.957519   \n",
       "..                                ...                              ...   \n",
       "305                          0.957326                         0.986575   \n",
       "306                          0.959173                         1.013243   \n",
       "307                          1.000271                         0.982115   \n",
       "308                          0.983111                         0.992385   \n",
       "309                          0.958838                         0.993042   \n",
       "\n",
       "                                     pre_mrna_sequence  \\\n",
       "0    GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "1    GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "2    GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "3    GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "4    GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "..                                                 ...   \n",
       "305  GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "306  GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "307  GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "308  GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "309  GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "\n",
       "                                sense_with_flank_120nt  \\\n",
       "0    GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "1    GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "2    GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "3    GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "4    GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...   \n",
       "..                                                 ...   \n",
       "305  TTGTCTTTTTCAGGTAATAGCCTGCAGCTGGTGTTTTGAGAAGCCC...   \n",
       "306  TGTCTTTTTCAGGTAATAGCCTGCAGCTGGTGTTTTGAGAAGCCCT...   \n",
       "307  GTCTTTTTCAGGTAATAGCCTGCAGCTGGTGTTTTGAGAAGCCCTA...   \n",
       "308  TCTTTTTCAGGTAATAGCCTGCAGCTGGTGTTTTGAGAAGCCCTAC...   \n",
       "309  CTTTTTCAGGTAATAGCCTGCAGCTGGTGTTTTGAGAAGCCCTACT...   \n",
       "\n",
       "     sense_avg_accessibility  \n",
       "0                   8.867892  \n",
       "1                   8.812901  \n",
       "2                   8.846318  \n",
       "3                   8.633534  \n",
       "4                   8.560287  \n",
       "..                       ...  \n",
       "305                      NaN  \n",
       "306                      NaN  \n",
       "307                      NaN  \n",
       "308                      NaN  \n",
       "309                      NaN  \n",
       "\n",
       "[8810 rows x 38 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Sequence</th>\n",
       "      <th>sense_start</th>\n",
       "      <th>sense_length</th>\n",
       "      <th>sense_start_from_end</th>\n",
       "      <th>Canonical Gene Name</th>\n",
       "      <th>Cell line organism</th>\n",
       "      <th>Inhibition(%)</th>\n",
       "      <th>sense_exon</th>\n",
       "      <th>sense_intron</th>\n",
       "      <th>sense_utr</th>\n",
       "      <th>...</th>\n",
       "      <th>on_target_fold_openness40_15</th>\n",
       "      <th>on_target_fold_openness_normalized40_15</th>\n",
       "      <th>exp_ps_hybr</th>\n",
       "      <th>Modification_min_distance_to_3prime</th>\n",
       "      <th>RNaseH1_Krel_score_R4a_krel_Krel</th>\n",
       "      <th>RNaseH1_Krel_score_R4b_krel_Krel</th>\n",
       "      <th>RNaseH1_Krel_score_R7_krel_Krel</th>\n",
       "      <th>pre_mrna_sequence</th>\n",
       "      <th>sense_with_flank_120nt</th>\n",
       "      <th>sense_avg_accessibility</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>GCAACCGGTGGGGCTGCGTC</td>\n",
       "      <td>0</td>\n",
       "      <td>20</td>\n",
       "      <td>8829</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-10.062500</td>\n",
       "      <td>-0.503125</td>\n",
       "      <td>-2923</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.978831</td>\n",
       "      <td>1.004691</td>\n",
       "      <td>1.025778</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>8.867892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CGCAACCGGTGGGGCTGCGT</td>\n",
       "      <td>1</td>\n",
       "      <td>20</td>\n",
       "      <td>8828</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.875000</td>\n",
       "      <td>-0.493750</td>\n",
       "      <td>-2941</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.949957</td>\n",
       "      <td>1.029123</td>\n",
       "      <td>1.015559</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>8.812901</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>GCGCAACCGGTGGGGCTGCG</td>\n",
       "      <td>2</td>\n",
       "      <td>20</td>\n",
       "      <td>8827</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.687500</td>\n",
       "      <td>-0.484375</td>\n",
       "      <td>-2956</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.980212</td>\n",
       "      <td>1.002733</td>\n",
       "      <td>0.990129</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>8.846318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TGCGCAACCGGTGGGGCTGC</td>\n",
       "      <td>3</td>\n",
       "      <td>20</td>\n",
       "      <td>8826</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.500000</td>\n",
       "      <td>-0.475000</td>\n",
       "      <td>-2907</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.968729</td>\n",
       "      <td>0.989912</td>\n",
       "      <td>0.972045</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>8.633534</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CTGCGCAACCGGTGGGGCTG</td>\n",
       "      <td>4</td>\n",
       "      <td>20</td>\n",
       "      <td>8825</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-9.312500</td>\n",
       "      <td>-0.465625</td>\n",
       "      <td>-2867</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.958289</td>\n",
       "      <td>0.988861</td>\n",
       "      <td>0.957519</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>8.560287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>305</th>\n",
       "      <td>TTCCCCAATCAAGATTTTTT</td>\n",
       "      <td>8805</td>\n",
       "      <td>20</td>\n",
       "      <td>24</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.333333</td>\n",
       "      <td>-0.016667</td>\n",
       "      <td>-2044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.991729</td>\n",
       "      <td>0.957326</td>\n",
       "      <td>0.986575</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>TTGTCTTTTTCAGGTAATAGCCTGCAGCTGGTGTTTTGAGAAGCCC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>306</th>\n",
       "      <td>TTTCCCCAATCAAGATTTTT</td>\n",
       "      <td>8806</td>\n",
       "      <td>20</td>\n",
       "      <td>23</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.300000</td>\n",
       "      <td>-0.015000</td>\n",
       "      <td>-2044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.978869</td>\n",
       "      <td>0.959173</td>\n",
       "      <td>1.013243</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>TGTCTTTTTCAGGTAATAGCCTGCAGCTGGTGTTTTGAGAAGCCCT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>307</th>\n",
       "      <td>TTTTCCCCAATCAAGATTTT</td>\n",
       "      <td>8807</td>\n",
       "      <td>20</td>\n",
       "      <td>22</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.266667</td>\n",
       "      <td>-0.013333</td>\n",
       "      <td>-2044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.068860</td>\n",
       "      <td>1.000271</td>\n",
       "      <td>0.982115</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>GTCTTTTTCAGGTAATAGCCTGCAGCTGGTGTTTTGAGAAGCCCTA...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>308</th>\n",
       "      <td>TTTTTCCCCAATCAAGATTT</td>\n",
       "      <td>8808</td>\n",
       "      <td>20</td>\n",
       "      <td>21</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.233333</td>\n",
       "      <td>-0.011667</td>\n",
       "      <td>-2044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.008057</td>\n",
       "      <td>0.983111</td>\n",
       "      <td>0.992385</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>TCTTTTTCAGGTAATAGCCTGCAGCTGGTGTTTTGAGAAGCCCTAC...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>309</th>\n",
       "      <td>TTTTTTCCCCAATCAAGATT</td>\n",
       "      <td>8809</td>\n",
       "      <td>20</td>\n",
       "      <td>20</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>human</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.200000</td>\n",
       "      <td>-0.010000</td>\n",
       "      <td>-2044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.985967</td>\n",
       "      <td>0.958838</td>\n",
       "      <td>0.993042</td>\n",
       "      <td>GACGCAGCCCCACCGGTTGCGCAGTCCCTCCCCGCCCCCGCTCTCC...</td>\n",
       "      <td>CTTTTTCAGGTAATAGCCTGCAGCTGGTGTTTTGAGAAGCCCTACT...</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8810 rows × 38 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 47
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:47:28.850704Z",
     "start_time": "2025-09-17T21:47:19.934003Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from asodesigner.features.RNaseH_features import rnaseh1_dict, compute_rnaseh1_score\n",
    "\n",
    "best_window_start_krel = {\n",
    "    'R4a_krel': {10: 0, 11: 0, 12: 0, 13: 0, 14: 0, 15: 0, 16: 0, 17: 3,  18: 2, 19: 4, 20: 3, 21: 0, 22: 0, 25:0},\n",
    "    'R4b_krel': {10: 0, 11: 0, 12: 0, 13: 0, 14: 0, 15: 0, 16: 0, 17: 1,  18: 3, 19: 1, 20: 3, 21: 0, 22: 0, 25:0},\n",
    "    'R7_krel':  {10: 0, 11: 0, 12: 0, 13: 0, 14: 0, 15: 0, 16: 3, 17: 2,  18: 4, 19: 6, 20: 4, 21: 0, 22: 0, 25:0},\n",
    "}\n",
    "\n",
    "for exp in ['R4a_krel', 'R4b_krel', 'R7_krel']:\n",
    "    weights = rnaseh1_dict(exp)\n",
    "\n",
    "    def score_row(row):\n",
    "        length = len(row['Sequence'])\n",
    "        pos = best_window_start_krel.get(exp, {}).get(length, 0)\n",
    "        return compute_rnaseh1_score(row['Sequence'], weights, window_start=pos)\n",
    "\n",
    "    col_name = f\"RNaseH1_Krel_score_{exp}\"\n",
    "    df_all[col_name] = df_all.apply(score_row, axis=1)\n",
    "\n",
    "RNaseH1_Krel_features_best = [f\"RNaseH1_Krel_score_{exp}\" for exp in ['R4a_krel', 'R4b_krel', 'R7_krel']]\n"
   ],
   "id": "4995154a94bcece1",
   "outputs": [],
   "execution_count": 60
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:47:33.426078Z",
     "start_time": "2025-09-17T21:47:33.423603Z"
    }
   },
   "cell_type": "code",
   "source": "df_all.loc[:, 'CAI_score_global_CDS'] = 0.8526221963673779 # TODO - replace with the calculation\n",
   "id": "aaff1f66a3bb1f47",
   "outputs": [],
   "execution_count": 61
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:47:33.618867Z",
     "start_time": "2025-09-17T21:47:33.616031Z"
    }
   },
   "cell_type": "code",
   "source": "df_all.columns",
   "id": "572e90ff6de288a4",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Sequence', 'sense_start', 'sense_length', 'sense_start_from_end',\n",
       "       'Canonical Gene Name', 'Cell line organism', 'Inhibition(%)',\n",
       "       'sense_exon', 'sense_intron', 'sense_utr', 'sense_type',\n",
       "       'Treatment_Period(hours)', 'ASO_volume(nM)', 'log_volume',\n",
       "       'normalized_start', 'normalized_sense_start_from_end', 'self_energy',\n",
       "       'internal_fold', 'gc_content', 'gc_content_3_prime_5',\n",
       "       'homooligo_count', 'hairpin_score', 'gc_skew', 'at_skew',\n",
       "       'nucleotide_diversity', 'stop_codon_count', 'at_rich_region_score',\n",
       "       'poly_pyrimidine_stretch', 'on_target_fold_openness40_15',\n",
       "       'on_target_fold_openness_normalized40_15', 'exp_ps_hybr',\n",
       "       'Modification_min_distance_to_3prime',\n",
       "       'RNaseH1_Krel_score_R4a_krel_Krel', 'RNaseH1_Krel_score_R4b_krel_Krel',\n",
       "       'RNaseH1_Krel_score_R7_krel_Krel', 'pre_mrna_sequence',\n",
       "       'sense_with_flank_120nt', 'sense_avg_accessibility',\n",
       "       'CAI_score_global_CDS', 'RNaseH1_Krel_score_R4a_krel',\n",
       "       'RNaseH1_Krel_score_R4b_krel', 'RNaseH1_Krel_score_R7_krel'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 62
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T21:57:50.115370Z",
     "start_time": "2025-09-17T21:57:50.096666Z"
    }
   },
   "cell_type": "code",
   "source": "malat_scores = model.predict(df_all[selected_features])\n",
   "id": "bc521ad6b8751d00",
   "outputs": [],
   "execution_count": 68
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T22:08:35.831581Z",
     "start_time": "2025-09-17T22:08:35.111859Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from asodesigner.util import get_antisense\n",
    "\n",
    "# Assuming you already have get_antisense(seq: str) -> str defined\n",
    "# get_antisense, for some reason numba doesn't work well\n",
    "tbl = str.maketrans(\"ACGTUacgtuNn\", \"TGCAAtgcaaNn\")\n",
    "\n",
    "df_all[\"score\"] = malat_scores\n",
    "\n",
    "df_all[\"sense\"] = df_all[SEQUENCE].astype(str).str.translate(tbl).str[::-1]\n",
    "(\n",
    "    df_all.assign(score=malat_scores)\n",
    "    .sort_values(\"score\", ascending=False)   # sort by score\n",
    "    .to_csv(\"malat_scores_model2.csv\", index=False)\n",
    ")\n"
   ],
   "id": "459df770ce0791ac",
   "outputs": [],
   "execution_count": 86
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T22:08:36.805397Z",
     "start_time": "2025-09-17T22:08:36.800379Z"
    }
   },
   "cell_type": "code",
   "source": "df_sorted = df_all.sort_values('score', ascending=False)\n",
   "id": "92d5a85bc6fc72ff",
   "outputs": [],
   "execution_count": 87
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T22:10:25.003177Z",
     "start_time": "2025-09-17T22:08:42.898Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# pip install requests\n",
    "import math, time, threading, urllib.parse, requests\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed\n",
    "\n",
    "UA = {\"User-Agent\": \"python-requests gggenome/greedy\"}\n",
    "\n",
    "def _ggg_hits_leq_json(seq, k, db=\"hg38\", timeout=60, retries=2):\n",
    "    \"\"\"Count hits with <=k mismatches via GGGenome JSON; fallback to CSV if needed.\"\"\"\n",
    "    s = str(seq).upper().replace(\"U\", \"T\")\n",
    "    q = urllib.parse.quote(s)\n",
    "    url_json = f\"https://gggenome.dbcls.jp/{db}/{k}/nogap/{q}.json\"\n",
    "    url_csv  = f\"https://gggenome.dbcls.jp/{db}/{k}/nogap/{q}.csv?download\"\n",
    "\n",
    "    for attempt in range(retries + 1):\n",
    "        try:\n",
    "            r = requests.get(url_json, headers=UA, timeout=timeout)\n",
    "            r.raise_for_status()\n",
    "            try:\n",
    "                data = r.json()\n",
    "            except ValueError:\n",
    "                raise RuntimeError(\"JSON parse failed\")\n",
    "            if isinstance(data, list):\n",
    "                return len(data)\n",
    "            if isinstance(data, dict):\n",
    "                if \"results\" in data and isinstance(data[\"results\"], list): return len(data[\"results\"])\n",
    "                if \"hits\" in data and isinstance(data[\"hits\"], list):       return len(data[\"hits\"])\n",
    "                return sum(len(v) for v in data.values() if isinstance(v, list))\n",
    "            return 0\n",
    "        except Exception:\n",
    "            # greedy CSV fallback\n",
    "            try:\n",
    "                r2 = requests.get(url_csv, headers=UA, timeout=timeout)\n",
    "                r2.raise_for_status()\n",
    "                return sum(1 for ln in r2.text.splitlines() if ln and not ln.startswith(\"#\"))\n",
    "            except Exception:\n",
    "                if attempt < retries:\n",
    "                    continue\n",
    "                return 0\n",
    "    return 0\n",
    "\n",
    "def _d123_for_sequence(seq, db=\"hg38\"):\n",
    "    s = str(seq).upper().replace(\"U\", \"T\")\n",
    "    if not s:\n",
    "        return (s, 0, 0, 0)\n",
    "    L = len(s)\n",
    "    k_allowed = max(0, math.floor(0.25 * L))  # GGGenome cap\n",
    "    k0 = _ggg_hits_leq_json(s, 0, db=db)\n",
    "    k1 = _ggg_hits_leq_json(s, 1, db=db) if k_allowed >= 1 else 0\n",
    "    k2 = _ggg_hits_leq_json(s, 2, db=db) if k_allowed >= 2 else 0\n",
    "    k3 = _ggg_hits_leq_json(s, 3, db=db) if k_allowed >= 3 else 0\n",
    "    d1 = max(0, k1 - k0)\n",
    "    d2 = max(0, k2 - k1)\n",
    "    d3 = max(0, k3 - k2)\n",
    "    return (s, d1, d2, d3)\n",
    "\n",
    "def add_gggenome_d123(main_df, seq_col=\"SEQUENCE\", db=\"hg38\", *, max_workers=32, print_every=10):\n",
    "    seqs = (main_df[seq_col].astype(str).str.upper().str.replace(\"U\", \"T\", regex=False))\n",
    "    uniq = seqs.dropna().unique().tolist()\n",
    "    N = len(uniq)\n",
    "    print(f\"[GGG] Unique sequences: {N} | db={db} | workers={max_workers}\")\n",
    "\n",
    "    cache = {}\n",
    "    lock = threading.Lock()\n",
    "    t0 = time.perf_counter()\n",
    "    errs = 0\n",
    "\n",
    "    with ThreadPoolExecutor(max_workers=max_workers) as ex:\n",
    "        futs = {ex.submit(_d123_for_sequence, s, db): s for s in uniq}\n",
    "        done = 0\n",
    "        for fut in as_completed(futs):\n",
    "            s = futs[fut]\n",
    "            try:\n",
    "                s_key, d1, d2, d3 = fut.result()\n",
    "            except Exception:\n",
    "                d1 = d2 = d3 = 0\n",
    "                with lock:\n",
    "                    errs += 1\n",
    "            with lock:\n",
    "                cache[s] = (d1, d2, d3)\n",
    "                done += 1\n",
    "                if (done == 1) or (done % print_every == 0) or (done == N):\n",
    "                    elapsed = time.perf_counter() - t0\n",
    "                    rps = done / elapsed if elapsed > 0 else 0.0\n",
    "                    print(f\"[GGG] {done}/{N} cached | ~{rps:.1f} seq/s | errors={errs}\")\n",
    "\n",
    "    main_df[\"ggg_d1\"] = seqs.map(lambda s: cache.get(s, (0, 0, 0))[0])\n",
    "    main_df[\"ggg_d2\"] = seqs.map(lambda s: cache.get(s, (0, 0, 0))[1])\n",
    "    main_df[\"ggg_d3\"] = seqs.map(lambda s: cache.get(s, (0, 0, 0))[2])\n",
    "\n",
    "    print(f\"[GGG] Finished in {time.perf_counter() - t0:.1f}s. Added columns: ggg_d1, ggg_d2, ggg_d3\")\n",
    "    return main_df\n",
    "\n",
    "\n",
    "# --- usage ---\n",
    "# main_df = main_df[main_df[SENSE_START] != -1]\n",
    "result = add_gggenome_d123(df_sorted[:100], seq_col='sense', db=\"hg38\", max_workers=100, print_every=2)"
   ],
   "id": "dbe6eb9130b4f08",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[GGG] Unique sequences: 100 | db=hg38 | workers=100\n",
      "[GGG] 1/100 cached | ~0.0 seq/s | errors=0\n",
      "[GGG] 2/100 cached | ~0.0 seq/s | errors=0\n",
      "[GGG] 4/100 cached | ~0.1 seq/s | errors=0\n",
      "[GGG] 6/100 cached | ~0.1 seq/s | errors=0\n",
      "[GGG] 8/100 cached | ~0.1 seq/s | errors=0\n",
      "[GGG] 10/100 cached | ~0.2 seq/s | errors=0\n",
      "[GGG] 12/100 cached | ~0.2 seq/s | errors=0\n",
      "[GGG] 14/100 cached | ~0.2 seq/s | errors=0\n",
      "[GGG] 16/100 cached | ~0.2 seq/s | errors=0\n",
      "[GGG] 18/100 cached | ~0.3 seq/s | errors=0\n",
      "[GGG] 20/100 cached | ~0.3 seq/s | errors=0\n",
      "[GGG] 22/100 cached | ~0.3 seq/s | errors=0\n",
      "[GGG] 24/100 cached | ~0.4 seq/s | errors=0\n",
      "[GGG] 26/100 cached | ~0.4 seq/s | errors=0\n",
      "[GGG] 28/100 cached | ~0.4 seq/s | errors=0\n",
      "[GGG] 30/100 cached | ~0.4 seq/s | errors=0\n",
      "[GGG] 32/100 cached | ~0.5 seq/s | errors=0\n",
      "[GGG] 34/100 cached | ~0.5 seq/s | errors=0\n",
      "[GGG] 36/100 cached | ~0.5 seq/s | errors=0\n",
      "[GGG] 38/100 cached | ~0.5 seq/s | errors=0\n",
      "[GGG] 40/100 cached | ~0.6 seq/s | errors=0\n",
      "[GGG] 42/100 cached | ~0.6 seq/s | errors=0\n",
      "[GGG] 44/100 cached | ~0.6 seq/s | errors=0\n",
      "[GGG] 46/100 cached | ~0.7 seq/s | errors=0\n",
      "[GGG] 48/100 cached | ~0.7 seq/s | errors=0\n",
      "[GGG] 50/100 cached | ~0.7 seq/s | errors=0\n",
      "[GGG] 52/100 cached | ~0.7 seq/s | errors=0\n",
      "[GGG] 54/100 cached | ~0.7 seq/s | errors=0\n",
      "[GGG] 56/100 cached | ~0.7 seq/s | errors=0\n",
      "[GGG] 58/100 cached | ~0.8 seq/s | errors=0\n",
      "[GGG] 60/100 cached | ~0.8 seq/s | errors=0\n",
      "[GGG] 62/100 cached | ~0.8 seq/s | errors=0\n",
      "[GGG] 64/100 cached | ~0.8 seq/s | errors=0\n",
      "[GGG] 66/100 cached | ~0.8 seq/s | errors=0\n",
      "[GGG] 68/100 cached | ~0.9 seq/s | errors=0\n",
      "[GGG] 70/100 cached | ~0.9 seq/s | errors=0\n",
      "[GGG] 72/100 cached | ~0.9 seq/s | errors=0\n",
      "[GGG] 74/100 cached | ~0.9 seq/s | errors=0\n",
      "[GGG] 76/100 cached | ~1.0 seq/s | errors=0\n",
      "[GGG] 78/100 cached | ~1.0 seq/s | errors=0\n",
      "[GGG] 80/100 cached | ~1.0 seq/s | errors=0\n",
      "[GGG] 82/100 cached | ~1.0 seq/s | errors=0\n",
      "[GGG] 84/100 cached | ~0.9 seq/s | errors=0\n",
      "[GGG] 86/100 cached | ~0.9 seq/s | errors=0\n",
      "[GGG] 88/100 cached | ~1.0 seq/s | errors=0\n",
      "[GGG] 90/100 cached | ~1.0 seq/s | errors=0\n",
      "[GGG] 92/100 cached | ~1.0 seq/s | errors=0\n",
      "[GGG] 94/100 cached | ~1.0 seq/s | errors=0\n",
      "[GGG] 96/100 cached | ~1.0 seq/s | errors=0\n",
      "[GGG] 98/100 cached | ~1.0 seq/s | errors=0\n",
      "[GGG] 100/100 cached | ~1.0 seq/s | errors=0\n",
      "[GGG] Finished in 95.7s. Added columns: ggg_d1, ggg_d2, ggg_d3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_823373/3595405422.py:86: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  main_df[\"ggg_d1\"] = seqs.map(lambda s: cache.get(s, (0, 0, 0))[0])\n",
      "/tmp/ipykernel_823373/3595405422.py:87: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  main_df[\"ggg_d2\"] = seqs.map(lambda s: cache.get(s, (0, 0, 0))[1])\n",
      "/tmp/ipykernel_823373/3595405422.py:88: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  main_df[\"ggg_d3\"] = seqs.map(lambda s: cache.get(s, (0, 0, 0))[2])\n"
     ]
    }
   ],
   "execution_count": 88
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T22:21:55.591734Z",
     "start_time": "2025-09-17T22:21:55.586341Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Show up to 200 characters per column\n",
    "pd.set_option(\"display.max_colwidth\", 200)\n",
    "\n",
    "# Show more columns across the screen\n",
    "pd.set_option(\"display.max_columns\", None)\n",
    "\n",
    "# Optionally, widen the console display\n",
    "pd.set_option(\"display.width\", 200)\n",
    "\n",
    "print(result[:30][[SEQUENCE, 'sense_start', 'ggg_d1', 'ggg_d2', 'at_skew', 'gc_content', 'on_target_fold_openness_normalized40_15', 'sense_avg_accessibility', 'RNaseH1_Krel_score_R7_krel']])\n"
   ],
   "id": "73da04a7639168ea",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                 Sequence  sense_start  ggg_d1  ggg_d2   at_skew  gc_content  on_target_fold_openness_normalized40_15  sense_avg_accessibility  RNaseH1_Krel_score_R7_krel\n",
      "5    TATCTTCTCTATTCTTTTCT         2005      43    1000 -0.733333        0.25                                -0.004375                 0.903107                    0.974312\n",
      "7    CCTATCTTCTCTATTCTTTT         2007       3       0 -0.714286        0.30                                -0.005792                 1.027160                    0.977993\n",
      "12   ATTTCCCTATCTTCTCTATT         2012       5       0 -0.571429        0.30                                -0.009583                 1.266860                    0.992284\n",
      "76   CTGCTTCCTACTTTTCAGGT         2076       2       0 -0.636364        0.45                                -0.084000                 1.498184                    0.973887\n",
      "337  TTTCGGCTTCTTTTATTCCA         5337       1      27 -0.692308        0.35                                -0.084667                 1.489497                    0.975788\n",
      "499  CTCTATTCTTTTCTTCGCCT         1999       0      36 -0.833333        0.40                                -0.017625                 0.603610                    0.954213\n",
      "1    TTCTCTATTCTTTTCTTCGC         2001       4     180 -0.846154        0.35                                -0.013375                 0.660595                    0.929154\n",
      "0    TCTCTATTCTTTTCTTCGCC         2000       1       0 -0.833333        0.40                                -0.015625                 0.651758                    0.951643\n",
      "2    CTTCTCTATTCTTTTCTTCG         2002      14       0 -0.846154        0.35                                -0.011125                 0.669695                    0.951424\n",
      "77   TCTGCTTCCTACTTTTCAGG         2077       1      39 -0.636364        0.45                                -0.083417                 1.349813                    0.971981\n",
      "6    CTATCTTCTCTATTCTTTTC         2006       5       0 -0.714286        0.30                                -0.005083                 0.963104                    0.958485\n",
      "336  TTCGGCTTCTTTTATTCCAG         5336       1      14 -0.666667        0.40                                -0.085458                 1.481122                    0.962050\n",
      "335  TCGGCTTCTTTTATTCCAGG         5335       0      10 -0.636364        0.45                                -0.086250                 1.471181                    0.958870\n",
      "78   TTCTGCTTCCTACTTTTCAG         2078       2      61 -0.666667        0.40                                -0.082833                 1.209630                    0.981337\n",
      "4    ATCTTCTCTATTCTTTTCTT         2004     333       0 -0.733333        0.25                                -0.006625                 0.841697                    0.956763\n",
      "3    TCTTCTCTATTCTTTTCTTC         2003      70       0 -0.857143        0.30                                -0.008875                 0.683592                    0.956535\n",
      "155  TTGTTTCTATCTTCCAATTT         2155       3       0 -0.600000        0.25                                -0.002917                 1.721309                    0.974491\n",
      "485  TCGCCTTCCCGTACTTCTGT         1985       0       0 -0.777778        0.55                                -0.066875                 1.455023                    0.983253\n",
      "153  GTTTCTATCTTCCAATTTTC         2153       1      65 -0.571429        0.30                                -0.007917                 1.482960                    0.945041\n",
      "486  TTCGCCTTCCCGTACTTCTG         1986       0       0 -0.777778        0.55                                -0.061750                 1.292820                    0.980251\n",
      "161  GTTATATTAGGTTCTCGTGT         8661       0       0 -0.538462        0.35                                -0.152625                 1.354807                    0.984114\n",
      "169  TTTCATTTTCTATCTTGTTT         2169      11       0 -0.750000        0.20                                -0.009417                 2.217595                    0.980994\n",
      "157  TCTTGTTTCTATCTTCCAAT         2157       2       0 -0.571429        0.30                                -0.003917                 1.741027                    0.988863\n",
      "141  CAATTTTCTTCTAAGTTTGT         2141       2       0 -0.466667        0.25                                -0.044167                 1.323397                    0.974970\n",
      "157  TATTAGGTTCTCGTGTAAAT         8657       0       4 -0.285714        0.30                                -0.148958                 1.311029                    1.000077\n",
      "5    CTTGAAGACTGTTGCTTGTT         7005       1      14 -0.500000        0.40                                -0.047500                 3.654411                    0.981246\n",
      "164  GCAGTTATATTAGGTTCTCG         8664       0       0 -0.333333        0.40                                -0.156125                 1.559141                    0.975641\n",
      "394  CATTTTCGTCTGCGTTTAGT         1894       0       5 -0.666667        0.40                                -0.057125                 2.461010                    0.979906\n",
      "11   TTTCCCTATCTTCTCTATTC         2011       4      85 -0.692308        0.35                                -0.008750                 1.470504                    0.940867\n",
      "10   TTCCCTATCTTCTCTATTCT         2010       4       0 -0.692308        0.35                                -0.007917                 1.466386                    0.939322\n"
     ]
    }
   ],
   "execution_count": 98
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T22:22:32.982136Z",
     "start_time": "2025-09-17T22:22:32.976787Z"
    }
   },
   "cell_type": "code",
   "source": "df_sorted.tail(30)[['on_target_fold_openness_normalized40_15']]",
   "id": "6b13e4688def2b34",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "     on_target_fold_openness_normalized40_15\n",
       "383                                -0.569417\n",
       "438                                -0.427958\n",
       "427                                -0.480875\n",
       "444                                -0.414875\n",
       "443                                -0.416208\n",
       "425                                -0.459792\n",
       "456                                -0.450292\n",
       "455                                -0.452917\n",
       "459                                -0.442417\n",
       "460                                -0.439792\n",
       "98                                 -0.515333\n",
       "99                                 -0.517250\n",
       "462                                -0.498375\n",
       "134                                -0.598625\n",
       "106                                -0.528042\n",
       "87                                 -0.418583\n",
       "344                                -0.480625\n",
       "463                                -0.461958\n",
       "441                                -0.418875\n",
       "464                                -0.473375\n",
       "422                                -0.411042\n",
       "436                                -0.435708\n",
       "142                                -0.442167\n",
       "432                                -0.439083\n",
       "437                                -0.431833\n",
       "428                                -0.491417\n",
       "439                                -0.424083\n",
       "105                                -0.538125\n",
       "103                                -0.530542\n",
       "104                                -0.534333"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>on_target_fold_openness_normalized40_15</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>383</th>\n",
       "      <td>-0.569417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>438</th>\n",
       "      <td>-0.427958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>427</th>\n",
       "      <td>-0.480875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td>-0.414875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>-0.416208</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>425</th>\n",
       "      <td>-0.459792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td>-0.450292</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td>-0.452917</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td>-0.442417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td>-0.439792</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>-0.515333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>99</th>\n",
       "      <td>-0.517250</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td>-0.498375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134</th>\n",
       "      <td>-0.598625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>106</th>\n",
       "      <td>-0.528042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>-0.418583</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>344</th>\n",
       "      <td>-0.480625</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td>-0.461958</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>441</th>\n",
       "      <td>-0.418875</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>-0.473375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>422</th>\n",
       "      <td>-0.411042</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>436</th>\n",
       "      <td>-0.435708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>142</th>\n",
       "      <td>-0.442167</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>432</th>\n",
       "      <td>-0.439083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>437</th>\n",
       "      <td>-0.431833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>428</th>\n",
       "      <td>-0.491417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>439</th>\n",
       "      <td>-0.424083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>105</th>\n",
       "      <td>-0.538125</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103</th>\n",
       "      <td>-0.530542</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>-0.534333</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 99
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "# Now let's generate LNA GFP sequences",
   "id": "826aff404f2adaab"
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T22:44:18.026502Z",
     "start_time": "2025-09-17T22:44:18.021438Z"
    }
   },
   "cell_type": "code",
   "source": [
    "GFP_IN_YEAST = (\n",
    "    'ATGGTtAGtAAaGGaGAaGAGTTgTTCACaGGaGTGGTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAGGGCGAGGGCGATGCCACCTACGGCAAGCTGACCCTGAAGTTCATCTGCACCACCGGCAAGCTGCCCGTGCCCTGGCCCACCCTCGTGACCACCCTGACCTACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGCACGACTTCTTCAAGTCCGCCATGCCCGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGCAACTACAAGACCCGCGCCGAGGTGAAGTTCGAGGGCGACACCCTGGTGAACCGCATCGAGCTGAAGGGCATCGACTTCAAGGAGGACGGCAACATCCTGGGGCACAAGCTGGAGTACAACTACAACAGCCACAACGTCTATATCATGGCCGACAAGCAGAAGAACGGCATCAAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCGACGGCCCCGTGCTGCTGCCCGACAACCACTACCTGAGCACCCAGTCCGCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACAAGGGTGCTGGGGCAggtacCCCTAAAGATCCAGCCAAACCTCCGGCCAcGGCACAAGTTGTGGGATGGCCACCGGTGAGATCATACCGGAAGAACGTGATGGTTTCCTGCCAAAAATCAAGCGGTGGCCCGGAGGCGGCGGCGTTCGTGAAGTAA'\n",
    "    .upper())\n",
    "\n",
    "GFP_IN_HUMAN = (\n",
    "\"atggtgagcaagggcgaggagctgttcaccggggtggtgcccatcctggtcgagctggacggcgacgtaaacggccacaagttcagcgtgtccggcgagggcgagggcgatgccacctacggcaagctgaccctgaagttcatctgcaccaccggcaagctgcccgtgccctggcccaccctcgtgaccaccctgacctacggcgtgcagtgcttcagccgctaccccgaccacatgaagcagcacgacttcttcaagtccgccatgcccgaaggctacgtccaggagcgcaccatcttcttcaaggacgacggcaactacaagacccgcgccgaggtgaagttcgagggcgacaccctggtgaaccgcatcgagctgaagggcatcgacttcaaggaggacggcaacatcctggggcacaagctggagtacaactacaacagccacaacgtctatatcatggccgacaagcagaagaacggcatcaaggtgaacttcaagatccgccacaacatcgaggacggcagcgtgcagctcgccgaccactaccagcagaacacccccatcggcgacggccccgtgctgctgcccgacaaccactacctgagcacccagtccgccctgagcaaagaccccaacgagaagcgcgatcacatggtcctgctggagttcgtgaccgccgccgggatcactctcggcatggacgagctgtacaagaagcttagccatggcttcccgccggaggtggaggagcaggatgatggcacgctgcccatgtcttgtgcccaggagagcgggatggaccgtcaccctgcagcctgtgcttctgctaggatcaatgtgaagcgacctgccgccacaaagaaggctggacaggctaagaagaagaaatgaggatcccgcgcgcgcatatgttaattaaccaactgcatggggatccacgcgttaagtcgacaatcaacctctggattacaaaatttgtgaaagattgactggtattcttaactatgttgctccttttacgctatgtggatacgctgctttaatgcctttgtatcatgctattgcttcccgtatggctttcattttctcctccttgtataaatcctggttgctgtctctttatgaggagttgtggcccgttgtcaggcaacgtggcgtggtgtgcactgtgtttgctgacgcaacccccactggttggggcattgccaccacctgtcagctcctttccgggactttcgctttccccctccctattgccacggcggaactcatcgccgcctgccttgcccgctgctggacaggggctcggctgttgggcactgacaattccgtggtgttgtcggggaaatcatcgtcctttccttggctgctcgcctgtgttgccacctggattctgcgcgggacgtccttctgctacgtcccttcggccctcaatccagcggaccttccttcccgcggcctgctgccggctctgcggcctcttccgcgtcttcgccttcgccctcagacgagtcggatctccctttgggccgcctccccgcgtcgactttaagaccaatgacttacaaggcagctgtagatcttagccactttttaaaagaaaaggggggactggaagggctaattcactcccaacgaagacaagatctgctttttgcttgtactgggtctctctggttagaccagatctgagcctgggagctctctggctaactagggaacccactgcttaagcctcaataaagcttgccttgagtgcttcaagtagtgtgtgcccgtctgttgtgtgactctggtaactagagatccctcagacccttttagtcagtgtggaaaatctctagcagtacgtatagtagttcatgtcatcttattattcagtatttataacttgcaaagaaatgaatatcagagagtgagagg\".upper()\n",
    ")\n",
    "# GFP + Degron x <= 842\n",
    "# NLS 843 <= x<= 893\n",
    "# 3UTR 1615 <= x <= 1848\n"
   ],
   "id": "ca6b83cac69a89e0",
   "outputs": [],
   "execution_count": 100
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T22:47:30.241963Z",
     "start_time": "2025-09-17T22:47:30.227786Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from asodesigner.util import get_antisense\n",
    "import pandas as pd\n",
    "\n",
    "GFP_YEAST_END = len(GFP_IN_YEAST) - 3\n",
    "GFP_HUMAN_END = 891  # TGA stop codon in NLS\n",
    "LNA_SIZE = 16\n",
    "\n",
    "def get_init_df(target_mrna, end):\n",
    "    candidates = []\n",
    "    sense_starts = []\n",
    "    sense_lengths = []\n",
    "    sense_starts_from_end = []\n",
    "\n",
    "    for i in range(0, len(target_mrna) - (LNA_SIZE - 1)):\n",
    "        target = target_mrna[i: i + LNA_SIZE]\n",
    "        candidates.append(get_antisense(str(target)))\n",
    "        sense_starts.append(i)\n",
    "        sense_lengths.append(LNA_SIZE)\n",
    "        sense_starts_from_end.append(end - i)\n",
    "    df = pd.DataFrame(\n",
    "        {SEQUENCE: candidates, SENSE_START: sense_starts,\n",
    "         SENSE_LENGTH: sense_lengths, \"sense_start_from_end\": sense_starts_from_end})\n",
    "    return df\n",
    "\n",
    "\n",
    "df_yeast = get_init_df(GFP_IN_YEAST, GFP_YEAST_END)\n",
    "df_human = get_init_df(GFP_IN_HUMAN, GFP_HUMAN_END)"
   ],
   "id": "64fa91eddd19f261",
   "outputs": [],
   "execution_count": 101
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T22:50:04.087234Z",
     "start_time": "2025-09-17T22:50:04.080796Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from asodesigner.util import get_antisense\n",
    "import pandas as pd\n",
    "\n",
    "df_yeast[CANONICAL_GENE] = 'YEAST_GFP'\n",
    "df_human[CANONICAL_GENE] = 'HUMAN_GFP'\n",
    "dataframes = [df_yeast, df_human]\n",
    "\n",
    "df_yeast['sense_exon'] = 1\n",
    "df_human['sense_exon'] = 1\n",
    "df_yeast['sense_intron'] = 0\n",
    "df_human['sense_intron'] = 0\n",
    "df_yeast['sense_utr'] = 0\n",
    "df_human['sense_utr'] = [1 if sense_start > 842 else 0 for sense_start in df_human[SENSE_START]]\n"
   ],
   "id": "f04111a58b965b0e",
   "outputs": [],
   "execution_count": 102
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T22:51:55.966698Z",
     "start_time": "2025-09-17T22:51:55.911635Z"
    }
   },
   "cell_type": "code",
   "source": [
    "%reload_ext autoreload\n",
    "%autoreload 2\n",
    "\n",
    "from asodesigner.process_utils import LocusInfo\n",
    "from scripts.data_genertion.data_handling import get_populate_fold\n",
    "\n",
    "genes_u = ['YEAST_GFP', 'HUMAN_GFP', 'MALAT1']\n",
    "gene_to_data = {'YEAST_GFP': LocusInfo(), 'HUMAN_GFP': LocusInfo(), 'MALAT1': LocusInfo()}\n",
    "gene_to_data['YEAST_GFP'].full_mrna = GFP_IN_YEAST\n",
    "gene_to_data['HUMAN_GFP'].full_mrna = GFP_IN_HUMAN"
   ],
   "id": "f96c8c63c1e68429",
   "outputs": [],
   "execution_count": 105
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T22:54:26.774808Z",
     "start_time": "2025-09-17T22:54:26.474341Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from scripts.data_genertion.data_handling import populate_features\n",
    "\n",
    "dfs = dict()\n",
    "dfs['YEAST_GFP'] = df_yeast\n",
    "dfs['HUMAN_GFP'] = df_human\n",
    "SEQUENCES = {}\n",
    "for target in ['YEAST_GFP', 'HUMAN_GFP']:\n",
    "    SEQUENCES[target] = gene_to_data[target].full_mrna\n",
    "\n",
    "for gene, df in dfs.items():\n",
    "    df[TREATMENT_PERIOD] = 24  # keep constant for all\n",
    "    df[VOLUME] = 1000  # keep constant for all\n",
    "    df['log_volume'] = np.log(df[VOLUME])\n",
    "    df['normalized_start'] = df[SENSE_START] / len(SEQUENCES[gene])\n",
    "    df['normalized_sense_start_from_end'] = df['sense_start_from_end'] / len(SEQUENCES[gene])\n",
    "    easy_to_populate = ['at_skew', 'gc_content', 'gc_content_3_prime_5', 'gc_skew', 'hairpin_score',\n",
    "                        'homooligo_count', 'internal_fold', 'nucleotide_diversity', 'self_energy', 'stop_codon_count',\n",
    "                        'at_rich_region_score', 'poly_pyrimidine_stretch']\n",
    "    populate_features(df, easy_to_populate)"
   ],
   "id": "35d54c2f22f5a3b5",
   "outputs": [],
   "execution_count": 107
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T23:23:20.783947Z",
     "start_time": "2025-09-17T23:23:20.763726Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def get_populate_fold(df, genes_u, gene_to_data, fold_variants=[(40, 15)]):\n",
    "    from asodesigner.fold import calculate_energies, get_weighted_energy\n",
    "    from asodesigner.util import get_antisense\n",
    "\n",
    "    all_data_human_gene_premrna_no_nan = df.copy()\n",
    "\n",
    "    # Comment out the long cases for quick running\n",
    "    for (window_size, step_size) in fold_variants:\n",
    "\n",
    "        on_target_fold = 'on_target_fold_openness' + str(window_size) + '_' + str(step_size)\n",
    "        on_target_fold_normalized = 'on_target_fold_openness_normalized' + str(window_size) + '_' + str(step_size)\n",
    "        all_data_human_gene_premrna_no_nan[on_target_fold] = np.zeros_like(all_data_human_gene_premrna_no_nan[SEQUENCE],\n",
    "                                                                           dtype=np.float64)\n",
    "        all_data_human_gene_premrna_no_nan[on_target_fold_normalized] = np.zeros_like(\n",
    "            all_data_human_gene_premrna_no_nan[SEQUENCE], dtype=np.float64)\n",
    "\n",
    "        for gene in genes_u:\n",
    "\n",
    "            target = gene_to_data[gene].full_mrna\n",
    "            gene_rows = all_data_human_gene_premrna_no_nan[all_data_human_gene_premrna_no_nan[CANONICAL_GENE] == gene]\n",
    "            energies = calculate_energies(str(target), step_size, window_size)\n",
    "\n",
    "            for index, row in gene_rows.iterrows():\n",
    "                antisense = row[SEQUENCE]\n",
    "                sense = get_antisense(antisense)\n",
    "                l = row[SENSE_LENGTH]\n",
    "                sense_start = row[SENSE_START]\n",
    "                mean_fold = get_weighted_energy(sense_start, l, step_size, energies, window_size)\n",
    "                mean_fold_end = get_weighted_energy(sense_start, l, step_size, energies, window_size)\n",
    "                mean_fold_start = get_weighted_energy(sense_start, l, step_size, energies, window_size)\n",
    "                if mean_fold > 100:\n",
    "                    print(energies)\n",
    "                    print(\"Weird: \", mean_fold)\n",
    "                    print(\"Sense_start \", sense_start)\n",
    "                    print(\"Sense_length \", l)\n",
    "                    print(\"Gene: \", gene)\n",
    "                    mean_fold = 0\n",
    "                all_data_human_gene_premrna_no_nan.loc[index, on_target_fold] = mean_fold\n",
    "                all_data_human_gene_premrna_no_nan.loc[index, on_target_fold_normalized] = mean_fold / l\n",
    "    return all_data_human_gene_premrna_no_nan"
   ],
   "id": "3eacc16cb4318ae9",
   "outputs": [],
   "execution_count": 129
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T23:23:43.200920Z",
     "start_time": "2025-09-17T23:23:42.334851Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from scripts.data_genertion.data_handling import get_populate_fold\n",
    "\n",
    "dfs['YEAST_GFP'] = df_yeast\n",
    "dfs['HUMAN_GFP'] = df_human\n",
    "\n",
    "for gene, df in dfs.items():\n",
    "    fold_variants = [(40, 15)]\n",
    "\n",
    "    df = get_populate_fold(df, ['YEAST_GFP', 'HUMAN_GFP'], gene_to_data, fold_variants=fold_variants)\n",
    "    dfs[gene] = df"
   ],
   "id": "8f9219459f3d66eb",
   "outputs": [],
   "execution_count": 132
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T23:24:30.735789Z",
     "start_time": "2025-09-17T23:24:30.713894Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from hybridization.hybridization_features import get_exp_psrna_hybridization\n",
    "\n",
    "for gene, df in dfs.items():\n",
    "    df.loc[:, 'exp_ps_hybr'] = [\n",
    "        get_exp_psrna_hybridization(antisense.replace('T', 'U'), temp=37) for\n",
    "        antisense in df[SEQUENCE]]"
   ],
   "id": "9b741973dc97c2cd",
   "outputs": [],
   "execution_count": 134
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T23:24:32.060073Z",
     "start_time": "2025-09-17T23:24:32.040810Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from features.mod_features import compute_mod_min_distance_to_3prime\n",
    "\n",
    "# generate LNA 16-mers\n",
    "for gene, df in dfs.items():\n",
    "    df.loc[:, 'Modification_min_distance_to_3prime'] = compute_mod_min_distance_to_3prime('LLLddddddddddLLL')"
   ],
   "id": "43e984282a8438c9",
   "outputs": [],
   "execution_count": 135
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T23:31:24.473767Z",
     "start_time": "2025-09-17T23:24:39.087486Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Missing: CAI_score_global_CDS, 'sense_avg_accessibility', RNaseH1_Krel_score_R7_krel, Modification_min_distance_to_3prime\n",
    "\n",
    "from yehuda_code.Folding_Functions import get_sense_with_flanks\n",
    "from yehuda_code.access_calculator import AccessCalculator\n",
    "\n",
    "FLANK_SIZE = 120\n",
    "ACCESS_SIZE = 13\n",
    "SEED_SIZE = 13\n",
    "SEED_SIZES = [SEED_SIZE * m for m in range(1, 4)]\n",
    "ACCESS_WIN_SIZE = 80\n",
    "\n",
    "\n",
    "def compute_sense_accessibility(row, flank_size, access_win_size, seed_sizes, access_size, min_gc=0, max_gc=100,\n",
    "                                gc_ranges=1):\n",
    "    try:\n",
    "        # Skip invalid rows\n",
    "        if row['sense_start'] == -1 or pd.isna(row['sense_with_flank_120nt']) or row['sense_with_flank_120nt'] == \"\":\n",
    "            return None\n",
    "\n",
    "        seq = row[f'sense_with_flank_{flank_size}nt']\n",
    "        sense_start = row['sense_start']\n",
    "        sense_length = row['sense_length']\n",
    "\n",
    "        # Calculate accessibility\n",
    "        df_access = AccessCalculator.calc(\n",
    "            seq, access_size,\n",
    "            min_gc, max_gc, gc_ranges,\n",
    "            access_win_size, seed_sizes\n",
    "        )\n",
    "\n",
    "        flank_start = max(0, sense_start - flank_size)\n",
    "        sense_start_in_flank = sense_start - flank_start\n",
    "        sense_end_in_flank = sense_start_in_flank + sense_length\n",
    "\n",
    "        if 0 <= sense_start_in_flank < len(df_access) and sense_end_in_flank <= len(df_access):\n",
    "            values = df_access['avg_access'].iloc[sense_start_in_flank:sense_end_in_flank].dropna()\n",
    "            return values.mean() if not values.empty else None\n",
    "        else:\n",
    "            return None\n",
    "\n",
    "    except Exception as e:\n",
    "        print(f\"Error at row {row.name} | seq start: {row['sense_start']} | error: {e}\")\n",
    "        return None\n",
    "\n",
    "\n",
    "for gene, df in dfs.items():\n",
    "    FLANKED_SENSE_COL = f'sense_with_flank_{FLANK_SIZE}nt'\n",
    "\n",
    "    val = gene_to_data[gene].full_mrna\n",
    "    df['pre_mrna_sequence'] = [val] * len(df)\n",
    "\n",
    "    # Create new column with flanked sequences\n",
    "    df[FLANKED_SENSE_COL] = df.apply(\n",
    "        lambda row: get_sense_with_flanks(\n",
    "            row['pre_mrna_sequence'],\n",
    "            row['sense_start'],\n",
    "            row['sense_length'],\n",
    "            flank_size=FLANK_SIZE\n",
    "        ) if row['sense_start'] != -1 else \"\",  # Handle cases where sense was not found\n",
    "        axis=1\n",
    "    )\n",
    "\n",
    "    batch_size = 500\n",
    "    for start_idx in range(0, len(df), batch_size):\n",
    "        end_idx = min(start_idx + batch_size, len(df))\n",
    "        batch = df.iloc[start_idx:end_idx].copy()\n",
    "\n",
    "        print(f\"Processing rows {start_idx} to {end_idx}...\")\n",
    "\n",
    "        batch['sense_avg_accessibility'] = batch.apply(\n",
    "            compute_sense_accessibility,\n",
    "            axis=1,\n",
    "            flank_size=FLANK_SIZE,\n",
    "            access_win_size=ACCESS_WIN_SIZE,\n",
    "            seed_sizes=SEED_SIZES,\n",
    "            access_size=ACCESS_SIZE,\n",
    "        )\n",
    "\n",
    "        # Save batch to the new folder\n",
    "        batch.to_csv(f\"out/{gene}/batch_{start_idx}_{end_idx}.csv\", index=False)\n"
   ],
   "id": "845a217a41c7d20d",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Processing rows 0 to 500...\n",
      "Processing rows 500 to 855...\n",
      "Processing rows 0 to 500...\n",
      "Processing rows 500 to 1000...\n",
      "Processing rows 1000 to 1500...\n",
      "Processing rows 1500 to 1911...\n"
     ]
    }
   ],
   "execution_count": 136
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T23:34:50.707372Z",
     "start_time": "2025-09-17T23:34:50.666966Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "\n",
    "# Load all batch files from the new output folder\n",
    "files = sorted(glob.glob(f\"out/HUMAN_GFP/batch_*.csv\"))\n",
    "df_human = pd.concat([pd.read_csv(f) for f in files], axis=0)\n",
    "\n",
    "files = sorted(glob.glob(f\"out/YEAST_GFP/batch_*.csv\"))\n",
    "df_yeast = pd.concat([pd.read_csv(f) for f in files], axis=0)"
   ],
   "id": "f5f8625cdf898242",
   "outputs": [],
   "execution_count": 142
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T23:31:26.887099Z",
     "start_time": "2025-09-17T23:31:24.661396Z"
    }
   },
   "cell_type": "code",
   "source": [
    "\n",
    "from asodesigner.features.RNaseH_features import rnaseh1_dict, compute_rnaseh1_score\n",
    "\n",
    "best_window_start_krel = {\n",
    "    'R4a_krel': {10: 0, 11: 0, 12: 0, 13: 0, 14: 0, 15: 0, 16: 0, 17: 3, 18: 2, 19: 4, 20: 3, 21: 0, 22: 0, 25: 0},\n",
    "    'R4b_krel': {10: 0, 11: 0, 12: 0, 13: 0, 14: 0, 15: 0, 16: 0, 17: 1, 18: 3, 19: 1, 20: 3, 21: 0, 22: 0, 25: 0},\n",
    "    'R7_krel': {10: 0, 11: 0, 12: 0, 13: 0, 14: 0, 15: 0, 16: 3, 17: 2, 18: 4, 19: 6, 20: 4, 21: 0, 22: 0, 25: 0},\n",
    "}\n",
    "\n",
    "for exp in ['R4a_krel', 'R4b_krel', 'R7_krel']:\n",
    "    weights = rnaseh1_dict(exp)\n",
    "\n",
    "\n",
    "    def score_row(row):\n",
    "        length = len(row['Sequence'])\n",
    "        pos = best_window_start_krel.get(exp, {}).get(length, 0)\n",
    "        return compute_rnaseh1_score(row['Sequence'], weights, window_start=pos)\n",
    "\n",
    "    col_name = f\"RNaseH1_Krel_score_{exp}\"\n",
    "\n",
    "    # Yeast should have similar motifs, perhaps\n",
    "    df_human[col_name] = df_human.apply(score_row, axis=1)\n",
    "    df_yeast[col_name] = df_yeast.apply(score_row, axis=1)\n",
    "\n",
    "RNaseH1_Krel_features_best = [f\"RNaseH1_Krel_score_{exp}\" for exp in ['R4a_krel', 'R4b_krel', 'R7_krel']]\n"
   ],
   "id": "cdec49cbc7f8911f",
   "outputs": [],
   "execution_count": 138
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T23:31:27.042785Z",
     "start_time": "2025-09-17T23:31:27.026616Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Number is probably accurate enough\n",
    "df_human.loc[:, 'CAI_score_global_CDS'] = 0.7879152581622387\n",
    "df_yeast.loc[:, 'CAI_score_global_CDS'] = 0.5590591814785562"
   ],
   "id": "c59da38fa425784e",
   "outputs": [],
   "execution_count": 139
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T23:31:27.422386Z",
     "start_time": "2025-09-17T23:31:27.324729Z"
    }
   },
   "cell_type": "code",
   "source": [
    "df_human_scores = model.predict(df_human[selected_features])\n",
    "\n",
    "# Assuming you already have get_antisense(seq: str) -> str defined\n",
    "# get_antisense, for some reason numba doesn't work well\n",
    "tbl = str.maketrans(\"ACGTUacgtuNn\", \"TGCAAtgcaaNn\")\n",
    "\n",
    "df_human[\"score\"] = df_human_scores\n",
    "\n",
    "df_human[\"sense\"] = df_human[SEQUENCE].astype(str).str.translate(tbl).str[::-1]\n",
    "(\n",
    "    df_human.assign(score=df_human_scores)\n",
    "    .sort_values(\"score\", ascending=False)  # sort by score\n",
    "    .to_csv(\"df_human_scores_model2.csv\", index=False)\n",
    ")\n",
    "\n",
    "df_sorted = df_human.sort_values('score', ascending=False)"
   ],
   "id": "c5a652dd4e63f0c1",
   "outputs": [],
   "execution_count": 140
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T23:37:33.190647Z",
     "start_time": "2025-09-17T23:37:24.637803Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# pip install requests\n",
    "import math, time, threading, urllib.parse, requests\n",
    "from concurrent.futures import ThreadPoolExecutor, as_completed\n",
    "\n",
    "UA = {\"User-Agent\": \"python-requests gggenome/greedy\"}\n",
    "\n",
    "def _ggg_hits_leq_json(seq, k, db=\"hg38\", timeout=180, retries=2):\n",
    "    \"\"\"Count hits with <=k mismatches via GGGenome JSON; fallback to CSV if needed.\"\"\"\n",
    "    s = str(seq).upper().replace(\"U\", \"T\")\n",
    "    q = urllib.parse.quote(s)\n",
    "    url_json = f\"https://gggenome.dbcls.jp/{db}/{k}/nogap/{q}.json\"\n",
    "    url_csv  = f\"https://gggenome.dbcls.jp/{db}/{k}/nogap/{q}.csv?download\"\n",
    "\n",
    "    for attempt in range(retries + 1):\n",
    "        try:\n",
    "            r = requests.get(url_json, headers=UA, timeout=timeout)\n",
    "            r.raise_for_status()\n",
    "            try:\n",
    "                data = r.json()\n",
    "            except ValueError:\n",
    "                raise RuntimeError(\"JSON parse failed\")\n",
    "            if isinstance(data, list):\n",
    "                return len(data)\n",
    "            if isinstance(data, dict):\n",
    "                if \"results\" in data and isinstance(data[\"results\"], list): return len(data[\"results\"])\n",
    "                if \"hits\" in data and isinstance(data[\"hits\"], list):       return len(data[\"hits\"])\n",
    "                return sum(len(v) for v in data.values() if isinstance(v, list))\n",
    "            return 0\n",
    "        except Exception:\n",
    "            # greedy CSV fallback\n",
    "            try:\n",
    "                r2 = requests.get(url_csv, headers=UA, timeout=timeout)\n",
    "                r2.raise_for_status()\n",
    "                return sum(1 for ln in r2.text.splitlines() if ln and not ln.startswith(\"#\"))\n",
    "            except Exception:\n",
    "                if attempt < retries:\n",
    "                    continue\n",
    "                return 0\n",
    "    return 0\n",
    "\n",
    "def _d123_for_sequence(seq, db=\"hg38\"):\n",
    "    s = str(seq).upper().replace(\"U\", \"T\")\n",
    "    if not s:\n",
    "        return (s, 0, 0, 0)\n",
    "    L = len(s)\n",
    "    k_allowed = max(0, math.floor(0.25 * L))  # GGGenome cap\n",
    "    k0 = _ggg_hits_leq_json(s, 0, db=db)\n",
    "    k1 = _ggg_hits_leq_json(s, 1, db=db) if k_allowed >= 1 else 0\n",
    "    k2 = _ggg_hits_leq_json(s, 2, db=db) if k_allowed >= 2 else 0\n",
    "    k3 = _ggg_hits_leq_json(s, 3, db=db) if k_allowed >= 3 else 0\n",
    "    d1 = max(0, k1 - k0)\n",
    "    d2 = max(0, k2 - k1)\n",
    "    d3 = max(0, k3 - k2)\n",
    "    return (s, d1, d2, d3)\n",
    "\n",
    "def add_gggenome_d123(main_df, seq_col=\"SEQUENCE\", db=\"hg38\", *, max_workers=32, print_every=10):\n",
    "    seqs = (main_df[seq_col].astype(str).str.upper().str.replace(\"U\", \"T\", regex=False))\n",
    "    uniq = seqs.dropna().unique().tolist()\n",
    "    N = len(uniq)\n",
    "    print(f\"[GGG] Unique sequences: {N} | db={db} | workers={max_workers}\")\n",
    "\n",
    "    cache = {}\n",
    "    lock = threading.Lock()\n",
    "    t0 = time.perf_counter()\n",
    "    errs = 0\n",
    "\n",
    "    with ThreadPoolExecutor(max_workers=max_workers) as ex:\n",
    "        futs = {ex.submit(_d123_for_sequence, s, db): s for s in uniq}\n",
    "        done = 0\n",
    "        for fut in as_completed(futs):\n",
    "            s = futs[fut]\n",
    "            try:\n",
    "                s_key, d1, d2, d3 = fut.result()\n",
    "            except Exception:\n",
    "                d1 = d2 = d3 = 0\n",
    "                with lock:\n",
    "                    errs += 1\n",
    "            with lock:\n",
    "                cache[s] = (d1, d2, d3)\n",
    "                done += 1\n",
    "                if (done == 1) or (done % print_every == 0) or (done == N):\n",
    "                    elapsed = time.perf_counter() - t0\n",
    "                    rps = done / elapsed if elapsed > 0 else 0.0\n",
    "                    print(f\"[GGG] {done}/{N} cached | ~{rps:.1f} seq/s | errors={errs}\")\n",
    "\n",
    "    main_df[\"ggg_d1\"] = seqs.map(lambda s: cache.get(s, (0, 0, 0))[0])\n",
    "    main_df[\"ggg_d2\"] = seqs.map(lambda s: cache.get(s, (0, 0, 0))[1])\n",
    "    main_df[\"ggg_d3\"] = seqs.map(lambda s: cache.get(s, (0, 0, 0))[2])\n",
    "\n",
    "    print(f\"[GGG] Finished in {time.perf_counter() - t0:.1f}s. Added columns: ggg_d1, ggg_d2, ggg_d3\")\n",
    "    return main_df\n",
    "\n",
    "\n",
    "# --- usage ---\n",
    "# main_df = main_df[main_df[SENSE_START] != -1]\n",
    "result = add_gggenome_d123(df_sorted[:100], seq_col='sense', db=\"hg38\", max_workers=100, print_every=2)"
   ],
   "id": "4dd142e4e4cf7114",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[GGG] Unique sequences: 100 | db=hg38 | workers=100\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001B[31m---------------------------------------------------------------------------\u001B[39m",
      "\u001B[31mKeyboardInterrupt\u001B[39m                         Traceback (most recent call last)",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[144]\u001B[39m\u001B[32m, line 70\u001B[39m, in \u001B[36madd_gggenome_d123\u001B[39m\u001B[34m(main_df, seq_col, db, max_workers, print_every)\u001B[39m\n\u001B[32m     69\u001B[39m done = \u001B[32m0\u001B[39m\n\u001B[32m---> \u001B[39m\u001B[32m70\u001B[39m \u001B[38;5;28;01mfor\u001B[39;00m fut \u001B[38;5;129;01min\u001B[39;00m as_completed(futs):\n\u001B[32m     71\u001B[39m     s = futs[fut]\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/concurrent/futures/_base.py:243\u001B[39m, in \u001B[36mas_completed\u001B[39m\u001B[34m(fs, timeout)\u001B[39m\n\u001B[32m    239\u001B[39m         \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mTimeoutError\u001B[39;00m(\n\u001B[32m    240\u001B[39m                 \u001B[33m'\u001B[39m\u001B[38;5;132;01m%d\u001B[39;00m\u001B[33m (of \u001B[39m\u001B[38;5;132;01m%d\u001B[39;00m\u001B[33m) futures unfinished\u001B[39m\u001B[33m'\u001B[39m % (\n\u001B[32m    241\u001B[39m                 \u001B[38;5;28mlen\u001B[39m(pending), total_futures))\n\u001B[32m--> \u001B[39m\u001B[32m243\u001B[39m waiter.event.wait(wait_timeout)\n\u001B[32m    245\u001B[39m \u001B[38;5;28;01mwith\u001B[39;00m waiter.lock:\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/threading.py:629\u001B[39m, in \u001B[36mEvent.wait\u001B[39m\u001B[34m(self, timeout)\u001B[39m\n\u001B[32m    628\u001B[39m \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;129;01mnot\u001B[39;00m signaled:\n\u001B[32m--> \u001B[39m\u001B[32m629\u001B[39m     signaled = \u001B[38;5;28mself\u001B[39m._cond.wait(timeout)\n\u001B[32m    630\u001B[39m \u001B[38;5;28;01mreturn\u001B[39;00m signaled\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/threading.py:327\u001B[39m, in \u001B[36mCondition.wait\u001B[39m\u001B[34m(self, timeout)\u001B[39m\n\u001B[32m    326\u001B[39m \u001B[38;5;28;01mif\u001B[39;00m timeout \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[32m--> \u001B[39m\u001B[32m327\u001B[39m     waiter.acquire()\n\u001B[32m    328\u001B[39m     gotit = \u001B[38;5;28;01mTrue\u001B[39;00m\n",
      "\u001B[31mKeyboardInterrupt\u001B[39m: ",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001B[31mKeyboardInterrupt\u001B[39m                         Traceback (most recent call last)",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[144]\u001B[39m\u001B[32m, line 96\u001B[39m\n\u001B[32m     91\u001B[39m     \u001B[38;5;28;01mreturn\u001B[39;00m main_df\n\u001B[32m     94\u001B[39m \u001B[38;5;66;03m# --- usage ---\u001B[39;00m\n\u001B[32m     95\u001B[39m \u001B[38;5;66;03m# main_df = main_df[main_df[SENSE_START] != -1]\u001B[39;00m\n\u001B[32m---> \u001B[39m\u001B[32m96\u001B[39m result = add_gggenome_d123(df_sorted[:\u001B[32m100\u001B[39m], seq_col=\u001B[33m'\u001B[39m\u001B[33msense\u001B[39m\u001B[33m'\u001B[39m, db=\u001B[33m\"\u001B[39m\u001B[33mhg38\u001B[39m\u001B[33m\"\u001B[39m, max_workers=\u001B[32m100\u001B[39m, print_every=\u001B[32m2\u001B[39m)\n",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[144]\u001B[39m\u001B[32m, line 67\u001B[39m, in \u001B[36madd_gggenome_d123\u001B[39m\u001B[34m(main_df, seq_col, db, max_workers, print_every)\u001B[39m\n\u001B[32m     64\u001B[39m t0 = time.perf_counter()\n\u001B[32m     65\u001B[39m errs = \u001B[32m0\u001B[39m\n\u001B[32m---> \u001B[39m\u001B[32m67\u001B[39m \u001B[38;5;28;01mwith\u001B[39;00m ThreadPoolExecutor(max_workers=max_workers) \u001B[38;5;28;01mas\u001B[39;00m ex:\n\u001B[32m     68\u001B[39m     futs = {ex.submit(_d123_for_sequence, s, db): s \u001B[38;5;28;01mfor\u001B[39;00m s \u001B[38;5;129;01min\u001B[39;00m uniq}\n\u001B[32m     69\u001B[39m     done = \u001B[32m0\u001B[39m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/concurrent/futures/_base.py:647\u001B[39m, in \u001B[36mExecutor.__exit__\u001B[39m\u001B[34m(self, exc_type, exc_val, exc_tb)\u001B[39m\n\u001B[32m    646\u001B[39m \u001B[38;5;28;01mdef\u001B[39;00m\u001B[38;5;250m \u001B[39m\u001B[34m__exit__\u001B[39m(\u001B[38;5;28mself\u001B[39m, exc_type, exc_val, exc_tb):\n\u001B[32m--> \u001B[39m\u001B[32m647\u001B[39m     \u001B[38;5;28mself\u001B[39m.shutdown(wait=\u001B[38;5;28;01mTrue\u001B[39;00m)\n\u001B[32m    648\u001B[39m     \u001B[38;5;28;01mreturn\u001B[39;00m \u001B[38;5;28;01mFalse\u001B[39;00m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/concurrent/futures/thread.py:235\u001B[39m, in \u001B[36mThreadPoolExecutor.shutdown\u001B[39m\u001B[34m(self, wait, cancel_futures)\u001B[39m\n\u001B[32m    233\u001B[39m \u001B[38;5;28;01mif\u001B[39;00m wait:\n\u001B[32m    234\u001B[39m     \u001B[38;5;28;01mfor\u001B[39;00m t \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mself\u001B[39m._threads:\n\u001B[32m--> \u001B[39m\u001B[32m235\u001B[39m         t.join()\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/threading.py:1119\u001B[39m, in \u001B[36mThread.join\u001B[39m\u001B[34m(self, timeout)\u001B[39m\n\u001B[32m   1116\u001B[39m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mRuntimeError\u001B[39;00m(\u001B[33m\"\u001B[39m\u001B[33mcannot join current thread\u001B[39m\u001B[33m\"\u001B[39m)\n\u001B[32m   1118\u001B[39m \u001B[38;5;28;01mif\u001B[39;00m timeout \u001B[38;5;129;01mis\u001B[39;00m \u001B[38;5;28;01mNone\u001B[39;00m:\n\u001B[32m-> \u001B[39m\u001B[32m1119\u001B[39m     \u001B[38;5;28mself\u001B[39m._wait_for_tstate_lock()\n\u001B[32m   1120\u001B[39m \u001B[38;5;28;01melse\u001B[39;00m:\n\u001B[32m   1121\u001B[39m     \u001B[38;5;66;03m# the behavior of a negative timeout isn't documented, but\u001B[39;00m\n\u001B[32m   1122\u001B[39m     \u001B[38;5;66;03m# historically .join(timeout=x) for x<0 has acted as if timeout=0\u001B[39;00m\n\u001B[32m   1123\u001B[39m     \u001B[38;5;28mself\u001B[39m._wait_for_tstate_lock(timeout=\u001B[38;5;28mmax\u001B[39m(timeout, \u001B[32m0\u001B[39m))\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/threading.py:1139\u001B[39m, in \u001B[36mThread._wait_for_tstate_lock\u001B[39m\u001B[34m(self, block, timeout)\u001B[39m\n\u001B[32m   1136\u001B[39m     \u001B[38;5;28;01mreturn\u001B[39;00m\n\u001B[32m   1138\u001B[39m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[32m-> \u001B[39m\u001B[32m1139\u001B[39m     \u001B[38;5;28;01mif\u001B[39;00m lock.acquire(block, timeout):\n\u001B[32m   1140\u001B[39m         lock.release()\n\u001B[32m   1141\u001B[39m         \u001B[38;5;28mself\u001B[39m._stop()\n",
      "\u001B[31mKeyboardInterrupt\u001B[39m: "
     ]
    }
   ],
   "execution_count": 144
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T23:36:32.954039Z",
     "start_time": "2025-09-17T23:36:32.932604Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Show up to 200 characters per column\n",
    "pd.set_option(\"display.max_colwidth\", 200)\n",
    "\n",
    "# Show more columns across the screen\n",
    "pd.set_option(\"display.max_columns\", None)\n",
    "\n",
    "# Optionally, widen the console display\n",
    "pd.set_option(\"display.width\", 200)\n",
    "\n",
    "print(result[:30][[SEQUENCE, 'sense_start', 'ggg_d1', 'ggg_d2', 'at_skew', 'gc_content', 'on_target_fold_openness_normalized40_15', 'sense_avg_accessibility', 'RNaseH1_Krel_score_R7_krel']])\n"
   ],
   "id": "d8b7215e6341764f",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "             Sequence  sense_start  ggg_d1  ggg_d2   at_skew  gc_content  on_target_fold_openness_normalized40_15  sense_avg_accessibility  RNaseH1_Krel_score_R7_krel\n",
      "389  TTTCTTTGCAAGTTAT         1889       0       0 -0.500000      0.2500                                -0.097656                 5.035191                    1.017614\n",
      "390  ATTTCTTTGCAAGTTA         1890       0       0 -0.333333      0.2500                                -0.098307                 5.253077                    1.022290\n",
      "388  TTCTTTGCAAGTTATA         1888       0       0 -0.333333      0.2500                                -0.097005                 4.828204                    0.990100\n",
      "391  CATTTCTTTGCAAGTT         1891       0       0 -0.454545      0.3125                                -0.098958                 5.573556                    0.971061\n",
      "387  TCTTTGCAAGTTATAA         1887       0       0 -0.166667      0.2500                                -0.096354                 4.632415                    0.998657\n",
      "386  CTTTGCAAGTTATAAA         1886       0       0  0.000000      0.2500                                -0.095703                 4.435598                    0.971862\n",
      "383  TGCAAGTTATAAATAC         1883       0       0  0.166667      0.2500                                -0.102474                 3.872907                    1.027571\n",
      "384  TTGCAAGTTATAAATA         1884       0       0  0.076923      0.1875                                -0.096419                 4.065347                    0.998109\n",
      "392  TCATTTCTTTGCAAGT         1892       0       0 -0.454545      0.3125                                -0.099609                 5.968641                    0.949912\n",
      "382  GCAAGTTATAAATACT         1882       0       0  0.166667      0.2500                                -0.108529                 3.770627                    1.038217\n",
      "381  CAAGTTATAAATACTG         1881       0       0  0.166667      0.2500                                -0.114583                 3.750716                    1.002482\n",
      "378  GTTATAAATACTGAAT         1878       0       0  0.076923      0.1875                                -0.155924                 3.753932                    0.967088\n",
      "385  TTTGCAAGTTATAAAT         1885       0       0 -0.076923      0.1875                                -0.095052                 4.239374                    0.963692\n",
      "379  AGTTATAAATACTGAA         1879       0       0  0.230769      0.1875                                -0.140299                 3.689726                    0.968562\n",
      "380  AAGTTATAAATACTGA         1880       0       0  0.230769      0.1875                                -0.120638                 3.704746                    0.996462\n",
      "377  TTATAAATACTGAATA         1877       0       0  0.142857      0.1250                                -0.171549                 3.928889                    0.964560\n",
      "403  CTCTCTGATATTCATT         1903       0       0 -0.454545      0.3125                                -0.161198                      NaN                    0.995907\n",
      "264  GGGCACACACTACTTG         1764       0       0  0.142857      0.5625                                -0.737956                 5.752672                    0.997139\n",
      "402  TCTCTGATATTCATTT         1902       0       0 -0.500000      0.2500                                -0.155078                      NaN                    0.981611\n",
      "324  TTTTCCACACTGACTA         1824       0       0 -0.200000      0.3750                                -0.448698                 5.831167                    0.973013\n",
      "397  GATATTCATTTCTTTG         1897       0       0 -0.500000      0.2500                                -0.117969                 7.240007                    0.959497\n",
      "393  TTCATTTCTTTGCAAG         1893       0       0 -0.454545      0.3125                                -0.100260                 6.305983                    0.960240\n",
      "400  TCTGATATTCATTTCT         1900       0       0 -0.500000      0.2500                                -0.142839                      NaN                    0.984611\n",
      "396  ATATTCATTTCTTTGC         1896       0       0 -0.500000      0.2500                                -0.110938                 6.996143                    0.926295\n",
      "401  CTCTGATATTCATTTC         1901       0       0 -0.454545      0.3125                                -0.148958                      NaN                    0.971986\n",
      "449  TGCTGGTAGTGGTCGG          538       0       0 -0.666667      0.6250                                -0.383724                 1.188746                    1.004494\n",
      "394  ATTCATTTCTTTGCAA         1894       0       0 -0.333333      0.2500                                -0.100911                 6.547222                    0.964742\n",
      "284  TCTTCTTCTTAGCCTG          873       0       0 -0.777778      0.4375                                -0.234049                 5.774980                    0.972656\n",
      "466  CCGTTCTTCTGCTTGT          466       0       0 -1.000000      0.5000                                -0.175456                 4.030456                    0.951062\n",
      "472  TTGATGCCGTTCTTCT          472       0       0 -0.777778      0.4375                                -0.165755                 4.318761                    0.977098\n"
     ]
    }
   ],
   "execution_count": 143
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T23:11:56.594883Z",
     "start_time": "2025-09-17T23:11:56.570294Z"
    }
   },
   "cell_type": "code",
   "source": "# Appendix for CAI",
   "id": "7fd981ffddb03542",
   "outputs": [],
   "execution_count": 115
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from math import prod\n",
    "\n",
    "# Yeast codon usage table (S. cerevisiae, from Codon Usage Database)\n",
    "# frequency per thousand, normalized to relative adaptiveness (w values)\n",
    "# For simplicity, store as preferred-codon relative adaptiveness\n",
    "yeast_codon_pref = {\n",
    "    'TTT':0.61,'TTC':1.0,'TTA':0.19,'TTG':0.45,'CTT':0.10,'CTC':0.08,'CTA':0.07,'CTG':0.09,\n",
    "    'ATT':0.36,'ATC':1.0,'ATA':0.12,'ATG':1.0,\n",
    "    'GTT':0.47,'GTC':0.15,'GTA':0.14,'GTG':0.25,\n",
    "    'TCT':0.55,'TCC':0.54,'TCA':0.37,'TCG':0.17,'AGT':0.39,'AGC':1.0,\n",
    "    'CCT':0.61,'CCC':0.21,'CCA':0.63,'CCG':0.15,\n",
    "    'ACT':0.60,'ACC':1.0,'ACA':0.36,'ACG':0.13,\n",
    "    'GCT':0.82,'GCC':0.49,'GCA':0.58,'GCG':0.13,\n",
    "    'TAT':0.57,'TAC':1.0,'TAA':0,'TAG':0,'TGA':0,\n",
    "    'CAT':0.57,'CAC':1.0,\n",
    "    'CAA':0.34,'CAG':1.0,\n",
    "    'AAT':0.53,'AAC':1.0,\n",
    "    'AAA':0.44,'AAG':1.0,\n",
    "    'GAT':0.46,'GAC':1.0,\n",
    "    'GAA':0.38,'GAG':1.0,\n",
    "    'TGT':0.44,'TGC':1.0,'TGG':1.0,\n",
    "    'CGT':0.56,'CGC':0.65,'CGA':0.14,'CGG':0.14,'AGA':0.05,'AGG':0.02,\n",
    "    'GGT':0.53,'GGC':0.49,'GGA':1.0,'GGG':0.27\n",
    "}\n",
    "\n",
    "# Input sequence (from user)\n",
    "seq = (\"ATGGTtAGtAAaGGaGAaGAGTTgTTCACaGGaGTGGTGCCCATCCTGGTCGAGCTGGACGGCGACGTAAACGGCCACAAGTTCAGCGTGTCCGGCGAGGGCGAGGGCGATGCCACCTACGGCAAGCTGACCCTGAAGTTCATCTGCACCACCGGCAAGCTGCCCGTGCCCTGGCCCACCCTCGTGACCACCCTGACCTACGGCGTGCAGTGCTTCAGCCGCTACCCCGACCACATGAAGCAGCACGACTTCTTCAAGTCCGCCATGCCCGAAGGCTACGTCCAGGAGCGCACCATCTTCTTCAAGGACGACGGCAACTACAAGACCCGCGCCGAGGTGAAGTTCGAGGGCGACACCCTGGTGAACCGCATCGAGCTGAAGGGCATCGACTTCAAGGAGGACGGCAACATCCTGGGGCACAAGCTGGAGTACAACTACAACAGCCACAACGTCTATATCATGGCCGACAAGCAGAAGAACGGCATCAAGGTGAACTTCAAGATCCGCCACAACATCGAGGACGGCAGCGTGCAGCTCGCCGACCACTACCAGCAGAACACCCCCATCGGCGACGGCCCCGTGCTGCTGCCCGACAACCACTACCTGAGCACCCAGTCCGCCCTGAGCAAAGACCCCAACGAGAAGCGCGATCACATGGTCCTGCTGGAGTTCGTGACCGCCGCCGGGATCACTCTCGGCATGGACGAGCTGTACAAGGGTGCTGGGGCAggtacCCCTAAAGATCCAGCCAAACCTCCGGCCAcGGCACAAGTTGTGGGATGGCCACCGGTGAGATCATACCGGAAGAACGTGATGGTTTCCTGCCAAAAATCAAGCGGTGGCCCGGAGGCGGCGGCGTTCGTGAAGTAA\").upper()\n",
    "\n",
    "# Split into codons\n",
    "codons = [seq[i:i+3] for i in range(0, len(seq), 3)]\n",
    "# Drop stop codon and any incomplete codons at the end\n",
    "valid_codons = [c for c in codons if c in yeast_codon_pref and yeast_codon_pref[c] > 0]\n",
    "\n",
    "# Calculate weights\n",
    "weights = [yeast_codon_pref[c] for c in valid_codons]\n",
    "\n",
    "# Compute CAI\n",
    "cai = prod(weights) ** (1/len(weights))\n",
    "cai\n"
   ],
   "id": "c325fee15269f921"
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": [
    "from math import prod\n",
    "\n",
    "# Human codon usage table (Homo sapiens, whole genome, from Kazusa)\n",
    "# Values are normalized as relative adaptiveness (w)\n",
    "# Frequencies per 1000 codons (rounded from Kazusa)\n",
    "human_freqs = {\n",
    "    'TTT':17.6,'TTC':20.3,'TTA':7.7,'TTG':12.9,'CTT':13.2,'CTC':19.6,'CTA':7.2,'CTG':39.6,\n",
    "    'ATT':16.0,'ATC':20.8,'ATA':7.5,'ATG':22.0,\n",
    "    'GTT':10.8,'GTC':14.6,'GTA':7.1,'GTG':28.1,\n",
    "    'TCT':15.2,'TCC':17.6,'TCA':12.2,'TCG':4.5,'AGT':11.9,'AGC':19.5,\n",
    "    'CCT':17.5,'CCC':19.8,'CCA':16.9,'CCG':6.9,\n",
    "    'ACT':13.1,'ACC':21.2,'ACA':15.1,'ACG':6.1,\n",
    "    'GCT':18.2,'GCC':27.7,'GCA':15.8,'GCG':7.4,\n",
    "    'TAT':12.0,'TAC':15.6,'TAA':1.0,'TAG':0.8,'TGA':1.6,\n",
    "    'CAT':10.9,'CAC':15.1,\n",
    "    'CAA':12.3,'CAG':34.2,\n",
    "    'AAT':17.0,'AAC':19.1,\n",
    "    'AAA':24.4,'AAG':32.5,\n",
    "    'GAT':22.3,'GAC':26.0,\n",
    "    'GAA':29.0,'GAG':39.6,\n",
    "    'TGT':10.8,'TGC':12.1,'TGG':13.2,\n",
    "    'CGT':4.7,'CGC':10.5,'CGA':6.2,'CGG':11.4,'AGA':12.1,'AGG':11.5,\n",
    "    'GGT':10.8,'GGC':22.2,'GGA':16.5,'GGG':16.5\n",
    "}\n",
    "\n",
    "# Normalize to get relative adaptiveness (w)\n",
    "human_w = {}\n",
    "for codon in human_freqs:\n",
    "    aa_group = [c for c in human_freqs if c[0] == codon[0] or True]  # lazy grouping\n",
    "# Let's just hard-code amino acid codon groupings\n",
    "aa_codons = {\n",
    "    'F':['TTT','TTC'],\n",
    "    'L':['TTA','TTG','CTT','CTC','CTA','CTG'],\n",
    "    'I':['ATT','ATC','ATA'],\n",
    "    'M':['ATG'],\n",
    "    'V':['GTT','GTC','GTA','GTG'],\n",
    "    'S':['TCT','TCC','TCA','TCG','AGT','AGC'],\n",
    "    'P':['CCT','CCC','CCA','CCG'],\n",
    "    'T':['ACT','ACC','ACA','ACG'],\n",
    "    'A':['GCT','GCC','GCA','GCG'],\n",
    "    'Y':['TAT','TAC'],\n",
    "    'H':['CAT','CAC'],\n",
    "    'Q':['CAA','CAG'],\n",
    "    'N':['AAT','AAC'],\n",
    "    'K':['AAA','AAG'],\n",
    "    'D':['GAT','GAC'],\n",
    "    'E':['GAA','GAG'],\n",
    "    'C':['TGT','TGC'],\n",
    "    'W':['TGG'],\n",
    "    'R':['CGT','CGC','CGA','CGG','AGA','AGG'],\n",
    "    'G':['GGT','GGC','GGA','GGG'],\n",
    "    'STOP':['TAA','TAG','TGA']\n",
    "}\n",
    "for aa, codon_list in aa_codons.items():\n",
    "    maxf = max(human_freqs[c] for c in codon_list)\n",
    "    for c in codon_list:\n",
    "        if human_freqs[c] > 0:\n",
    "            human_w[c] = human_freqs[c]/maxf\n",
    "        else:\n",
    "            human_w[c] = 0\n",
    "\n",
    "# Input sequence (human GFP variant)\n",
    "seq_human = (\"atggtgagcaagggcgaggagctgttcaccggggtggtgcccatcctggtcgagctggacggcgacgtaaacggccacaagttcagcgtgtccggcgagggcgagggcgatgccacctacggcaagctgaccctgaagttcatctgcaccaccggcaagctgcccgtgccctggcccaccctcgtgaccaccctgacctacggcgtgcagtgcttcagccgctaccccgaccacatgaagcagcacgacttcttcaagtccgccatgcccgaaggctacgtccaggagcgcaccatcttcttcaaggacgacggcaactacaagacccgcgccgaggtgaagttcgagggcgacaccctggtgaaccgcatcgagctgaagggcatcgacttcaaggaggacggcaacatcctggggcacaagctggagtacaactacaacagccacaacgtctatatcatggccgacaagcagaagaacggcatcaaggtgaacttcaagatccgccacaacatcgaggacggcagcgtgcagctcgccgaccactaccagcagaacacccccatcggcgacggccccgtgctgctgcccgacaaccactacctgagcacccagtccgccctgagcaaagaccccaacgagaagcgcgatcacatggtcctgctggagttcgtgaccgccgccgggatcactctcggcatggacgagctgtacaagaagcttagccatggcttcccgccggaggtggaggagcaggatgatggcacgctgcccatgtcttgtgcccaggagagcgggatggaccgtcaccctgcagcctgtgcttctgctaggatcaatgtgaagcgacctgccgccacaaagaaggctggacaggctaagaagaagaaatgaggatcccgcgcgcgcatatgttaattaaccaactgcatggggatccacgcgttaagtcgacaatcaacctctggattacaaaatttgtgaaagattgactggtattcttaactatgttgctccttttacgctatgtggatacgctgctttaatgcctttgtatcatgctattgcttcccgtatggctttcattttctcctccttgtataaatcctggttgctgtctctttatgaggagttgtggcccgttgtcaggcaacgtggcgtggtgtgcactgtgtttgctgacgcaacccccactggttggggcattgccaccacctgtcagctcctttccgggactttcgctttccccctccctattgccacggcggaactcatcgccgcctgccttgcccgctgctggacaggggctcggctgttgggcactgacaattccgtggtgttgtcggggaaatcatcgtcctttccttggctgctcgcctgtgttgccacctggattctgcgcgggacgtccttctgctacgtcccttcggccctcaatccagcggaccttccttcccgcggcctgctgccggctctgcggcctcttccgcgtcttcgccttcgccctcagacgagtcggatctccctttgggccgcctccccgcgtcgactttaagaccaatgacttacaaggcagctgtagatcttagccactttttaaaagaaaaggggggactggaagggctaattcactcccaacgaagacaagatctgctttttgcttgtactgggtctctctggttagaccagatctgagcctgggagctctctggctaactagggaacccactgcttaagcctcaataaagcttgccttgagtgcttcaagtagtgtgtgcccgtctgttgtgtgactctggtaactagagatccctcagacccttttagtcagtgtggaaaatctctagcagtacgtatagtagttcatgtcatcttattattcagtatttataacttgcaaagaaatgaatatcagagagtgagagg\")\n",
    "\n",
    "# Split into codons\n",
    "codons_h = [seq_human[i:i+3].upper() for i in range(0, len(seq_human), 3)]\n",
    "valid_codons_h = [c for c in codons_h if c in human_w and human_w[c] > 0]\n",
    "\n",
    "weights_h = [human_w[c] for c in valid_codons_h]\n",
    "cai_h = prod(weights_h) ** (1/len(weights_h))\n",
    "cai_h\n"
   ],
   "id": "58d620bc7716dfdb"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
