{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "d4e0823384ae0a08",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-18T21:36:31.301082Z",
     "start_time": "2025-04-18T21:36:30.912231Z"
    }
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "\n",
    "project_root = os.path.abspath(os.path.join(os.getcwd(), \"../..\"))\n",
    "if project_root not in sys.path:\n",
    "    sys.path.insert(0, project_root)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "d949d2d5",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1562/3800543916.py:6: DtypeWarning: Columns (20) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  aso_df = pd.read_csv(DATA_PATH / 'data_asoptimizer_updated.csv')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>ISIS</th>\n",
       "      <th>Target_gene</th>\n",
       "      <th>Cell_line</th>\n",
       "      <th>Density(cells/well)</th>\n",
       "      <th>Transfection</th>\n",
       "      <th>ASO_volume(nM)</th>\n",
       "      <th>Treatment_Period(hours)</th>\n",
       "      <th>Primer_probe_set</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>...</th>\n",
       "      <th>Inhibition(%)</th>\n",
       "      <th>seq_length</th>\n",
       "      <th>Canonical Gene Name</th>\n",
       "      <th>Cell line organism</th>\n",
       "      <th>Transcript</th>\n",
       "      <th>Location_in_sequence</th>\n",
       "      <th>Location_div_by_length</th>\n",
       "      <th>true_length_of_seq</th>\n",
       "      <th>mod_scan</th>\n",
       "      <th>cell_line_uniform</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>540733</td>\n",
       "      <td>K-RAS</td>\n",
       "      <td>A431</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>free uptake</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>24</td>\n",
       "      <td>RTS3496_MGB</td>\n",
       "      <td>GCTAAAACAAATGCTA</td>\n",
       "      <td>...</td>\n",
       "      <td>33.0</td>\n",
       "      <td>16</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41212.0</td>\n",
       "      <td>0.901972</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>A431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>540747</td>\n",
       "      <td>K-RAS</td>\n",
       "      <td>A431</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>free uptake</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>24</td>\n",
       "      <td>RTS3496_MGB</td>\n",
       "      <td>TATAATGGTGAATATC</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>16</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23686.0</td>\n",
       "      <td>0.518395</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>A431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>540806</td>\n",
       "      <td>K-RAS</td>\n",
       "      <td>A431</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>free uptake</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>24</td>\n",
       "      <td>RTS3496_MGB</td>\n",
       "      <td>GCATGAAGATTTCTGG</td>\n",
       "      <td>...</td>\n",
       "      <td>62.0</td>\n",
       "      <td>16</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>43363.0</td>\n",
       "      <td>0.949049</td>\n",
       "      <td>16</td>\n",
       "      <td>1</td>\n",
       "      <td>A431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>651479</td>\n",
       "      <td>K-RAS</td>\n",
       "      <td>A431</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>free uptake</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>24</td>\n",
       "      <td>RTS3496_MGB</td>\n",
       "      <td>GGTGAATATCTTCAAA</td>\n",
       "      <td>...</td>\n",
       "      <td>28.0</td>\n",
       "      <td>16</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>23680.0</td>\n",
       "      <td>0.518264</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>A431</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>651490</td>\n",
       "      <td>K-RAS</td>\n",
       "      <td>A431</td>\n",
       "      <td>5000.0</td>\n",
       "      <td>free uptake</td>\n",
       "      <td>2000.0</td>\n",
       "      <td>24</td>\n",
       "      <td>RTS3496_MGB</td>\n",
       "      <td>CACTTGTACTAGTATG</td>\n",
       "      <td>...</td>\n",
       "      <td>36.0</td>\n",
       "      <td>16</td>\n",
       "      <td>KRAS</td>\n",
       "      <td>human</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41168.0</td>\n",
       "      <td>0.901009</td>\n",
       "      <td>16</td>\n",
       "      <td>0</td>\n",
       "      <td>A431</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   index    ISIS Target_gene Cell_line  Density(cells/well) Transfection  \\\n",
       "0      0  540733       K-RAS      A431               5000.0  free uptake   \n",
       "1      1  540747       K-RAS      A431               5000.0  free uptake   \n",
       "2      2  540806       K-RAS      A431               5000.0  free uptake   \n",
       "3      3  651479       K-RAS      A431               5000.0  free uptake   \n",
       "4      4  651490       K-RAS      A431               5000.0  free uptake   \n",
       "\n",
       "   ASO_volume(nM)  Treatment_Period(hours) Primer_probe_set          Sequence  \\\n",
       "0          2000.0                       24      RTS3496_MGB  GCTAAAACAAATGCTA   \n",
       "1          2000.0                       24      RTS3496_MGB  TATAATGGTGAATATC   \n",
       "2          2000.0                       24      RTS3496_MGB  GCATGAAGATTTCTGG   \n",
       "3          2000.0                       24      RTS3496_MGB  GGTGAATATCTTCAAA   \n",
       "4          2000.0                       24      RTS3496_MGB  CACTTGTACTAGTATG   \n",
       "\n",
       "   ... Inhibition(%) seq_length Canonical Gene Name Cell line organism  \\\n",
       "0  ...          33.0         16                KRAS              human   \n",
       "1  ...           7.0         16                KRAS              human   \n",
       "2  ...          62.0         16                KRAS              human   \n",
       "3  ...          28.0         16                KRAS              human   \n",
       "4  ...          36.0         16                KRAS              human   \n",
       "\n",
       "  Transcript Location_in_sequence  Location_div_by_length  true_length_of_seq  \\\n",
       "0        NaN              41212.0                0.901972                  16   \n",
       "1        NaN              23686.0                0.518395                  16   \n",
       "2        NaN              43363.0                0.949049                  16   \n",
       "3        NaN              23680.0                0.518264                  16   \n",
       "4        NaN              41168.0                0.901009                  16   \n",
       "\n",
       "  mod_scan cell_line_uniform  \n",
       "0        0              A431  \n",
       "1        0              A431  \n",
       "2        1              A431  \n",
       "3        0              A431  \n",
       "4        0              A431  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# conda activate aso_design\n",
    "import pandas as pd\n",
    "from scripts.data_genertion.consts import *\n",
    "from asodesigner.file_utils import read_human_genome_fasta_dict\n",
    "from asodesigner.consts import *\n",
    "aso_df = pd.read_csv(DATA_PATH / 'data_asoptimizer_updated.csv')\n",
    "aso_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89f64fd7",
   "metadata": {},
   "source": [
    "Reading the A expression file that contain the RNA expression matrix for each cell line"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "ad93114f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    DepMap_ID  TSPAN6 (7105)  TNMD (64102)  DPM1 (8813)  SCYL3 (57147)  \\\n",
      "0  ACH-001113       4.956577      0.000000     7.577648       3.179411   \n",
      "1  ACH-001289       4.954992      0.617243     7.334747       2.783576   \n",
      "2  ACH-001339       3.421952      0.000000     7.546069       2.615880   \n",
      "3  ACH-001979       4.651643      0.000000     5.946408       2.454515   \n",
      "4  ACH-002438       4.336705      0.000000     6.879387       2.262824   \n",
      "\n",
      "   FIRRM (55732)  FGR (2268)  CFH (3075)  FUCA2 (2519)  GCLC (2729)  ...  \\\n",
      "0       4.765742    0.037483    1.510547      3.103039     6.697482  ...   \n",
      "1       3.736280    0.000000    0.149060      3.855960     4.377061  ...   \n",
      "2       4.476233    0.064571    1.397491      6.833915     3.980336  ...   \n",
      "3       1.852111    0.000000    7.680796      6.156866     5.042574  ...   \n",
      "4       3.256491    0.000000    2.253926      6.895436     3.470992  ...   \n",
      "\n",
      "   H3C3 (8352)  CIMIP3 (114841037)  SMIM42 (117981789)  NPBWR1 (2831)  \\\n",
      "0     0.093516                 0.0                 0.0       0.414727   \n",
      "1     0.672134                 0.0                 0.0       0.029840   \n",
      "2     0.689369                 0.0                 0.0       0.127768   \n",
      "3     0.000000                 0.0                 0.0       0.012012   \n",
      "4     0.278721                 0.0                 0.0       0.024008   \n",
      "\n",
      "   ACTL10 (170487)  PANO1 (101927423)  HRURF (120766137)  PRRC2B (84726)  \\\n",
      "0         0.077634           0.411901                0.0        5.134808   \n",
      "1         0.000000           1.026871                0.0        5.951231   \n",
      "2         0.105594           0.540575                0.0        4.205971   \n",
      "3         0.000000           0.729221                0.0        4.922916   \n",
      "4         0.984538           0.549131                0.0        4.949465   \n",
      "\n",
      "   F8A2 (474383)  F8A1 (8263)  \n",
      "0       1.214541     4.315653  \n",
      "1       0.007227     4.251060  \n",
      "2       0.030894     2.783448  \n",
      "3       0.150926     4.661556  \n",
      "4       0.000000     6.202939  \n",
      "\n",
      "[5 rows x 19206 columns]\n"
     ]
    }
   ],
   "source": [
    "expr = pd.read_csv(\"OmicsExpressionProteinCodingGenesTPMLogp1.csv\")\n",
    "expr = expr.rename(columns={expr.columns[0]: \"DepMap_ID\"})\n",
    "print(expr.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d65f5c41",
   "metadata": {},
   "source": [
    "loading the Metadata mapping ACH IDs to readable cell line names."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "8e3c09a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = pd.read_csv(\"Model.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6158b6d4",
   "metadata": {},
   "source": [
    "Merge the two df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "ec4e44a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       cell line  TSPAN6 (7105)  TNMD (64102)  DPM1 (8813)  SCYL3 (57147)  \\\n",
      "0        LC1SQSF       4.956577      0.000000     7.577648       3.179411   \n",
      "1       COGAR359       4.954992      0.617243     7.334747       2.783576   \n",
      "2        COLO794       3.421952      0.000000     7.546069       2.615880   \n",
      "3          NZM11       4.651643      0.000000     5.946408       2.454515   \n",
      "4  CCLFPEDS0026T       4.336705      0.000000     6.879387       2.262824   \n",
      "\n",
      "   FIRRM (55732)  FGR (2268)  CFH (3075)  FUCA2 (2519)  GCLC (2729)  ...  \\\n",
      "0       4.765742    0.037483    1.510547      3.103039     6.697482  ...   \n",
      "1       3.736280    0.000000    0.149060      3.855960     4.377061  ...   \n",
      "2       4.476233    0.064571    1.397491      6.833915     3.980336  ...   \n",
      "3       1.852111    0.000000    7.680796      6.156866     5.042574  ...   \n",
      "4       3.256491    0.000000    2.253926      6.895436     3.470992  ...   \n",
      "\n",
      "   CIMIP3 (114841037)  SMIM42 (117981789)  NPBWR1 (2831)  ACTL10 (170487)  \\\n",
      "0                 0.0                 0.0       0.414727         0.077634   \n",
      "1                 0.0                 0.0       0.029840         0.000000   \n",
      "2                 0.0                 0.0       0.127768         0.105594   \n",
      "3                 0.0                 0.0       0.012012         0.000000   \n",
      "4                 0.0                 0.0       0.024008         0.984538   \n",
      "\n",
      "   PANO1 (101927423)  HRURF (120766137)  PRRC2B (84726)  F8A2 (474383)  \\\n",
      "0           0.411901                0.0        5.134808       1.214541   \n",
      "1           1.026871                0.0        5.951231       0.007227   \n",
      "2           0.540575                0.0        4.205971       0.030894   \n",
      "3           0.729221                0.0        4.922916       0.150926   \n",
      "4           0.549131                0.0        4.949465       0.000000   \n",
      "\n",
      "   F8A1 (8263)     ModelID  \n",
      "0     4.315653  ACH-001113  \n",
      "1     4.251060  ACH-001289  \n",
      "2     2.783448  ACH-001339  \n",
      "3     4.661556  ACH-001979  \n",
      "4     6.202939  ACH-002438  \n",
      "\n",
      "[5 rows x 19207 columns]\n"
     ]
    }
   ],
   "source": [
    "expr_named = expr.merge(\n",
    "    model[[\"ModelID\", \"StrippedCellLineName\"]],\n",
    "    left_on=\"DepMap_ID\", right_on=\"ModelID\", how=\"left\"\n",
    ")\n",
    "expr_named = expr_named.drop(columns=[\"DepMap_ID\"])\n",
    "expr_named = expr_named.rename(columns={\"StrippedCellLineName\": \"cell line\"})\n",
    "cols = [\"cell line\"] + [c for c in expr_named.columns if c != \"cell line\"]\n",
    "expr_named = expr_named[cols]\n",
    "print(expr_named.head())\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c78bac10",
   "metadata": {},
   "source": [
    "Reading the catalog of RNA-binding proteins (RBPs) and their motifs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "ac0874b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "motifs= pd.read_csv(\"RBS_motifs_Homo_sapiens.csv\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "add33130",
   "metadata": {},
   "source": [
    "merging all the df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "dfad3853",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Rows after merge: 5478052\n",
      "  cell line      Gene_raw Expression Gene_name          Gene_id Mutated  \\\n",
      "0   LC1SQSF  RBM5 (10181)   5.591937      RBM5  ENSG00000003756      no   \n",
      "1   LC1SQSF  RBM5 (10181)   5.591937      RBM5  ENSG00000003756      no   \n",
      "2   LC1SQSF  RBM5 (10181)   5.591937      RBM5  ENSG00000003756      no   \n",
      "3   LC1SQSF  RBM5 (10181)   5.591937      RBM5  ENSG00000003756      no   \n",
      "4   LC1SQSF  RBM5 (10181)   5.591937      RBM5  ENSG00000003756      no   \n",
      "\n",
      "       Organism      Motif  Len  \\\n",
      "0  Homo_sapiens    AAAAAAA    7   \n",
      "1  Homo_sapiens     AGGUAA    6   \n",
      "2  Homo_sapiens    CCCCCCC    7   \n",
      "3  Homo_sapiens  CUCUUCUCU    9   \n",
      "4  Homo_sapiens    GAAGGAA    7   \n",
      "\n",
      "                              Experiment_description Database    Pubmed  \\\n",
      "0  Homopolymer binding assay with recombinant pro...        S  11029660   \n",
      "1                       Chemical shift mapping (NMR)        S  22162216   \n",
      "2  Homopolymer binding assay with recombinant pro...        S  11029660   \n",
      "3  EMSA and UV cross-linking with recombinant pro...        S  18840686   \n",
      "4                                         RNAcompete        C  23846655   \n",
      "\n",
      "  Experiment_description.1    Family Matrix_id       Score  \n",
      "0                    other  ZF_RANBP      1151  1.000000**  \n",
      "1             SOLUTION NMR  ZF_RANBP      1152  1.000000**  \n",
      "2                    other  ZF_RANBP      1150  1.000000**  \n",
      "3    UV cross-linking;EMSA  ZF_RANBP      1148  1.000000**  \n",
      "4               RNAcompete  ZF_RANBP  M053_0.6  0.177741**  \n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "# --- 1) Helper: canonicalize gene symbols (strip parentheses, trim, uppercase) ---\n",
    "def canonical_symbol(s: str) -> str:\n",
    "    s = str(s).strip()\n",
    "    # remove trailing \" (12345)\" or any parentheses block at the end\n",
    "    s = re.sub(r\"\\s*\\(.*?\\)\\s*$\", \"\", s)\n",
    "    # collapse internal whitespace\n",
    "    s = re.sub(r\"\\s+\", \"\", s)\n",
    "    return s.upper()\n",
    "\n",
    "# --- 2) Melt expression to long and add a canonical gene column ---\n",
    "expr_long = expr_named.melt(\n",
    "    id_vars=[\"cell line\"],\n",
    "    var_name=\"Gene_raw\",\n",
    "    value_name=\"Expression\"\n",
    ")\n",
    "expr_long[\"Gene_name\"] = expr_long[\"Gene_raw\"].apply(canonical_symbol)\n",
    "\n",
    "# (optional) keep only real genes (drop the 'cell line' row if it leaked)\n",
    "expr_long = expr_long[expr_long[\"Gene_name\"].ne(\"CELLLINE\")]\n",
    "\n",
    "\n",
    "# make sure the column exists; if it's differently named, change here\n",
    "assert \"Gene_name\" in motifs.columns, \"Expected 'Gene_name' column in motifs file.\"\n",
    "\n",
    "motifs = motifs.copy()\n",
    "motifs[\"Gene_name\"] = motifs[\"Gene_name\"].apply(canonical_symbol)\n",
    "\n",
    "\n",
    "\n",
    "# --- 4) Merge on the canonical symbol ---\n",
    "merged = expr_long.merge(motifs, on=\"Gene_name\", how=\"inner\")\n",
    "\n",
    "# sanity check\n",
    "print(\"Rows after merge:\", len(merged))\n",
    "print(merged.head())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f62ac8ed",
   "metadata": {},
   "source": [
    "Filter merged table to only ASO-relevant cell lines (only for the cell lines in our ASO data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "f3270497",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Original size: 5478052\n",
      "Filtered size: 29277\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "def _norm_cell(v: str) -> str:\n",
    "    # Uppercase + remove spaces/hyphens/underscores/slashes\n",
    "    s = str(v).upper()\n",
    "    s = re.sub(r\"[\\s\\-\\_/]+\", \"\", s)\n",
    "    return s\n",
    "\n",
    "# Unique cell lines from ASO file (normalized)\n",
    "aso_cell_lines = {_norm_cell(c) for c in aso_df[\"Cell_line\"].dropna()}\n",
    "\n",
    "# Add A549 manually (normalized)\n",
    "aso_cell_lines.add(_norm_cell(\"A549\"))\n",
    "\n",
    "# Normalize the \"cell line\" column in merged and filter\n",
    "merged_norm = merged[\"cell line\"].astype(str).map(_norm_cell)\n",
    "filtered = merged[ merged_norm.isin(aso_cell_lines) ].copy()\n",
    "\n",
    "print(\"Original size:\", len(merged))\n",
    "print(\"Filtered size:\", len(filtered))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0b617395",
   "metadata": {},
   "source": [
    "save as csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d9b151f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "filtered.to_csv(\"RBP_motifs_for_each_cell_line_with_mRNA_expression.csv\",index=False )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "aso_design",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
