{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {
    "collapsed": true,
    "ExecuteTime": {
     "end_time": "2025-09-14T21:00:25.608191Z",
     "start_time": "2025-09-14T21:00:15.434374Z"
    }
   },
   "source": [
    "from xgboost import XGBRegressor\n",
    "\n",
    "from scripts.data_genertion.consts import *\n",
    "from scripts.features.feature_extraction import load_all_features\n",
    "\n",
    "main_df = load_all_features()\n",
    "main_df[SEQUENCE] = main_df[SEQUENCE].astype(str)"
   ],
   "outputs": [],
   "execution_count": 1
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-14T21:00:25.768939Z",
     "start_time": "2025-09-14T21:00:25.611114Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from asodesigner.read_human_genome import get_locus_to_data_dict\n",
    "import pickle\n",
    "from asodesigner.consts import CACHE_DIR\n",
    "\n",
    "genes_u = ['HIF1A', 'APOL1', 'YAP1', 'SOD1', 'SNCA', 'IRF4', 'KRAS', 'KLKB1', 'SNHG14', 'DGAT2', 'IRF5', 'HTRA1',\n",
    "           'MYH7', 'MALAT1', 'HSD17B13']\n",
    "cache_path = CACHE_DIR / 'gene_to_data_simple_cache.pickle'\n",
    "if not cache_path.exists():\n",
    "    gene_to_data = get_locus_to_data_dict(include_introns=True, gene_subset=genes_u)\n",
    "    with open(cache_path, 'wb') as f:\n",
    "        pickle.dump(gene_to_data, f)\n",
    "else:\n",
    "    with open(cache_path, 'rb') as f:\n",
    "        gene_to_data = pickle.load(f)"
   ],
   "id": "bb12a6b006aa5221",
   "outputs": [],
   "execution_count": 2
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-14T21:00:40.579128Z",
     "start_time": "2025-09-14T21:00:25.807395Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from scripts.data_genertion.data_handling import get_populated_df_with_structure_features\n",
    "\n",
    "main_df = get_populated_df_with_structure_features(main_df, genes_u, gene_to_data)"
   ],
   "id": "ba6fd9b3f8198eda",
   "outputs": [],
   "execution_count": 3
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-14T21:00:40.599714Z",
     "start_time": "2025-09-14T21:00:40.593295Z"
    }
   },
   "cell_type": "code",
   "source": "main_df[[SENSE_TYPE, 'sense_intron', 'sense_exon', 'sense_utr', SENSE_START, CANONICAL_GENE]]",
   "id": "5353e83b2f861243",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "      sense_type  sense_intron  sense_exon  sense_utr  sense_start  \\\n",
       "0         intron             1           0          0        41212   \n",
       "1         intron             1           0          0        23686   \n",
       "2         intron             1           0          0        43363   \n",
       "3         intron             1           0          0        23680   \n",
       "4         intron             1           0          0        41168   \n",
       "...          ...           ...         ...        ...          ...   \n",
       "34760     intron             1           0          0         7827   \n",
       "34761     intron             1           0          0         8250   \n",
       "34762     intron             1           0          0         8335   \n",
       "34763        utr             0           1          1        13848   \n",
       "34764     intron             1           0          0         5361   \n",
       "\n",
       "      Canonical Gene Name  \n",
       "0                    KRAS  \n",
       "1                    KRAS  \n",
       "2                    KRAS  \n",
       "3                    KRAS  \n",
       "4                    KRAS  \n",
       "...                   ...  \n",
       "34760               APOL1  \n",
       "34761               APOL1  \n",
       "34762               APOL1  \n",
       "34763               APOL1  \n",
       "34764               APOL1  \n",
       "\n",
       "[29987 rows x 6 columns]"
      ],
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sense_type</th>\n",
       "      <th>sense_intron</th>\n",
       "      <th>sense_exon</th>\n",
       "      <th>sense_utr</th>\n",
       "      <th>sense_start</th>\n",
       "      <th>Canonical Gene Name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41212</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23686</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>43363</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23680</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41168</td>\n",
       "      <td>KRAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34760</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>7827</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34761</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8250</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34762</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8335</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34763</th>\n",
       "      <td>utr</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>13848</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34764</th>\n",
       "      <td>intron</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>5361</td>\n",
       "      <td>APOL1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>29987 rows × 6 columns</p>\n",
       "</div>"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 4
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-14T21:00:40.642757Z",
     "start_time": "2025-09-14T21:00:40.633681Z"
    }
   },
   "cell_type": "code",
   "source": [
    "main_df.loc[:, 'mrna_length'] = [len(gene_to_data[gene].full_mrna) for gene in main_df[CANONICAL_GENE]]\n",
    "main_df['normalized_sense_start_from_end'] = main_df['sense_start_from_end'] / main_df['mrna_length']"
   ],
   "id": "39c480fb731b31f2",
   "outputs": [],
   "execution_count": 5
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-14T21:00:40.793077Z",
     "start_time": "2025-09-14T21:00:40.689229Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "first_filtered = main_df.copy()\n",
    "\n",
    "first_filtered = first_filtered[first_filtered['Cell line organism'] == 'human']\n",
    "# filtered = filtered[filtered[CANONICAL_GENE] == 'MALAT1']\n",
    "first_filtered = first_filtered.dropna(subset=[INHIBITION]).copy()\n",
    "# filtered = filtered.dropna(subset=[DENSITY_UPDATED]).copy()\n",
    "log_correction = 1.01\n",
    "first_filtered.loc[:, 'log_inhibition'] = -np.log(log_correction - first_filtered[INHIBITION] / 100)\n",
    "first_filtered = first_filtered[~first_filtered[CELL_LINE].isin([\n",
    "    'Hela',  # scanning modifications\n",
    "    'Human Neuronal Cell',  # scanning modifications;\n",
    "    'CC-2580',  # scanning modifications\n",
    "    'SH-SY5Y'  # non pure PS based\n",
    "])]\n",
    "# first_filtered = first_filtered[~((first_filtered[CELL_LINE] == 'A431')\n",
    "#                                   & (first_filtered[CANONICAL_GENE] == 'SOD1'))] # (Kind of) Scanning modifications\n",
    "\n",
    "mean_metric = 'log_inhibition'\n",
    "# mean_metric = INHIBITION\n",
    "\n",
    "# Ensure you remove only the columns you don't want to group by\n",
    "# cols_except_inhibition = [c for c in first_filtered.columns\n",
    "#                           if c not in [INHIBITION, mean_metric, 'index']]\n",
    "\n",
    "# Group by all remaining columns and average log_inhibition\n",
    "collapsed = (\n",
    "    first_filtered\n",
    "    .groupby(['ISIS', VOLUME, TREATMENT_PERIOD, DENSITY_UPDATED, CHEMICAL_PATTERN], as_index=False)[mean_metric]\n",
    "    .mean()\n",
    ")\n",
    "\n",
    "# first_filtered['mean_inhibition'] = first_filtered.groupby('ISIS')[mean_metric].transform('mean')\n",
    "# first_filtered = first_filtered.drop_duplicates(subset='ISIS').assign(**{mean_metric: first_filtered['mean_inhibition']}).drop(\n",
    "#     columns='mean_inhibition')"
   ],
   "id": "e500000f5f4ea91f",
   "outputs": [],
   "execution_count": 6
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-15T21:52:15.567085Z",
     "start_time": "2025-09-15T21:52:14.970209Z"
    }
   },
   "cell_type": "code",
   "source": [
    "from sklearn.feature_selection import mutual_info_regression, f_regression\n",
    "\n",
    "# filtered = filtered[filtered['Linkage'] == 'phosphorothioate']\n",
    "\n",
    "moe_pattern = 'MMMMMddddddddddMMMMM'\n",
    "# filtered = filtered[filtered['Chemical_Pattern'] == moe_pattern].copy()\n",
    "# filtered = filtered[filtered[VOLUME] == 2500].copy()\n",
    "filtered = first_filtered.copy()\n",
    "\n",
    "filtered['log_volume'] = np.log(filtered[VOLUME])\n",
    "filtered['log_treatment'] = np.log(filtered[TREATMENT_PERIOD])\n",
    "li = filtered['log_inhibition']\n",
    "\n",
    "# filtered['correct_log_inhibition'] = filtered['log_inhibition'] / (filtered['log_volume'])\n",
    "filtered['correct_log_inhibition'] = li / (filtered[VOLUME] / (filtered[VOLUME] + 10))\n",
    "filtered['correct_log_inhibition2'] = li / (filtered[VOLUME] * 0.0000601 + 0.537)\n",
    "filtered['correct_log_inhibition3'] = filtered['correct_log_inhibition2'] / (-0.032*  filtered[TREATMENT_PERIOD] + 2.36)\n",
    "# filtered['correct_log_inhibition'] = li\n",
    "\n",
    "\n",
    "\n",
    "filtered_original = filtered\n",
    "\n",
    "ALL_CELL_LINES = filtered_original[CELL_LINE].unique()\n",
    "non_hepa_cancer = filtered_original[\n",
    "    # (~((filtered_original[CELL_LINE] == 'HepG2') & (filtered_original[TRANSFECTION].str.contains('Lipo')))) &\n",
    "    (filtered_original[CELL_LINE] != 'HepG2')  # experiment settings too different\n",
    "    & (filtered_original[CELL_LINE] != 'HepaRG')  # not similar to cancer\n",
    "    & (filtered_original[CELL_LINE] != 'A-431')\n",
    "    & (filtered_original[CELL_LINE] != 'SNU-449')\n",
    "    # & (filtered_original[CELL_LINE] != 'MM.1R')\n",
    "    # & (filtered_original[CELL_LINE] != 'KARPAS-229')\n",
    "\n",
    "    ]\n",
    "\n",
    "metric = 'correct_log_inhibition2'\n",
    "\n",
    "features = (\n",
    "    # [TREATMENT_PERIOD] +\n",
    "    [\n",
    "        col\n",
    "        for col in non_hepa_cancer.select_dtypes(include=[\"number\"]).columns\n",
    "        if col not in [INHIBITION, \"log_inhibition\", \"correct_log_inhibition\", \"correct_log_inhibition2\", \"ISIS\", 'log_volume', 'index', 'Transcript']\n",
    "    ]\n",
    ")\n",
    "\n",
    "X = non_hepa_cancer[features]\n",
    "y = non_hepa_cancer[metric]\n"
   ],
   "id": "92adec4331d1a545",
   "outputs": [],
   "execution_count": 43
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-14T21:00:40.916118Z",
     "start_time": "2025-09-14T21:00:40.905452Z"
    }
   },
   "cell_type": "code",
   "source": [
    "cols = [c for c in X.columns if \"off_target\" in c]\n",
    "X[cols] = X[cols].fillna(0)"
   ],
   "id": "cd4e8280612a0125",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3676029/2996463637.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  X[cols] = X[cols].fillna(0)\n"
     ]
    }
   ],
   "execution_count": 8
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-14T21:18:57.726866Z",
     "start_time": "2025-09-14T21:18:46.609945Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# Spearman-first version (target vs features + feature-feature Spearman)\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# --- prep (numeric + median impute; avoids dropping useful cols) ---\n",
    "Xc = X.select_dtypes(include=[np.number]).copy()\n",
    "Xc = Xc.fillna(Xc.median(numeric_only=True))\n",
    "yc = pd.Series(y, index=Xc.index).astype(float)\n",
    "\n",
    "# --- Spearman with target (like your mi_df) ---\n",
    "rho = Xc.corrwith(yc, method=\"spearman\")                       # feature ↔ target\n",
    "spearman_df = (\n",
    "    rho.to_frame(\"spearman\")\n",
    "       .assign(abs_spearman=lambda d: d[\"spearman\"].abs())\n",
    "       .sort_values(\"abs_spearman\", ascending=False)\n",
    "       .reset_index()\n",
    "       .rename(columns={\"index\": \"feature\"})\n",
    ")\n",
    "\n",
    "# --- Spearman between features (matrix + sorted upper-triangle pairs) ---\n",
    "rho_ff = Xc.corr(method=\"spearman\").astype(float).abs()        # |feature ↔ feature|\n",
    "pairs_ff = (\n",
    "    rho_ff.where(np.triu(np.ones(rho_ff.shape, dtype=bool), 1))\n",
    "          .stack()\n",
    "          .rename(\"abs_spearman\")\n",
    "          .reset_index()\n",
    "          .rename(columns={\"level_0\": \"f1\", \"level_1\": \"f2\"})\n",
    "          .sort_values(\"abs_spearman\", ascending=False)\n",
    "    # .query(\"abs_spearman >= 0.85\")  # uncomment to view highly correlated pairs\n",
    ")\n",
    "\n",
    "# spearman_df -> ranked features by |Spearman with target|\n",
    "# rho_ff       -> full |Spearman| correlation matrix between features\n",
    "# pairs_ff     -> tidy table of feature pairs with highest |Spearman|\n"
   ],
   "id": "b2eb64fc91f1dddd",
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/michael/anaconda3/envs/aso_design/lib/python3.11/site-packages/pandas/core/nanops.py:1632: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  return spearmanr(a, b)[0]\n"
     ]
    }
   ],
   "execution_count": 17
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-14T21:19:33.193538Z",
     "start_time": "2025-09-14T21:19:23.063374Z"
    }
   },
   "cell_type": "code",
   "source": [
    "def select_spearman_decorrelated(Xc, spearman_df, target_n=30, rho_cut=0.85):\n",
    "    corr = Xc.corr(method=\"spearman\").abs()\n",
    "    selected = []\n",
    "    for f in spearman_df[\"feature\"]:\n",
    "        if len(selected) >= target_n:\n",
    "            break\n",
    "        if not selected or corr.loc[f, selected].max() < rho_cut:\n",
    "            selected.append(f)\n",
    "    return selected\n",
    "\n",
    "selected = select_spearman_decorrelated(Xc, spearman_df, target_n=50, rho_cut=0.85)"
   ],
   "id": "77f64e565ab5ec63",
   "outputs": [],
   "execution_count": 20
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-14T21:20:01.050367Z",
     "start_time": "2025-09-14T21:20:01.042662Z"
    }
   },
   "cell_type": "code",
   "source": "selected",
   "id": "c9d2fdc10e2b102e",
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['on_target_fold_openness40_15',\n",
       " 'Modification_min_distance_to_3prime',\n",
       " 'index',\n",
       " 'Modification_skew_index',\n",
       " 'CAI_score_global_CDS',\n",
       " 'sense_avg_accessibility',\n",
       " 'at_skew',\n",
       " 'Modification_mean_gap',\n",
       " 'Modification_3prime_run',\n",
       " 'Modification_pos_std',\n",
       " 'md_ps_hybr',\n",
       " 'ENC_score_global_CDS',\n",
       " 'premRNA_flexible_dinucleotide_fraction_20',\n",
       " 'normalized_sense_start_from_end',\n",
       " 'internal_fold',\n",
       " 'Treatment_Period(hours)',\n",
       " 'ASO_gc_block_length',\n",
       " 'premRNA_gc_content_20',\n",
       " 'gc_skew',\n",
       " 'premRNA_at_rich_region_score_40',\n",
       " 'premRNA_flexible_dinucleotide_fraction_50',\n",
       " 'Modification_symmetry_score',\n",
       " 'moe_hybr',\n",
       " 'premRNA_gc_skew_20',\n",
       " 'off_target.top200.cutoff600.premRNA_TPM',\n",
       " 'premRNA_ggg_counts_40',\n",
       " 'premRNA_at_skew_20',\n",
       " 'premRNA_ggg_counts_70',\n",
       " 'premRNA_gc_content_50',\n",
       " 'ASO_hairpin_tm',\n",
       " 'Modification_max_block_length',\n",
       " 'mod_scan',\n",
       " 'premRNA_ggg_counts_20',\n",
       " 'premRNA_gc_skew_70',\n",
       " 'premRNA_at_skew_40',\n",
       " 'gc_content_3_prime_5',\n",
       " 'ASO_at_rich_region_score',\n",
       " 'premRNA_cg_dinucleotide_fraction_30',\n",
       " 'premRNA_stop_codon_count_30',\n",
       " 'ASO_flexible_dinucleotide_fraction',\n",
       " 'premRNA_gc_block_length_40',\n",
       " 'premRNA_gc_block_length_70',\n",
       " 'premRNA_cg_dinucleotide_fraction_70',\n",
       " 'off_target_feature_50_log',\n",
       " 'dsm_su95_rev_wGU_pos1384t37Falseon_target_energy_max600',\n",
       " 'premRNA_stop_codon_count_70',\n",
       " 'self_energy',\n",
       " 'Modification_char_entropy',\n",
       " 'RNaseH1_Krel_score_R7_krel',\n",
       " 'sense_utr']"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 21
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-14T21:49:31.975108Z",
     "start_time": "2025-09-14T21:49:31.529563Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.feature_selection import mutual_info_regression\n",
    "\n",
    "def select_mi_spearman_decorrelated(Xc, yc, top_k=80, target_n=40, random_state=42):\n",
    "    \"\"\"\n",
    "    Iteratively:\n",
    "      1) Compute MI on remaining features\n",
    "      2) Keep top_k by MI\n",
    "      3) Pick the feature with the lowest max |Spearman| corr\n",
    "         (vs already-selected; if none selected yet, vs others in top_k)\n",
    "      4) Remove it and repeat until target_n are selected\n",
    "    Returns (selected_list, selection_df)\n",
    "    \"\"\"\n",
    "    assert target_n <= Xc.shape[1]\n",
    "    remaining = Xc.columns.tolist()\n",
    "    selected = []\n",
    "    records = []\n",
    "\n",
    "    for step in range(target_n):\n",
    "        # MI on remaining\n",
    "        mi = mutual_info_regression(Xc[remaining], yc, random_state=random_state)\n",
    "        mi_s = pd.Series(mi, index=remaining)\n",
    "\n",
    "        # top_k candidates by MI\n",
    "        candidates = mi_s.nlargest(min(top_k, len(remaining))).index.tolist()\n",
    "\n",
    "        # Spearman block over (candidates + selected)\n",
    "        cols = candidates + selected\n",
    "        corr = Xc[cols].corr(method=\"spearman\").abs()\n",
    "\n",
    "        if selected:\n",
    "            # max corr of each candidate vs the selected set\n",
    "            max_corr_to_sel = corr.loc[candidates, selected].max(axis=1).fillna(0.0)\n",
    "        else:\n",
    "            # no selected yet: use within-candidates redundancy\n",
    "            cc = corr.loc[candidates, candidates].copy()\n",
    "            np.fill_diagonal(cc.values, 0.0)\n",
    "            max_corr_to_sel = cc.max(axis=1)\n",
    "\n",
    "        # choose candidate with minimal max corr; tie-break by higher MI\n",
    "        mc_min = max_corr_to_sel.min()\n",
    "        tie = [c for c in candidates if max_corr_to_sel[c] == mc_min]\n",
    "        best = pd.Series({c: mi_s[c] for c in tie}).idxmax()\n",
    "\n",
    "        print(\"Selected feature: \", best)\n",
    "        selected.append(best)\n",
    "        remaining.remove(best)\n",
    "        records.append({\n",
    "            \"rank\": step+1,\n",
    "            \"feature\": best,\n",
    "            \"max_spearman_to_selected\": float(mc_min),\n",
    "            \"mi_at_pick\": float(mi_s[best])\n",
    "        })\n",
    "\n",
    "    return selected, pd.DataFrame(records)\n",
    "\n",
    "# usage:\n",
    "selected_30, summary = select_mi_spearman_decorrelated(Xc, yc, top_k=80, target_n=40)\n",
    "print(selected_30)\n",
    "print(summary)\n"
   ],
   "id": "378052e38efa04cb",
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "Input X contains NaN.",
     "output_type": "error",
     "traceback": [
      "\u001B[31m---------------------------------------------------------------------------\u001B[39m",
      "\u001B[31mValueError\u001B[39m                                Traceback (most recent call last)",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[33]\u001B[39m\u001B[32m, line 59\u001B[39m\n\u001B[32m     56\u001B[39m     \u001B[38;5;28;01mreturn\u001B[39;00m selected, pd.DataFrame(records)\n\u001B[32m     58\u001B[39m \u001B[38;5;66;03m# usage:\u001B[39;00m\n\u001B[32m---> \u001B[39m\u001B[32m59\u001B[39m selected_30, summary = select_mi_spearman_decorrelated(Xc, yc, top_k=\u001B[32m80\u001B[39m, target_n=\u001B[32m40\u001B[39m)\n\u001B[32m     60\u001B[39m \u001B[38;5;28mprint\u001B[39m(selected_30)\n\u001B[32m     61\u001B[39m \u001B[38;5;28mprint\u001B[39m(summary)\n",
      "\u001B[36mCell\u001B[39m\u001B[36m \u001B[39m\u001B[32mIn[33]\u001B[39m\u001B[32m, line 22\u001B[39m, in \u001B[36mselect_mi_spearman_decorrelated\u001B[39m\u001B[34m(Xc, yc, top_k, target_n, random_state)\u001B[39m\n\u001B[32m     18\u001B[39m records = []\n\u001B[32m     20\u001B[39m \u001B[38;5;28;01mfor\u001B[39;00m step \u001B[38;5;129;01min\u001B[39;00m \u001B[38;5;28mrange\u001B[39m(target_n):\n\u001B[32m     21\u001B[39m     \u001B[38;5;66;03m# MI on remaining\u001B[39;00m\n\u001B[32m---> \u001B[39m\u001B[32m22\u001B[39m     mi = mutual_info_regression(Xc[remaining], yc, random_state=random_state)\n\u001B[32m     23\u001B[39m     mi_s = pd.Series(mi, index=remaining)\n\u001B[32m     25\u001B[39m     \u001B[38;5;66;03m# top_k candidates by MI\u001B[39;00m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/site-packages/sklearn/utils/_param_validation.py:218\u001B[39m, in \u001B[36mvalidate_params.<locals>.decorator.<locals>.wrapper\u001B[39m\u001B[34m(*args, **kwargs)\u001B[39m\n\u001B[32m    212\u001B[39m \u001B[38;5;28;01mtry\u001B[39;00m:\n\u001B[32m    213\u001B[39m     \u001B[38;5;28;01mwith\u001B[39;00m config_context(\n\u001B[32m    214\u001B[39m         skip_parameter_validation=(\n\u001B[32m    215\u001B[39m             prefer_skip_nested_validation \u001B[38;5;129;01mor\u001B[39;00m global_skip_validation\n\u001B[32m    216\u001B[39m         )\n\u001B[32m    217\u001B[39m     ):\n\u001B[32m--> \u001B[39m\u001B[32m218\u001B[39m         \u001B[38;5;28;01mreturn\u001B[39;00m func(*args, **kwargs)\n\u001B[32m    219\u001B[39m \u001B[38;5;28;01mexcept\u001B[39;00m InvalidParameterError \u001B[38;5;28;01mas\u001B[39;00m e:\n\u001B[32m    220\u001B[39m     \u001B[38;5;66;03m# When the function is just a wrapper around an estimator, we allow\u001B[39;00m\n\u001B[32m    221\u001B[39m     \u001B[38;5;66;03m# the function to delegate validation to the estimator, but we replace\u001B[39;00m\n\u001B[32m    222\u001B[39m     \u001B[38;5;66;03m# the name of the estimator by the name of the function in the error\u001B[39;00m\n\u001B[32m    223\u001B[39m     \u001B[38;5;66;03m# message to avoid confusion.\u001B[39;00m\n\u001B[32m    224\u001B[39m     msg = re.sub(\n\u001B[32m    225\u001B[39m         \u001B[33mr\u001B[39m\u001B[33m\"\u001B[39m\u001B[33mparameter of \u001B[39m\u001B[33m\\\u001B[39m\u001B[33mw+ must be\u001B[39m\u001B[33m\"\u001B[39m,\n\u001B[32m    226\u001B[39m         \u001B[33mf\u001B[39m\u001B[33m\"\u001B[39m\u001B[33mparameter of \u001B[39m\u001B[38;5;132;01m{\u001B[39;00mfunc.\u001B[34m__qualname__\u001B[39m\u001B[38;5;132;01m}\u001B[39;00m\u001B[33m must be\u001B[39m\u001B[33m\"\u001B[39m,\n\u001B[32m    227\u001B[39m         \u001B[38;5;28mstr\u001B[39m(e),\n\u001B[32m    228\u001B[39m     )\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/site-packages/sklearn/feature_selection/_mutual_info.py:441\u001B[39m, in \u001B[36mmutual_info_regression\u001B[39m\u001B[34m(X, y, discrete_features, n_neighbors, copy, random_state, n_jobs)\u001B[39m\n\u001B[32m    325\u001B[39m \u001B[38;5;129m@validate_params\u001B[39m(\n\u001B[32m    326\u001B[39m     {\n\u001B[32m    327\u001B[39m         \u001B[33m\"\u001B[39m\u001B[33mX\u001B[39m\u001B[33m\"\u001B[39m: [\u001B[33m\"\u001B[39m\u001B[33marray-like\u001B[39m\u001B[33m\"\u001B[39m, \u001B[33m\"\u001B[39m\u001B[33msparse matrix\u001B[39m\u001B[33m\"\u001B[39m],\n\u001B[32m   (...)\u001B[39m\u001B[32m    345\u001B[39m     n_jobs=\u001B[38;5;28;01mNone\u001B[39;00m,\n\u001B[32m    346\u001B[39m ):\n\u001B[32m    347\u001B[39m \u001B[38;5;250m    \u001B[39m\u001B[33;03m\"\"\"Estimate mutual information for a continuous target variable.\u001B[39;00m\n\u001B[32m    348\u001B[39m \n\u001B[32m    349\u001B[39m \u001B[33;03m    Mutual information (MI) [1]_ between two random variables is a non-negative\u001B[39;00m\n\u001B[32m   (...)\u001B[39m\u001B[32m    439\u001B[39m \u001B[33;03m    array([0.117, 2.645, 0.0287])\u001B[39;00m\n\u001B[32m    440\u001B[39m \u001B[33;03m    \"\"\"\u001B[39;00m\n\u001B[32m--> \u001B[39m\u001B[32m441\u001B[39m     \u001B[38;5;28;01mreturn\u001B[39;00m _estimate_mi(\n\u001B[32m    442\u001B[39m         X,\n\u001B[32m    443\u001B[39m         y,\n\u001B[32m    444\u001B[39m         discrete_features=discrete_features,\n\u001B[32m    445\u001B[39m         discrete_target=\u001B[38;5;28;01mFalse\u001B[39;00m,\n\u001B[32m    446\u001B[39m         n_neighbors=n_neighbors,\n\u001B[32m    447\u001B[39m         copy=copy,\n\u001B[32m    448\u001B[39m         random_state=random_state,\n\u001B[32m    449\u001B[39m         n_jobs=n_jobs,\n\u001B[32m    450\u001B[39m     )\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/site-packages/sklearn/feature_selection/_mutual_info.py:271\u001B[39m, in \u001B[36m_estimate_mi\u001B[39m\u001B[34m(X, y, discrete_features, discrete_target, n_neighbors, copy, random_state, n_jobs)\u001B[39m\n\u001B[32m    202\u001B[39m \u001B[38;5;28;01mdef\u001B[39;00m\u001B[38;5;250m \u001B[39m\u001B[34m_estimate_mi\u001B[39m(\n\u001B[32m    203\u001B[39m     X,\n\u001B[32m    204\u001B[39m     y,\n\u001B[32m   (...)\u001B[39m\u001B[32m    211\u001B[39m     n_jobs=\u001B[38;5;28;01mNone\u001B[39;00m,\n\u001B[32m    212\u001B[39m ):\n\u001B[32m    213\u001B[39m \u001B[38;5;250m    \u001B[39m\u001B[33;03m\"\"\"Estimate mutual information between the features and the target.\u001B[39;00m\n\u001B[32m    214\u001B[39m \n\u001B[32m    215\u001B[39m \u001B[33;03m    Parameters\u001B[39;00m\n\u001B[32m   (...)\u001B[39m\u001B[32m    269\u001B[39m \u001B[33;03m           Data Sets\". PLoS ONE 9(2), 2014.\u001B[39;00m\n\u001B[32m    270\u001B[39m \u001B[33;03m    \"\"\"\u001B[39;00m\n\u001B[32m--> \u001B[39m\u001B[32m271\u001B[39m     X, y = check_X_y(X, y, accept_sparse=\u001B[33m\"\u001B[39m\u001B[33mcsc\u001B[39m\u001B[33m\"\u001B[39m, y_numeric=\u001B[38;5;129;01mnot\u001B[39;00m discrete_target)\n\u001B[32m    272\u001B[39m     n_samples, n_features = X.shape\n\u001B[32m    274\u001B[39m     \u001B[38;5;28;01mif\u001B[39;00m \u001B[38;5;28misinstance\u001B[39m(discrete_features, (\u001B[38;5;28mstr\u001B[39m, \u001B[38;5;28mbool\u001B[39m)):\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/site-packages/sklearn/utils/validation.py:1368\u001B[39m, in \u001B[36mcheck_X_y\u001B[39m\u001B[34m(X, y, accept_sparse, accept_large_sparse, dtype, order, copy, force_writeable, force_all_finite, ensure_all_finite, ensure_2d, allow_nd, multi_output, ensure_min_samples, ensure_min_features, y_numeric, estimator)\u001B[39m\n\u001B[32m   1362\u001B[39m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mValueError\u001B[39;00m(\n\u001B[32m   1363\u001B[39m         \u001B[33mf\u001B[39m\u001B[33m\"\u001B[39m\u001B[38;5;132;01m{\u001B[39;00mestimator_name\u001B[38;5;132;01m}\u001B[39;00m\u001B[33m requires y to be passed, but the target y is None\u001B[39m\u001B[33m\"\u001B[39m\n\u001B[32m   1364\u001B[39m     )\n\u001B[32m   1366\u001B[39m ensure_all_finite = _deprecate_force_all_finite(force_all_finite, ensure_all_finite)\n\u001B[32m-> \u001B[39m\u001B[32m1368\u001B[39m X = check_array(\n\u001B[32m   1369\u001B[39m     X,\n\u001B[32m   1370\u001B[39m     accept_sparse=accept_sparse,\n\u001B[32m   1371\u001B[39m     accept_large_sparse=accept_large_sparse,\n\u001B[32m   1372\u001B[39m     dtype=dtype,\n\u001B[32m   1373\u001B[39m     order=order,\n\u001B[32m   1374\u001B[39m     copy=copy,\n\u001B[32m   1375\u001B[39m     force_writeable=force_writeable,\n\u001B[32m   1376\u001B[39m     ensure_all_finite=ensure_all_finite,\n\u001B[32m   1377\u001B[39m     ensure_2d=ensure_2d,\n\u001B[32m   1378\u001B[39m     allow_nd=allow_nd,\n\u001B[32m   1379\u001B[39m     ensure_min_samples=ensure_min_samples,\n\u001B[32m   1380\u001B[39m     ensure_min_features=ensure_min_features,\n\u001B[32m   1381\u001B[39m     estimator=estimator,\n\u001B[32m   1382\u001B[39m     input_name=\u001B[33m\"\u001B[39m\u001B[33mX\u001B[39m\u001B[33m\"\u001B[39m,\n\u001B[32m   1383\u001B[39m )\n\u001B[32m   1385\u001B[39m y = _check_y(y, multi_output=multi_output, y_numeric=y_numeric, estimator=estimator)\n\u001B[32m   1387\u001B[39m check_consistent_length(X, y)\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/site-packages/sklearn/utils/validation.py:1105\u001B[39m, in \u001B[36mcheck_array\u001B[39m\u001B[34m(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_writeable, force_all_finite, ensure_all_finite, ensure_non_negative, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, estimator, input_name)\u001B[39m\n\u001B[32m   1099\u001B[39m     \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mValueError\u001B[39;00m(\n\u001B[32m   1100\u001B[39m         \u001B[33mf\u001B[39m\u001B[33m\"\u001B[39m\u001B[33mFound array with dim \u001B[39m\u001B[38;5;132;01m{\u001B[39;00marray.ndim\u001B[38;5;132;01m}\u001B[39;00m\u001B[33m,\u001B[39m\u001B[33m\"\u001B[39m\n\u001B[32m   1101\u001B[39m         \u001B[33mf\u001B[39m\u001B[33m\"\u001B[39m\u001B[33m while dim <= 2 is required\u001B[39m\u001B[38;5;132;01m{\u001B[39;00mcontext\u001B[38;5;132;01m}\u001B[39;00m\u001B[33m.\u001B[39m\u001B[33m\"\u001B[39m\n\u001B[32m   1102\u001B[39m     )\n\u001B[32m   1104\u001B[39m \u001B[38;5;28;01mif\u001B[39;00m ensure_all_finite:\n\u001B[32m-> \u001B[39m\u001B[32m1105\u001B[39m     _assert_all_finite(\n\u001B[32m   1106\u001B[39m         array,\n\u001B[32m   1107\u001B[39m         input_name=input_name,\n\u001B[32m   1108\u001B[39m         estimator_name=estimator_name,\n\u001B[32m   1109\u001B[39m         allow_nan=ensure_all_finite == \u001B[33m\"\u001B[39m\u001B[33mallow-nan\u001B[39m\u001B[33m\"\u001B[39m,\n\u001B[32m   1110\u001B[39m     )\n\u001B[32m   1112\u001B[39m \u001B[38;5;28;01mif\u001B[39;00m copy:\n\u001B[32m   1113\u001B[39m     \u001B[38;5;28;01mif\u001B[39;00m _is_numpy_namespace(xp):\n\u001B[32m   1114\u001B[39m         \u001B[38;5;66;03m# only make a copy if `array` and `array_orig` may share memory`\u001B[39;00m\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/site-packages/sklearn/utils/validation.py:120\u001B[39m, in \u001B[36m_assert_all_finite\u001B[39m\u001B[34m(X, allow_nan, msg_dtype, estimator_name, input_name)\u001B[39m\n\u001B[32m    117\u001B[39m \u001B[38;5;28;01mif\u001B[39;00m first_pass_isfinite:\n\u001B[32m    118\u001B[39m     \u001B[38;5;28;01mreturn\u001B[39;00m\n\u001B[32m--> \u001B[39m\u001B[32m120\u001B[39m _assert_all_finite_element_wise(\n\u001B[32m    121\u001B[39m     X,\n\u001B[32m    122\u001B[39m     xp=xp,\n\u001B[32m    123\u001B[39m     allow_nan=allow_nan,\n\u001B[32m    124\u001B[39m     msg_dtype=msg_dtype,\n\u001B[32m    125\u001B[39m     estimator_name=estimator_name,\n\u001B[32m    126\u001B[39m     input_name=input_name,\n\u001B[32m    127\u001B[39m )\n",
      "\u001B[36mFile \u001B[39m\u001B[32m~/anaconda3/envs/aso_design/lib/python3.11/site-packages/sklearn/utils/validation.py:169\u001B[39m, in \u001B[36m_assert_all_finite_element_wise\u001B[39m\u001B[34m(X, xp, allow_nan, msg_dtype, estimator_name, input_name)\u001B[39m\n\u001B[32m    152\u001B[39m \u001B[38;5;28;01mif\u001B[39;00m estimator_name \u001B[38;5;129;01mand\u001B[39;00m input_name == \u001B[33m\"\u001B[39m\u001B[33mX\u001B[39m\u001B[33m\"\u001B[39m \u001B[38;5;129;01mand\u001B[39;00m has_nan_error:\n\u001B[32m    153\u001B[39m     \u001B[38;5;66;03m# Improve the error message on how to handle missing values in\u001B[39;00m\n\u001B[32m    154\u001B[39m     \u001B[38;5;66;03m# scikit-learn.\u001B[39;00m\n\u001B[32m    155\u001B[39m     msg_err += (\n\u001B[32m    156\u001B[39m         \u001B[33mf\u001B[39m\u001B[33m\"\u001B[39m\u001B[38;5;130;01m\\n\u001B[39;00m\u001B[38;5;132;01m{\u001B[39;00mestimator_name\u001B[38;5;132;01m}\u001B[39;00m\u001B[33m does not accept missing values\u001B[39m\u001B[33m\"\u001B[39m\n\u001B[32m    157\u001B[39m         \u001B[33m\"\u001B[39m\u001B[33m encoded as NaN natively. For supervised learning, you might want\u001B[39m\u001B[33m\"\u001B[39m\n\u001B[32m   (...)\u001B[39m\u001B[32m    167\u001B[39m         \u001B[33m\"\u001B[39m\u001B[33m#estimators-that-handle-nan-values\u001B[39m\u001B[33m\"\u001B[39m\n\u001B[32m    168\u001B[39m     )\n\u001B[32m--> \u001B[39m\u001B[32m169\u001B[39m \u001B[38;5;28;01mraise\u001B[39;00m \u001B[38;5;167;01mValueError\u001B[39;00m(msg_err)\n",
      "\u001B[31mValueError\u001B[39m: Input X contains NaN."
     ]
    }
   ],
   "execution_count": 33
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T13:27:58.837463Z",
     "start_time": "2025-09-17T13:27:56.512131Z"
    }
   },
   "cell_type": "code",
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from sklearn.feature_selection import mutual_info_regression\n",
    "from scipy.stats import spearmanr\n",
    "\n",
    "# --- slice & clean ---\n",
    "mask_cell = (filtered_original[CELL_LINE] == 'A431') & (filtered_original[CANONICAL_GENE] == 'MALAT1')\n",
    "Xmmr = filtered_original.loc[mask_cell, features].copy()\n",
    "yMMR = pd.to_numeric(filtered_original.loc[mask_cell, 'correct_log_inhibition3'], errors='coerce')\n",
    "\n",
    "# drop feature columns with any NaNs; then drop rows with NaN in y (if any)\n",
    "Xmmr = Xmmr.loc[:, Xmmr.notna().all(axis=0)]\n",
    "row_mask = yMMR.notna()\n",
    "Xmmr = Xmmr.loc[row_mask]\n",
    "y = yMMR.loc[row_mask].astype(float).values\n",
    "\n",
    "# --- Mutual Information ---\n",
    "mi_vals = mutual_info_regression(Xmmr.astype(float), y, random_state=42)\n",
    "\n",
    "# --- Spearman rho & p-value ---\n",
    "rhos, pvals = [], []\n",
    "for col in Xmmr.columns:\n",
    "    x = pd.to_numeric(Xmmr[col], errors='coerce').values\n",
    "    rho, p = spearmanr(x, y)  # no NaNs expected after cleaning\n",
    "    rhos.append(rho)\n",
    "    pvals.append(p)\n",
    "\n",
    "# --- Assemble result ---\n",
    "mi_df = (pd.DataFrame({\n",
    "            \"feature\": Xmmr.columns,\n",
    "            \"mi\": mi_vals,\n",
    "            \"spearman_rho\": rhos,\n",
    "            \"spearman_p\": pvals\n",
    "        })\n",
    "        .sort_values(\"mi\", ascending=False)\n",
    "        .reset_index(drop=True))\n",
    "\n",
    "pd.set_option(\"display.max_columns\", None)     # show all columns\n",
    "pd.set_option(\"display.width\", 200)            # wider print\n",
    "pd.set_option(\"display.max_colwidth\", None)    # don’t truncate cell contents\n",
    "\n",
    "print(mi_df.head(20))\n"
   ],
   "id": "d55f99bcd2be4a7f",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                      feature        mi  spearman_rho    spearman_p\n",
      "0                     correct_log_inhibition3  4.611676      1.000000  0.000000e+00\n",
      "1                      Location_div_by_length  0.181713      0.124697  4.704783e-12\n",
      "2                                 sense_start  0.181711      0.124697  4.704783e-12\n",
      "3                        Location_in_sequence  0.180527      0.124697  4.704783e-12\n",
      "4                            normalized_start  0.180127      0.124697  4.704783e-12\n",
      "5             normalized_sense_start_from_end  0.180126     -0.124697  4.704783e-12\n",
      "6                        sense_start_from_end  0.179877     -0.124697  4.704783e-12\n",
      "7                                     index.1  0.152745      0.227935  2.833219e-37\n",
      "8                on_target_fold_openness40_15  0.139374      0.291598  6.606943e-61\n",
      "9                               mfe_window_45  0.136296      0.303088  6.869733e-66\n",
      "10    on_target_fold_openness_normalized40_15  0.132296      0.286547  8.679579e-59\n",
      "11                               mfe_edges_45  0.125944      0.304144  2.332091e-66\n",
      "12                                 min_mfe_45  0.112289      0.307027  1.191491e-67\n",
      "13  premRNA_flexible_dinucleotide_fraction_70  0.105876      0.300883  6.470777e-65\n",
      "14                      premRNA_gc_content_60  0.105230     -0.296919  3.475227e-63\n",
      "15    off_target.top200.cutoff600.premRNA_log  0.100998      0.276308  1.261883e-54\n",
      "16                      premRNA_gc_content_70  0.098023     -0.297655  1.665405e-63\n",
      "17                premRNA_sequence_entropy_70  0.095624     -0.232576  8.883255e-39\n",
      "18                  premRNA_purine_content_50  0.093825      0.211411  3.455027e-32\n",
      "19  premRNA_flexible_dinucleotide_fraction_50  0.093255      0.299324  3.122214e-64\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_3676029/2076979898.py:24: ConstantInputWarning: An input array is constant; the correlation coefficient is not defined.\n",
      "  rho, p = spearmanr(x, y)  # no NaNs expected after cleaning\n"
     ]
    }
   ],
   "execution_count": 47
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-14T21:49:51.460377Z",
     "start_time": "2025-09-14T21:49:41.004299Z"
    }
   },
   "cell_type": "code",
   "source": [
    "# usage:\n",
    "selected_30, summary = select_mi_spearman_decorrelated(Xmmr, yMMR, top_k=80, target_n=40)\n",
    "print(selected_30)\n",
    "print(summary)"
   ],
   "id": "5c79ef2d5b6bb054",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Selected feature:  Treatment_Period(hours)\n",
      "Selected feature:  normalized_sense_start_from_end\n",
      "Selected feature:  ENC_score_global_CDS\n",
      "Selected feature:  Modification_min_distance_to_3prime\n",
      "Selected feature:  CAI_score_global_CDS\n",
      "Selected feature:  ASO_hairpin_tm\n",
      "Selected feature:  premRNA_cg_dinucleotide_fraction_70\n",
      "Selected feature:  premRNA_ggg_counts_20\n",
      "Selected feature:  ASO_at_rich_region_score\n",
      "Selected feature:  Modification_mean_gap\n",
      "Selected feature:  RNaseH1_Krel_score_R7_krel\n",
      "Selected feature:  premRNA_gc_skew_70\n",
      "Selected feature:  premRNA_at_skew_40\n",
      "Selected feature:  ASO_gc_block_length\n",
      "Selected feature:  premRNA_flexible_dinucleotide_fraction_20\n",
      "Selected feature:  at_skew\n",
      "Selected feature:  self_energy\n",
      "Selected feature:  gc_content_3_prime_5\n",
      "Selected feature:  premRNA_stop_codon_count_70\n",
      "Selected feature:  sense_avg_accessibility\n",
      "Selected feature:  ASO_flexible_dinucleotide_fraction\n",
      "Selected feature:  gc_skew\n",
      "Selected feature:  premRNA_gc_block_length_40\n",
      "Selected feature:  internal_fold\n",
      "Selected feature:  dsm_su95_rev_wGU_pos1384t37Falseon_target_energy_max600\n",
      "Selected feature:  Modification_skew_index\n",
      "Selected feature:  premRNA_at_rich_region_score_40\n",
      "Selected feature:  premRNA_ggg_counts_70\n",
      "Selected feature:  on_target_fold_openness40_15\n",
      "Selected feature:  sense_utr\n",
      "Selected feature:  premRNA_gc_content_50\n",
      "Selected feature:  premRNA_gc_skew_20\n",
      "Selected feature:  premRNA_gc_block_length_70\n",
      "Selected feature:  md_ps_hybr\n",
      "Selected feature:  premRNA_stop_codon_count_30\n",
      "Selected feature:  premRNA_gc_content_20\n",
      "Selected feature:  premRNA_at_skew_20\n",
      "Selected feature:  premRNA_ggg_counts_40\n",
      "Selected feature:  premRNA_cg_dinucleotide_fraction_30\n",
      "Selected feature:  premRNA_flexible_dinucleotide_fraction_50\n",
      "['Treatment_Period(hours)', 'normalized_sense_start_from_end', 'ENC_score_global_CDS', 'Modification_min_distance_to_3prime', 'CAI_score_global_CDS', 'ASO_hairpin_tm', 'premRNA_cg_dinucleotide_fraction_70', 'premRNA_ggg_counts_20', 'ASO_at_rich_region_score', 'Modification_mean_gap', 'RNaseH1_Krel_score_R7_krel', 'premRNA_gc_skew_70', 'premRNA_at_skew_40', 'ASO_gc_block_length', 'premRNA_flexible_dinucleotide_fraction_20', 'at_skew', 'self_energy', 'gc_content_3_prime_5', 'premRNA_stop_codon_count_70', 'sense_avg_accessibility', 'ASO_flexible_dinucleotide_fraction', 'gc_skew', 'premRNA_gc_block_length_40', 'internal_fold', 'dsm_su95_rev_wGU_pos1384t37Falseon_target_energy_max600', 'Modification_skew_index', 'premRNA_at_rich_region_score_40', 'premRNA_ggg_counts_70', 'on_target_fold_openness40_15', 'sense_utr', 'premRNA_gc_content_50', 'premRNA_gc_skew_20', 'premRNA_gc_block_length_70', 'md_ps_hybr', 'premRNA_stop_codon_count_30', 'premRNA_gc_content_20', 'premRNA_at_skew_20', 'premRNA_ggg_counts_40', 'premRNA_cg_dinucleotide_fraction_30', 'premRNA_flexible_dinucleotide_fraction_50']\n",
      "    rank                                            feature  \\\n",
      "0      1                            Treatment_Period(hours)   \n",
      "1      2                    normalized_sense_start_from_end   \n",
      "2      3                               ENC_score_global_CDS   \n",
      "3      4                Modification_min_distance_to_3prime   \n",
      "4      5                               CAI_score_global_CDS   \n",
      "5      6                                     ASO_hairpin_tm   \n",
      "6      7                premRNA_cg_dinucleotide_fraction_70   \n",
      "7      8                              premRNA_ggg_counts_20   \n",
      "8      9                           ASO_at_rich_region_score   \n",
      "9     10                              Modification_mean_gap   \n",
      "10    11                         RNaseH1_Krel_score_R7_krel   \n",
      "11    12                                 premRNA_gc_skew_70   \n",
      "12    13                                 premRNA_at_skew_40   \n",
      "13    14                                ASO_gc_block_length   \n",
      "14    15          premRNA_flexible_dinucleotide_fraction_20   \n",
      "15    16                                            at_skew   \n",
      "16    17                                        self_energy   \n",
      "17    18                               gc_content_3_prime_5   \n",
      "18    19                        premRNA_stop_codon_count_70   \n",
      "19    20                            sense_avg_accessibility   \n",
      "20    21                 ASO_flexible_dinucleotide_fraction   \n",
      "21    22                                            gc_skew   \n",
      "22    23                         premRNA_gc_block_length_40   \n",
      "23    24                                      internal_fold   \n",
      "24    25  dsm_su95_rev_wGU_pos1384t37Falseon_target_ener...   \n",
      "25    26                            Modification_skew_index   \n",
      "26    27                    premRNA_at_rich_region_score_40   \n",
      "27    28                              premRNA_ggg_counts_70   \n",
      "28    29                       on_target_fold_openness40_15   \n",
      "29    30                                          sense_utr   \n",
      "30    31                              premRNA_gc_content_50   \n",
      "31    32                                 premRNA_gc_skew_20   \n",
      "32    33                         premRNA_gc_block_length_70   \n",
      "33    34                                         md_ps_hybr   \n",
      "34    35                        premRNA_stop_codon_count_30   \n",
      "35    36                              premRNA_gc_content_20   \n",
      "36    37                                 premRNA_at_skew_20   \n",
      "37    38                              premRNA_ggg_counts_40   \n",
      "38    39                premRNA_cg_dinucleotide_fraction_30   \n",
      "39    40          premRNA_flexible_dinucleotide_fraction_50   \n",
      "\n",
      "    max_spearman_to_selected  mi_at_pick  \n",
      "0                   0.000000    0.018248  \n",
      "1                   0.000000    0.213281  \n",
      "2                   0.000000    0.022302  \n",
      "3                   0.000000    0.000000  \n",
      "4                   0.000000    0.005406  \n",
      "5                   0.020492    0.072234  \n",
      "6                   0.051751    0.084772  \n",
      "7                   0.091392    0.001311  \n",
      "8                   0.177150    0.000000  \n",
      "9                   0.198871    0.047448  \n",
      "10                  0.228914    0.111072  \n",
      "11                  0.277724    0.159606  \n",
      "12                  0.300362    0.140855  \n",
      "13                  0.303485    0.035198  \n",
      "14                  0.313376    0.086627  \n",
      "15                  0.350126    0.080905  \n",
      "16                  0.360640    0.103239  \n",
      "17                  0.372973    0.009848  \n",
      "18                  0.393469    0.085771  \n",
      "19                  0.410350    0.140097  \n",
      "20                  0.424588    0.010060  \n",
      "21                  0.428838    0.103809  \n",
      "22                  0.482824    0.073010  \n",
      "23                  0.547802    0.038446  \n",
      "24                  0.550882    0.086649  \n",
      "25                  0.562697    0.062955  \n",
      "26                  0.580594    0.056398  \n",
      "27                  0.602208    0.027621  \n",
      "28                  0.669004    0.190419  \n",
      "29                  0.711868    0.038569  \n",
      "30                  0.712287    0.134547  \n",
      "31                  0.716204    0.135841  \n",
      "32                  0.733898    0.065754  \n",
      "33                  0.780706    0.130658  \n",
      "34                  0.786430    0.049915  \n",
      "35                  0.805935    0.108594  \n",
      "36                  0.808958    0.133469  \n",
      "37                  0.824650    0.036136  \n",
      "38                  0.826928    0.028200  \n",
      "39                  0.836705    0.115499  \n"
     ]
    }
   ],
   "execution_count": 34
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T16:17:02.858517Z",
     "start_time": "2025-09-17T16:17:01.336851Z"
    }
   },
   "cell_type": "code",
   "source": [
    "mask = (filtered_original[CELL_LINE] == 'A431') &(filtered_original[CANONICAL_GENE] == 'KRAS')\n",
    "Xmmr = filtered_original[mask][features]\n",
    "Xmmr = Xmmr.loc[:, Xmmr.notna().all(axis=0)]\n",
    "\n",
    "yMMR = filtered_original[mask][INHIBITION]\n",
    "\n",
    "mi = mutual_info_regression(Xmmr, yMMR, random_state=42)\n",
    "\n",
    "mi_df = (pd.Series(mi, index=Xmmr.columns, name=\"score\")\n",
    "           .sort_values(ascending=False)\n",
    "           .reset_index()\n",
    "           .rename(columns={\"index\": \"feature\"}))"
   ],
   "id": "850fe4953e417605",
   "outputs": [],
   "execution_count": 48
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-17T16:17:11.191880Z",
     "start_time": "2025-09-17T16:17:11.187521Z"
    }
   },
   "cell_type": "code",
   "source": [
    "mi_df = mi_df.sort_values(\"score\", ascending=False).reset_index(drop=True)\n",
    "print(mi_df.head(20))"
   ],
   "id": "343616cc740cdf1d",
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                     feature     score\n",
      "0                                    correct_log_inhibition3  3.589216\n",
      "1                                                    index.1  0.531335\n",
      "2                                             ASO_volume(nM)  0.325728\n",
      "3                                              log_treatment  0.182698\n",
      "4                    off_target.top100.cutoff600.premRNA_TPM  0.166716\n",
      "5                                    Treatment_Period(hours)  0.164405\n",
      "6                                                exp_ps_hybr  0.149815\n",
      "7                                           exp_ps_hybr_norm  0.147853\n",
      "8                         premRNA_dispersed_repeats_score_30  0.145306\n",
      "9     dsm_su95_rev_wGU_pos1384t37Trueon_target_energy_max600  0.143503\n",
      "10   dsm_su95_rev_wGU_pos1382t37Falseon_target_energy_max600  0.139901\n",
      "11   dsm_su95_rev_woGU_pos1384t37Trueon_target_energy_max600  0.139507\n",
      "12  dsm_su95_rev_woGU_pos1382t37Falseon_target_energy_max600  0.139139\n",
      "13   dsm_su95_rev_wGU_pos1384t37Falseon_target_energy_max600  0.138567\n",
      "14                        premRNA_dispersed_repeats_score_50  0.137879\n",
      "15                   off_target.top200.cutoff600.premRNA_TPM  0.137355\n",
      "16   dsm_su95_rev_woGU_pos1382t37Trueon_target_energy_max600  0.136987\n",
      "17    dsm_su95_rev_wGU_pos1382t37Trueon_target_energy_max600  0.135889\n",
      "18                                        premRNA_gc_skew_40  0.132984\n",
      "19  dsm_su95_rev_woGU_pos1384t37Falseon_target_energy_max600  0.132588\n"
     ]
    }
   ],
   "execution_count": 50
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# usage:\n",
    "selected_30, summary = select_mi_spearman_decorrelated(Xmmr, yMMR, top_k=80, target_n=40)\n",
    "print(selected_30)\n",
    "print(summary)"
   ],
   "id": "35132d3129dde243",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "Xmmr = filtered_original[filtered_original[CELL_LINE] == 'SK-MEL-28'][features]\n",
    "Xmmr = Xmmr.loc[:, Xmmr.notna().all(axis=0)]\n",
    "\n",
    "yMMR = filtered_original[filtered_original[CELL_LINE] == 'SK-MEL-28'][INHIBITION]\n",
    "\n",
    "mi = mutual_info_regression(Xmmr, yMMR, random_state=42)\n",
    "\n",
    "mi_df = (pd.Series(mi, index=Xmmr.columns, name=\"score\")\n",
    "           .sort_values(ascending=False)\n",
    "           .reset_index()\n",
    "           .rename(columns={\"index\": \"feature\"}))"
   ],
   "id": "4e6ae2d9927d8f85",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "# usage:\n",
    "selected_30, summary = select_mi_spearman_decorrelated(Xmmr, yMMR, top_k=80, target_n=40)\n",
    "print(selected_30)\n",
    "print(summary)"
   ],
   "id": "9f1454b08f666933",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-09-14T21:21:59.847731Z",
     "start_time": "2025-09-14T21:21:59.844194Z"
    }
   },
   "cell_type": "code",
   "source": [
    "selected = ['on_target_fold_openness40_15',\n",
    " 'Modification_min_distance_to_3prime',\n",
    " 'Modification_skew_index',\n",
    " 'CAI_score_global_CDS',\n",
    " 'sense_avg_accessibility',\n",
    " 'at_skew',\n",
    " 'Modification_mean_gap',\n",
    " 'Modification_3prime_run',\n",
    " 'Modification_pos_std',\n",
    " 'md_ps_hybr',\n",
    " 'ENC_score_global_CDS',\n",
    " 'premRNA_flexible_dinucleotide_fraction_20',\n",
    " 'normalized_sense_start_from_end',\n",
    " 'internal_fold',\n",
    " 'Treatment_Period(hours)',\n",
    " 'ASO_gc_block_length',\n",
    " 'premRNA_gc_content_20',\n",
    " 'gc_skew',\n",
    " 'premRNA_at_rich_region_score_40',\n",
    " 'premRNA_flexible_dinucleotide_fraction_50',\n",
    " 'Modification_symmetry_score',\n",
    " 'moe_hybr',\n",
    " 'premRNA_gc_skew_20',\n",
    " 'off_target.top200.cutoff600.premRNA_TPM',\n",
    " 'premRNA_ggg_counts_40',\n",
    " 'premRNA_at_skew_20',\n",
    " 'premRNA_ggg_counts_70',\n",
    " 'premRNA_gc_content_50',\n",
    " 'ASO_hairpin_tm',\n",
    " 'Modification_max_block_length',\n",
    " 'premRNA_ggg_counts_20',\n",
    " 'premRNA_gc_skew_70',\n",
    " 'premRNA_at_skew_40',\n",
    " 'gc_content_3_prime_5',\n",
    " 'ASO_at_rich_region_score',\n",
    " 'premRNA_cg_dinucleotide_fraction_30',\n",
    " 'premRNA_stop_codon_count_30',\n",
    " 'ASO_flexible_dinucleotide_fraction',\n",
    " 'premRNA_gc_block_length_40',\n",
    " 'premRNA_gc_block_length_70',\n",
    " 'premRNA_cg_dinucleotide_fraction_70',\n",
    " 'off_target_feature_50_log',\n",
    " 'dsm_su95_rev_wGU_pos1384t37Falseon_target_energy_max600',\n",
    " 'premRNA_stop_codon_count_70',\n",
    " 'self_energy',\n",
    " 'Modification_char_entropy',\n",
    " 'RNaseH1_Krel_score_R7_krel',\n",
    " 'sense_utr']"
   ],
   "id": "4c52dca30cefced8",
   "outputs": [],
   "execution_count": 23
  },
  {
   "metadata": {},
   "cell_type": "code",
   "outputs": [],
   "execution_count": null,
   "source": "",
   "id": "2e1a375aa7f2bd1a"
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
