{
 "cells": [
  {
   "cell_type": "code",
   "id": "initial_id",
   "metadata": {},
   "source": [
    "import numpy as np\n",
    "import pandas\n",
    "\n",
    "from asodesigner.consts import *\n",
    "from data_genertion.consts import *\n",
    "from asodesigner.util import get_gc_content"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e7476383712c015f",
   "metadata": {},
   "source": [
    "all_data = pandas.read_csv(DATA_PATH / 'data_from_article_fixed.csv')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "487a08973b8f0765",
   "metadata": {},
   "source": [
    "all_data['gc_content'] = [get_gc_content(seq) for seq in all_data[SEQUENCE]]"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "bae16c55212d4eb",
   "metadata": {},
   "source": [
    "all_data_no_nan = all_data.dropna(subset=[INHIBITION]).copy()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "2fa16f3b8b80e403",
   "metadata": {},
   "source": [
    "all_data_no_nan.loc[:, 'log_inhibition'] = -np.log(-all_data_no_nan[INHIBITION] + 100.001)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "927c4701d045a3f6",
   "metadata": {},
   "source": [
    "import ViennaRNA as RNA\n",
    "from scripts.util.print import print_correlations\n",
    "\n",
    "all_data_no_nan.loc[:, 'internal_fold'] = [RNA.fold(antisense)[1] for antisense in all_data_no_nan[SEQUENCE]]\n",
    "\n",
    "print_correlations(all_data_no_nan, 'internal_fold', INHIBITION)\n",
    "print_correlations(all_data_no_nan, 'internal_fold', 'log_inhibition')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "f8890aae9a6f949a",
   "metadata": {},
   "source": [
    "from scripts.data_genertion.data_handling import get_unique_human_genes\n",
    "\n",
    "genes_u = get_unique_human_genes(all_data)\n",
    "print(genes_u)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "172ca9c520918fd0",
   "metadata": {},
   "source": [
    "from asodesigner.read_human_genome import get_locus_to_data_dict\n",
    "import pickle\n",
    "from asodesigner.consts import CACHE_DIR\n",
    "\n",
    "cache_path = CACHE_DIR / 'gene_to_data_simple_cache.pickle'\n",
    "if not cache_path.exists():\n",
    "    gene_to_data = get_locus_to_data_dict(include_introns=True, gene_subset=genes_u)\n",
    "    with open(cache_path, 'wb') as f:\n",
    "        pickle.dump(gene_to_data, f)\n",
    "else:\n",
    "    with open(cache_path, 'rb') as f:\n",
    "        gene_to_data = pickle.load(f)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "80625f8551213237",
   "metadata": {},
   "source": [
    "for gene_name in genes_u:\n",
    "    if gene_name not in gene_to_data:\n",
    "        print(f\"gene : {gene_name} not found\")"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "ff3753dda27d2a9b",
   "metadata": {},
   "source": [
    "from data_genertion.data_handling import get_populated_df_with_structure_features\n",
    "all_data_human_gene = get_populated_df_with_structure_features(all_data, genes_u, gene_to_data)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "204275300de82357",
   "metadata": {},
   "source": [
    "all_data_human_gene_premrna_org = all_data_human_gene[all_data_human_gene[SENSE_START] != -1]\n",
    "len(all_data_human_gene_premrna_org)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "616d2f4befb35b63",
   "metadata": {},
   "source": [
    "from scripts.data_genertion.risearch_hybridization import populate_with_risearch_hybridization\n",
    "\n",
    "all_data_human_gene_premrna = all_data_human_gene_premrna_org.copy()  # so Jupyter doesn't shout at me\n",
    "\n",
    "populate_with_risearch_hybridization(all_data_human_gene_premrna, genes_u, gene_to_data)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import pandas as pd\n",
    "all_data_human_gene_premrna = pd.read_csv('out.csv')"
   ],
   "id": "9377add1da5b5b56",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "all_data_human_gene_premrna.loc[:, 'log_inhibition'] = -np.log(-all_data_human_gene_premrna[INHIBITION] + 100.001)\n",
    "all_data_human_gene_premrna['Target_gene']"
   ],
   "id": "734cbd45385700a4",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "print(all_data_human_gene_premrna.columns)",
   "id": "6cc23322d5a22bb9",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "test = all_data_human_gene_premrna\n",
    "# test = all_data_human_gene_premrna[all_data_human_gene_premrna['Linkage'] == 'phosphorothioate']\n",
    "# test = all_data_human_gene_premrna[all_data_human_gene_premrna['Target_gene'] == 'APOL1']\n",
    "# test = test.sort_values(by=['3on_target_energy_max800'], ascending=[False])\n",
    "\n",
    "# for cutoff in ['800', '900', '700']:\n",
    "for dsm in ['dsm_su95_rev_wGU_pos']:\n",
    "# for dsm in ['dsm_su95_rev_woGU_pos']:\n",
    "    # for dsm in ['dsm_su95_rev_woGU_pos', 'dsm_su95_rev_wGU_pos', 'dsm_slh04_woGU_pos']:\n",
    "\n",
    "    # for base in [0, 120, 126, 132, 138, 144]:\n",
    "    # for base in [0, 100, 138, 150]:\n",
    "    for base in [100, 138, 150]:\n",
    "        for cutoff in [400]:\n",
    "            for metric in ['on_target_energy_max']:\n",
    "                for temp in [37, 50]:\n",
    "                    for transpose in [True, False]:\n",
    "                        # for metric in ['on_target_energy_max', 'on_target_energy_sum', 'on_target_energy_fits']:\n",
    "                        #     print_correlations(test, dsm + str(base) + '6' + metric + str(cutoff), 'log_inhibition')\n",
    "                        # print_correlations(test, dsm + str(base) + '6' + 't' + str(temp) + metric + str(cutoff), 'log_inhibition')\n",
    "                        # print_correlations(test, dsm + str(base) + '5' + 't' + str(temp) + metric + str(cutoff), 'log_inhibition')\n",
    "                        print_correlations(test, dsm + str(base) + '4' + 't' + str(temp) + str(transpose) + metric + str(cutoff), 'log_inhibition')\n",
    "                        print_correlations(test, dsm + str(base) + '6' + 't' + str(temp)  +str(transpose) +  metric + str(cutoff), 'log_inhibition')\n",
    "                        # print_correlations(test, dsm + str(base) + '2' + 't' + str(temp)  +str(transpose) +  metric + str(cutoff), 'log_inhibition')\n",
    "                        # print_correlations(test, dsm + str(base) + '2' + 't' + str(temp) + metric + str(cutoff), 'log_inhibition')\n",
    "                        # print_correlations(test, dsm + str(base) + '0' + metric + str(cutoff), 'log_inhibition')\n",
    "                        # print_correlations(test, metric + str(cutoff), 'log_inhibition')\n",
    "    print_correlations(test, 'dsm_su95_rev_wGU_pos1382t50Falseon_target_energy_max400', 'log_inhibition')"
   ],
   "id": "bc4c4daaa5bfd6a2",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "len(set(list(test['dsm_slh04_woGU_pos1384t37Falseon_target_energy_max400'])))",
   "id": "92b426e2da9cc0df",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "len(set(list(test['dsm_su95_rev_wGU_pos1442on_target_energy_max800'])))",
   "id": "3fa221574e04d708",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.scatter(test['dsm_su95_rev_woGU_pos1384t50Trueon_target_energy_max400'], test['log_inhibition'], alpha=0.02)\n",
    "# plt.scatter(test['dsm_su95_rev_wGU_pos1504on_target_energy_max800'], test['log_inhibition'], alpha=0.05)\n",
    "# plt.scatter(test['dsm_su95_rev_wGU_pos1505on_target_energy_max800'], test['log_inhibition'], alpha=0.05)\n",
    "plt.show()"
   ],
   "id": "af2317d3bf8def8e",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "9539b29b38b1bbdf",
   "metadata": {},
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "test = all_data_human_gene_premrna[all_data_human_gene_premrna[CANONICAL_GENE] == 'SNCA']\n",
    "test = test.sort_values(by=['dsm_slh04_woGU_pos04t37Falseon_target_energy_max800'], ascending=[False])\n",
    "print(test.tail(5))\n",
    "plt.scatter(test['dsm_slh04_woGU_pos04t37Falseon_target_energy_max800'], test['log_inhibition'], alpha=0.05)\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "debcba8576949852",
   "metadata": {},
   "source": [
    "from scripts.data_genertion.data_handling import get_populate_fold\n",
    "from asodesigner.fold import calculate_energies, get_weighted_energy\n",
    "from asodesigner.util import get_antisense\n",
    "\n",
    "# fold_variants = [(40, 15), (60, 15), (80, 15), (100, 15), (60, 25), (80, 25), (100, 25), (80, 40), (100, 40)]\n",
    "fold_variants = [(40, 15)]\n",
    "all_data_human_gene_premrna_no_nan = get_populate_fold(all_data_human_gene_premrna, genes_u, gene_to_data, fold_variants)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "86268291af9b29e0",
   "metadata": {},
   "source": [
    "fold_feature_names = ['on_target_fold_openness', 'on_target_fold_openness_normalized']\n",
    "fold_features = []\n",
    "for (window_size, step_size) in fold_variants:\n",
    "    for feature_name in fold_feature_names:\n",
    "        fold_features.append(feature_name + str(window_size) + \"_\" + str(step_size))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "b60ca6d6e03a00b1",
   "metadata": {},
   "source": [
    "test = all_data_human_gene_premrna_no_nan\n",
    "# test = all_data_human_gene_premrna_no_nan[all_data_human_gene_premrna_no_nan[INHIBITION] > 0]\n",
    "\n",
    "\n",
    "for feature_name in fold_features:\n",
    "    print_correlations(test, feature_name, INHIBITION, p_value_threshold=0.05)\n",
    "    print_correlations(test, feature_name, 'log_inhibition', p_value_threshold=0.05)\n",
    "\n",
    "feature = fold_features[1]\n",
    "plt.scatter(test[feature], test[INHIBITION], alpha=0.02)\n",
    "plt.show()\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "85994a253c23a952",
   "metadata": {},
   "source": [
    "from scripts.data_genertion.data_handling import populate_features\n",
    "\n",
    "populate_features(all_data_human_gene_premrna_no_nan, 'self_energy')\n",
    "\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'self_energy', INHIBITION)\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'self_energy', 'log_inhibition')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "a0ea8eb3606d0d1e",
   "metadata": {},
   "source": [
    "populate_features(all_data_human_gene_premrna_no_nan, 'internal_fold')\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'internal_fold', INHIBITION)\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'internal_fold', 'log_inhibition')\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e15457978aab3485",
   "metadata": {},
   "source": [
    "all_data_human_gene_premrna_no_nan.loc[:, 'gc_content_5_prime_5'] = [get_gc_content(sequence[:5]) for sequence in\n",
    "                                                                     all_data_human_gene_premrna_no_nan[SEQUENCE]]\n",
    "all_data_human_gene_premrna_no_nan.loc[:, 'gc_content_3_prime_5'] = [get_gc_content(sequence[-5:]) for sequence in\n",
    "                                                                     all_data_human_gene_premrna_no_nan[SEQUENCE]]\n",
    "all_data_human_gene_premrna_no_nan.loc[:, 'first_nucleotide'] = [sequence[0] for sequence in\n",
    "                                                                 all_data_human_gene_premrna_no_nan[SEQUENCE]]\n",
    "all_data_human_gene_premrna_no_nan.loc[:, 'second_nucleotide'] = [sequence[2] for sequence in\n",
    "                                                                  all_data_human_gene_premrna_no_nan[SEQUENCE]]\n",
    "\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'gc_content_5_prime_5', INHIBITION)\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'gc_content_5_prime_5', 'log_inhibition')\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'gc_content_3_prime_5', 'log_inhibition')\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'gc_content_3_prime_5', 'log_inhibition')\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "648c5e86e0e0ffc",
   "metadata": {},
   "source": [
    "import primer3\n",
    "\n",
    "all_data_human_gene_premrna_no_nan.loc[:, 'self_energy'] = [primer3.calc_homodimer(antisense).dg for antisense in\n",
    "                                                            all_data_human_gene_premrna_no_nan[SEQUENCE]]\n",
    "all_data_human_gene_premrna_no_nan.loc[:, 'self_energy_log'] = -np.log(\n",
    "    2500 - all_data_human_gene_premrna_no_nan.loc[:, 'self_energy'])\n",
    "\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'self_energy', INHIBITION)\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'self_energy', 'log_inhibition')\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'self_energy_log', INHIBITION)\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'self_energy_log', 'log_inhibition')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "e5d10cef04619d72",
   "metadata": {},
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.scatter(all_data_human_gene_premrna_no_nan['self_energy_log'], all_data_human_gene_premrna_no_nan['log_inhibition'],\n",
    "            alpha=0.1)\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "191d96dd326c8dd1",
   "metadata": {},
   "source": [
    "print_correlations(all_data_human_gene_premrna_no_nan, TREATMENT_PERIOD, INHIBITION)\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, TREATMENT_PERIOD, 'log_inhibition')\n",
    "plt.scatter(all_data_human_gene_premrna_no_nan[TREATMENT_PERIOD], all_data_human_gene_premrna_no_nan[INHIBITION],\n",
    "            alpha=0.05)\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "19239ad556b46874",
   "metadata": {},
   "source": [
    "tmp = all_data_human_gene_premrna_no_nan.copy()\n",
    "tmp = tmp.dropna(subset=[DENSITY]).copy()\n",
    "tmp.loc[:, 'log_density'] = np.log(tmp[DENSITY])\n",
    "\n",
    "print_correlations(tmp, DENSITY, INHIBITION)\n",
    "print_correlations(tmp, DENSITY, 'log_inhibition')\n",
    "print_correlations(tmp, 'log_density', INHIBITION)\n",
    "print_correlations(tmp, 'log_density', 'log_inhibition')\n",
    "plt.scatter(tmp[DENSITY], tmp[INHIBITION], alpha=0.05)\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "4ff456402d070799",
   "metadata": {},
   "source": [
    "all_data_human_gene_premrna_no_nan.loc[:, 'mrna_length'] = [len(gene_to_data[gene].full_mrna) for gene in\n",
    "                                                            all_data_human_gene_premrna_no_nan[CANONICAL_GENE]]\n",
    "all_data_human_gene_premrna_no_nan.loc[:, 'normalized_start'] = all_data_human_gene_premrna_no_nan[SENSE_START] / \\\n",
    "                                                                all_data_human_gene_premrna_no_nan['mrna_length']\n",
    "\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'normalized_start', INHIBITION)\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'normalized_start', 'log_inhibition')\n",
    "\n",
    "plt.scatter(all_data_human_gene_premrna_no_nan['normalized_start'],\n",
    "            all_data_human_gene_premrna_no_nan['log_inhibition'], alpha=0.05)\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from hybridization.hybridization_features import get_exp_psrna_hybridization_normalized, get_exp_psrna_hybridization\n",
    "from hybridization.md_weights import get_md_psrna_hybridization, get_md_psrna_hybridization_normalized\n",
    "import matplotlib.pyplot as plt\n",
    "# table from article: https://pubs.acs.org/doi/10.1021/acs.jpcb.4c08344\n",
    "\n",
    "all_data_human_gene_premrna_no_nan.loc[:, 'ps_hybr'] = [get_md_psrna_hybridization(antisense.replace('T', 'U'), algo='gb') for\n",
    "                                                        antisense in all_data_human_gene_premrna_no_nan[SEQUENCE]]\n",
    "all_data_human_gene_premrna_no_nan.loc[:, 'ps_hybr_norm'] = [get_md_psrna_hybridization_normalized(antisense.replace('T', 'U'), algo='gb') for\n",
    "                                                        antisense in all_data_human_gene_premrna_no_nan[SEQUENCE]]\n",
    "\n",
    "all_data_human_gene_premrna_no_nan.loc[:, 'exp_ps_hybr'] = [get_exp_psrna_hybridization(antisense.replace('T', 'U'), temp=37) for\n",
    "                                                        antisense in all_data_human_gene_premrna_no_nan[SEQUENCE]]\n",
    "all_data_human_gene_premrna_no_nan.loc[:, 'exp_ps_hybr_norm'] = [get_exp_psrna_hybridization_normalized(antisense.replace('T', 'U'), temp=37) for\n",
    "                                                        antisense in all_data_human_gene_premrna_no_nan[SEQUENCE]]\n",
    "\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'ps_hybr', INHIBITION)\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'ps_hybr', 'log_inhibition')\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'exp_ps_hybr', INHIBITION)\n",
    "print_correlations(all_data_human_gene_premrna_no_nan, 'exp_ps_hybr', 'log_inhibition')\n",
    "\n",
    "plt.scatter(all_data_human_gene_premrna_no_nan['exp_ps_hybr'],all_data_human_gene_premrna_no_nan['log_inhibition'],alpha=0.05)\n",
    "plt.plot()"
   ],
   "id": "78f717d1e749d86d",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from hybridization.md_weights import get_md_2_moe_hybridization, get_md_2_moe_hybridization_norm\n",
    "\n",
    "moe_pattern = 'MMMMMddddddddddMMMMM'\n",
    "filtered = all_data_human_gene_premrna_no_nan.copy()\n",
    "# filtered = filtered[filtered['Chemical_Pattern'] == moe_pattern]\n",
    "filtered = filtered[filtered['Linkage'] == 'phosphorothioate']\n",
    "filtered = filtered[filtered['Cell_line'] == 'HepG2']\n",
    "filtered = filtered[filtered['Modification'] == 'MOE/5-methylcytosines/deoxy']\n",
    "\n",
    "def get_moe(pattern):\n",
    "    moe = []\n",
    "    for i, n in enumerate(pattern):\n",
    "        if n == \"M\":\n",
    "           moe.append(i)\n",
    "    return moe\n",
    "\n",
    "filtered.loc[:, 'moe_hybr'] = [get_md_2_moe_hybridization(row[0].replace('T', 'U'), get_moe(row[1])) for\n",
    "                               row in zip(filtered[SEQUENCE], filtered['Chemical_Pattern'])]\n",
    "\n",
    "filtered.loc[:, 'moe_hybr_norm'] = [get_md_2_moe_hybridization_norm(row[0].replace('T', 'U'), get_moe(row[1])) for\n",
    "                               row in zip(filtered[SEQUENCE], filtered['Chemical_Pattern'])]\n",
    "\n",
    "print_correlations(filtered, 'moe_hybr_norm', INHIBITION)\n",
    "print_correlations(filtered, 'dsm_su95_rev_wGU_pos1384t37Falseon_target_energy_max400', INHIBITION)\n",
    "print_correlations(filtered, 'ps_hybr', INHIBITION)\n",
    "# print_correlations(filtered, 'moe_hybr_log', INHIBITION)\n",
    "# print_correlations(filtered, 'moe_hybr_log', 'log_inhibition')\n",
    "\n",
    "plt.scatter(filtered['moe_hybr_norm'] ,\n",
    "            filtered['log_inhibition'], alpha=0.05)\n",
    "plt.show()"
   ],
   "id": "4812134e2e713dab",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "5684d18b0cbaad98",
   "metadata": {},
   "source": [
    "from sklearn.neighbors import KNeighborsRegressor\n",
    "from sklearn.kernel_ridge import KernelRidge\n",
    "from sklearn.svm import SVR\n",
    "from itertools import product\n",
    "from scripts.data_genertion.risearch_hybridization import RNA_TABLES\n",
    "from sklearn.neural_network import MLPRegressor\n",
    "from scipy import stats\n",
    "from sklearn.linear_model import LinearRegression\n",
    "from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor\n",
    "from sklearn.model_selection import train_test_split, cross_val_score\n",
    "from sklearn.metrics import make_scorer\n",
    "import pandas as pd\n",
    "from scipy.stats import pearsonr, spearmanr\n",
    "\n",
    "\n",
    "filtered = all_data_human_gene_premrna_no_nan.copy()\n",
    "filtered = filtered[filtered['Cell line organism'] == 'human']\n",
    "print(len(filtered))\n",
    "# filtered.loc[:, 'moe_hybr'] = [get_md_2_moe_hybridization(row[0].replace('T', 'U'), get_moe(row[1])) for\n",
    "#                                row in zip(filtered[SEQUENCE], filtered['Chemical_Pattern'])]\n",
    "filtered = filtered.dropna(subset=[INHIBITION]).copy()\n",
    "filtered = filtered.dropna(subset=[DENSITY]).copy()\n",
    "\n",
    "filtered['log_volume'] = np.log(filtered[VOLUME])\n",
    "filtered.loc[:, 'log_inhibition'] = -np.log(-filtered[INHIBITION] + 101)\n",
    "filtered.loc[:, 'log_density'] = np.log(filtered[DENSITY])\n",
    "filtered.loc[:, SENSE_LENGTH] = [len(sequence) for sequence in filtered[SEQUENCE]]\n",
    "filtered.loc[:, SENSE_LENGTH] = filtered.loc[:, SENSE_LENGTH].astype(float)\n",
    "filtered['gc_deviation'] = np.abs(0.5 - filtered['gc_content'])\n",
    "filtered['self_energy_log_squared'] = filtered['self_energy_log'] ** 2\n",
    "\n",
    "ONE_HOT_FEATURE_NAMES = [\n",
    "                         # CELL_LINE,\n",
    "                         DENSITY,\n",
    "                         TRANSFECTION,\n",
    "                         'first_nucleotide', 'second_nucleotide',\n",
    "                         # MODIFICATION,\n",
    "                         SENSE_TYPE]\n",
    "for feature in ONE_HOT_FEATURE_NAMES:\n",
    "    filtered = pd.get_dummies(filtered, columns=[feature]).copy()\n",
    "\n",
    "\n",
    "# filtered['self_energy_squared'] = filtered['self_energy'] ** 2\n",
    "\n",
    "\n",
    "# features_base = ['log_volume', TREATMENT_PERIOD, VOLUME]\n",
    "features_base = ['gc_content', 'gc_deviation',\n",
    "                 # SENSE_LENGTH,\n",
    "                 # 'normalized_start',\n",
    "                 # SENSE_START,\n",
    "                 'self_energy_log',\n",
    "                 'self_energy_log_squared', 'internal_fold', 'gc_content_3_prime_5', 'gc_content_5_prime_5',\n",
    "                 # 'dsm_su95_rev_wGU_pos05on_target_energy_max800',\n",
    "                 # 'dsm_slh04_woGU_pos04t50on_target_energy_max800',\n",
    "                 # 'dsm_slh04_woGU_pos1384t50Trueon_target_energy_max800',\n",
    "                 # 'dsm_su95_rev_wGU_pos02on_target_energy_max800'\n",
    "                 ]\n",
    "# dsm_su95_rev_wGU_pos\n",
    "experiment_features = [VOLUME, 'log_volume', TREATMENT_PERIOD, 'log_density']\n",
    "# experiment_features = []\n",
    "\n",
    "one_hot_encoded_features = []\n",
    "for one_hot_feature in ONE_HOT_FEATURE_NAMES:\n",
    "    one_hot_encoded_features.append(\n",
    "        [feature for feature in filtered.columns if one_hot_feature in feature and one_hot_feature != feature])\n",
    "# Flatten the list of list to a single big list\n",
    "flat_one_hot_encoded_features = [feature for sublist in one_hot_encoded_features for feature in sublist]\n",
    "\n",
    "features = features_base + experiment_features + fold_features + flat_one_hot_encoded_features\n",
    "\n",
    "# filtered = all_data_human_gene_premrna_no_nan\n",
    "common_chemical_pattern = 'CCCddddddddddCCC'\n",
    "moe_pattern = 'MMMMMddddddddddMMMMM'\n",
    "\n",
    "# filtered = filtered[filtered['Chemical_Pattern'] == common_chemical_pattern]\n",
    "# filtered = filtered[filtered['Chemical_Pattern'] == moe_pattern]\n",
    "# filtered = filtered[filtered['Linkage'] == 'phosphorothioate']\n",
    "# filtered = filtered[filtered['Cell_line'] == 'HepaRG']\n",
    "\n",
    "# filtered = filtered[filtered['Linkage'] == 'phosphorothioate']\n",
    "print(len(filtered))\n",
    "# filtered = filtered[filtered['Cell_line'] == 'HepG2']\n",
    "print(len(filtered))\n",
    "# filtered = filtered[filtered['Modification'] == 'MOE/5-methylcytosines/deoxy']\n",
    "print(len(filtered))\n",
    "# filtered = filtered[filtered[VOLUME] == 2500]\n",
    "print(len(filtered))\n",
    "\n",
    "\n",
    "def get_model(my_feature, filtered, features):\n",
    "    current_features = list(features)\n",
    "    current_features.append(my_feature)\n",
    "\n",
    "    # print(\"Used features: \", current_features)\n",
    "\n",
    "\n",
    "    # filtered = filtered[filtered[CANONICAL_GENE] == 'IRF5']\n",
    "    # filtered = filtered[filtered[INHIBITION] > 0]\n",
    "    # filtered = filtered[filtered[VOLUME] == 4000]\n",
    "    # filtered = filtered[filtered[INHIBITION] > 70]\n",
    "    # print(\"Length filtered: \", len(filtered))\n",
    "    # train, test = train_test_split(filtered, test_size=0.2, random_state=0, stratify=filtered[CANONICAL_GENE])\n",
    "    train, test = train_test_split(filtered, test_size=0.2)\n",
    "\n",
    "    print(\"Test: \", len(test))\n",
    "    print(\"Train: \", len(train))\n",
    "    X = train[current_features]\n",
    "    y = train['log_inhibition']\n",
    "    # x_tag = X\n",
    "    # y_tag = y\n",
    "    #\n",
    "    # weight = VOLUME\n",
    "    #\n",
    "    # # Multiply the values via numpy without changing the column names\n",
    "    # multiplier = np.sqrt(X[weight]).values.reshape(-1, 1)\n",
    "    # X_tag = pd.DataFrame(\n",
    "    #     X[features].values * multiplier,\n",
    "    #     index=X.index,\n",
    "    #     columns=X[features].columns)\n",
    "    #\n",
    "    # y_tag = y * np.sqrt(X[weight])\n",
    "\n",
    "    # model = LinearRegression()\n",
    "    # model.fit(X_tag, y_tag)\n",
    "    # model = KNeighborsRegressor(n_neighbors=20)\n",
    "    # model = RandomForestRegressor(n_estimators=100, max_depth=3, min_samples_leaf=10, n_jobs=-1)\n",
    "    model = RandomForestRegressor()\n",
    "\n",
    "\n",
    "\n",
    "    # model = RandomForestRegressor(random_state=42, n_jobs=-1)\n",
    "\n",
    "    # model.fit(X,y)\n",
    "    model.fit(train[current_features], train['log_inhibition'])\n",
    "    return model, X, y, current_features, test, train\n",
    "\n",
    "models = []\n",
    "my_features = []\n",
    "minimal_scores = [800]\n",
    "for dsm, base, prefix, score, temp, transpose  in list(product(RNA_TABLES, [138], [4,6], minimal_scores, [37, 50], [True, False])):\n",
    "    my_feature = f\"{dsm}{base}{prefix}t{temp}{transpose}on_target_energy_max{score}\"\n",
    "    my_features.append(my_feature)\n",
    "\n",
    "# my_features.append('moe_hybr')\n",
    "my_features.append('ps_hybr')\n",
    "my_features.append('exp_ps_hybr')\n",
    "# my_features.append('dsm_slh04_woGU_pos1382t37Falseon_target_energy_max400')\n",
    "# my_features.append('dsm_su95_rev_wGU_pos1382t37Falseon_target_energy_max400')\n",
    "# my_features.append('dsm_su95_rev_woGU_pos1382t37Falseon_target_energy_max400')\n",
    "\n",
    "for my_feature in my_features:\n",
    "    models.append(get_model(my_feature, filtered, features))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from sklearn.model_selection import KFold\n",
    "from scipy.stats import hmean\n",
    "cv = KFold(n_splits=5, shuffle=True, random_state=42)\n",
    "\n",
    "\n",
    "from scipy import stats\n",
    "def spearman_score(y_true, y_pred):\n",
    "    return stats.spearmanr(y_true, y_pred).correlation ** 2\n",
    "spearman_scorer = make_scorer(spearman_score, greater_is_better=True)\n",
    "\n",
    "feature_score = []\n",
    "for model, X, y, current_features, test, train in models:\n",
    "    print(current_features[-1])\n",
    "    # baseline_score = cross_val_score(model, X, y, cv=cv, n_jobs=-1).mean()\n",
    "    # print(\"Baseline score: \", baseline_score)\n",
    "\n",
    "    # spearman_score_num = cross_val_score(model, X[current_features], y, cv=cv, scoring=spearman_scorer, n_jobs=-1).mean()\n",
    "    # print(\"Spearman score train: \", spearman_score_num)\n",
    "\n",
    "    spearman_score_num = cross_val_score(model, filtered[current_features], filtered['log_inhibition'], cv=cv, scoring=spearman_scorer, n_jobs=-1).mean()\n",
    "    print(\"Spearman score everything: \", spearman_score_num)\n",
    "\n",
    "    feature_score.append((current_features[-1], spearman_score_num))\n",
    "# feature_score\n"
   ],
   "id": "e890fade54964163",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from scipy.stats import pearsonr, spearmanr\n",
    "\n",
    "for model, X, y, current_features, test, train in models:\n",
    "    train_filtered = train.copy()\n",
    "    test_filtered = test.copy()\n",
    "\n",
    "    predicted = model.predict(test_filtered[current_features])\n",
    "\n",
    "    # test_features = test_filtered[features]\n",
    "    # predicted = model.predict(test_features)\n",
    "    # y_test = test_filtered['log_inhibition'] / np.sqrt(test_filtered[weight])\n",
    "    y_test = test_filtered['log_inhibition']\n",
    "\n",
    "    print(current_features[-1])\n",
    "    print(\"Train correlations:\")\n",
    "    train_predicted = model.predict(train_filtered[current_features])\n",
    "    corr, p_value = pearsonr(train_filtered['log_inhibition'], train_predicted)\n",
    "    print(\"Pearson Corr: \", corr, \" p_value: \", p_value)\n",
    "    corr, p_value = spearmanr(train_filtered['log_inhibition'], train_predicted)\n",
    "    print(\"Spearman Corr: \", corr, \" p_value: \", p_value)\n",
    "\n",
    "    print(\"Test correlations:\")\n",
    "    corr, p_value = pearsonr(y_test, predicted)\n",
    "    print(\"TPearson Corr: \", corr, \" p_value: \", p_value)\n",
    "    corr, p_value = spearmanr(y_test, predicted)\n",
    "    print(\"TSpearman Corr: \", corr, \" p_value: \", p_value)\n",
    "#\n",
    "#\n",
    "# plt.scatter(predicted, y_test, alpha=0.2)\n",
    "# # plt.scatter(y_test, test['gc_content'])\n",
    "# x = np.linspace(np.min(predicted), np.max(predicted), 100)\n",
    "# plt.plot(x, x, color='red')\n",
    "#\n",
    "# plt.xlabel('Predicted Inhibition')\n",
    "# plt.ylabel('Actual Inhibition')\n",
    "# plt.title('Predicted vs Actual Inhibition')\n",
    "# plt.show()"
   ],
   "id": "54d9b9c6d1dde079",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "from scipy.stats import pearsonr, spearmanr\n",
    "\n",
    "for model, X, y, current_features, test, train in models:\n",
    "    train_filtered = train.copy()\n",
    "    test_filtered = test.copy()\n",
    "\n",
    "    predicted = model.predict(test_filtered[current_features])\n",
    "\n",
    "    # test_features = test_filtered[features]\n",
    "    # predicted = model.predict(test_features)\n",
    "    # y_test = test_filtered['log_inhibition'] / np.sqrt(test_filtered[weight])\n",
    "    y_test = test_filtered['log_inhibition']\n",
    "\n",
    "    print(current_features[-1])\n",
    "    print(\"Train correlations:\")\n",
    "    train_predicted = model.predict(train_filtered[current_features])\n",
    "    corr, p_value = pearsonr(train_filtered['log_inhibition'], train_predicted)\n",
    "    print(\"Pearson Corr: \", corr, \" p_value: \", p_value)\n",
    "    corr, p_value = spearmanr(train_filtered['log_inhibition'], train_predicted)\n",
    "    print(\"Spearman Corr: \", corr, \" p_value: \", p_value)\n",
    "\n",
    "    print(\"Test correlations:\")\n",
    "    corr, p_value = pearsonr(y_test, predicted)\n",
    "    print(\"TPearson Corr: \", corr, \" p_value: \", p_value)\n",
    "    corr, p_value = spearmanr(y_test, predicted)\n",
    "    print(\"TSpearman Corr: \", corr, \" p_value: \", p_value)\n",
    "#\n",
    "#\n",
    "# plt.scatter(predicted, y_test, alpha=0.2)\n",
    "# # plt.scatter(y_test, test['gc_content'])\n",
    "# x = np.linspace(np.min(predicted), np.max(predicted), 100)\n",
    "# plt.plot(x, x, color='red')\n",
    "#\n",
    "# plt.xlabel('Predicted Inhibition')\n",
    "# plt.ylabel('Actual Inhibition')\n",
    "# plt.title('Predicted vs Actual Inhibition')\n",
    "# plt.show()"
   ],
   "id": "d80c58b779500381",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "d7d7b1a80ad203f8",
   "metadata": {},
   "source": [
    "from sklearn.model_selection import permutation_test_score\n",
    "\n",
    "score, permutation_scores, pvalue = permutation_test_score(model, X, y, cv=5, n_permutations=100, scoring='r2')\n",
    "print(score, permutation_scores, pvalue)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "682802cc8e2ed3b",
   "metadata": {},
   "source": [
    "plt.scatter(filtered['self_energy_log_squared'], filtered['log_inhibition'], alpha=0.1)\n",
    "plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "5c3ef6ca9d3286d",
   "metadata": {},
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import statsmodels.api as sm\n",
    "from scipy.stats import rankdata\n",
    "\n",
    "\n",
    "def partial_corr(x, y, z):\n",
    "    # Regress x on z\n",
    "    X = sm.add_constant(z)\n",
    "    model_x = sm.OLS(x, X).fit()\n",
    "    res_x = model_x.resid\n",
    "\n",
    "    # Regress y on z\n",
    "    model_y = sm.OLS(y, X).fit()\n",
    "    res_y = model_y.resid\n",
    "\n",
    "    # Compute correlation between the residuals\n",
    "    return np.corrcoef(res_x, res_y)[0, 1]\n",
    "\n",
    "\n",
    "def partial_spearman_corr(x, y, z):\n",
    "    # Convert variables to ranks\n",
    "    rx = rankdata(x)\n",
    "    ry = rankdata(y)\n",
    "    rz = rankdata(z)\n",
    "\n",
    "    # Regress rx on rz\n",
    "    X = sm.add_constant(rz)\n",
    "    model_x = sm.OLS(rx, X).fit()\n",
    "    res_x = model_x.resid\n",
    "\n",
    "    # Regress ry on rz\n",
    "    model_y = sm.OLS(ry, X).fit()\n",
    "    res_y = model_y.resid\n",
    "\n",
    "    # Compute the Pearson correlation between the residuals\n",
    "    return np.corrcoef(res_x, res_y)[0, 1]\n",
    "\n",
    "\n",
    "for gene in genes_u:\n",
    "    t = filtered[filtered[CANONICAL_GENE] == gene]\n",
    "    print(\"Gene:\", gene)\n",
    "    try:\n",
    "        pcorr = partial_spearman_corr(t['self_energy'], t['log_inhibition'], t['gc_content'])\n",
    "        print(pcorr)\n",
    "    except:\n",
    "        pass"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "efb116eceffac737",
   "metadata": {},
   "source": [
    "from sklearn.ensemble import RandomForestRegressor\n",
    "\n",
    "rf = RandomForestRegressor(random_state=42, n_jobs=-1)\n",
    "rf.fit(filtered[features], filtered['log_inhibition'])\n",
    "\n",
    "# Get the feature importances\n",
    "importances = rf.feature_importances_\n",
    "importance_df = pd.DataFrame({\n",
    "    'Feature': X.columns,\n",
    "    'Importance': importances\n",
    "}).sort_values(by='Importance', ascending=False)\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "bf6b3b05a6ff46b1",
   "metadata": {},
   "source": [
    "for index, row in importance_df.iterrows():\n",
    "    print(f\"{row['Feature']:<40} {row['Importance']:.3}\")\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.width', None)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "from sklearn.inspection import permutation_importance\n",
    "\n",
    "def spearman_score(y_true, y_pred):\n",
    "    return stats.spearmanr(y_true, y_pred).correlation ** 2\n",
    "spearman_scorer = make_scorer(spearman_score, greater_is_better=True)\n",
    "\n",
    "for model, X, y, current_features, test, train, in models:\n",
    "\n",
    "    result = permutation_importance(\n",
    "        model,\n",
    "        X[current_features],\n",
    "        y,\n",
    "        n_repeats=100,       # shuffle each feature 10 times\n",
    "        random_state=42,\n",
    "        scoring=spearman_scorer,\n",
    "        n_jobs=-1,\n",
    "    )\n",
    "\n",
    "    imp_df = pd.DataFrame({\n",
    "        'feature': current_features,\n",
    "        'importance': result.importances_mean,\n",
    "        'std': result.importances_std\n",
    "    }).sort_values('importance', ascending=False)\n",
    "\n",
    "    # spearman_score_num = cross_val_score(model, filtered[current_features], filtered['log_inhibition'], cv=cv, scoring=spearman_scorer, n_jobs=-1).mean()\n",
    "    # print(\"Spearman score everything: \", spearman_score_num)\n",
    "    print(imp_df.head(10))"
   ],
   "id": "7225d6a995f56490",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "jupyter": {}
   },
   "cell_type": "code",
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.width', None)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "from sklearn.inspection import permutation_importance\n",
    "\n",
    "def spearman_score(y_true, y_pred):\n",
    "    return stats.spearmanr(y_true, y_pred).correlation ** 2\n",
    "spearman_scorer = make_scorer(spearman_score, greater_is_better=True)\n",
    "\n",
    "for model, X, y, current_features, test, train, in models:\n",
    "\n",
    "    result = permutation_importance(\n",
    "        model,\n",
    "        X[current_features],\n",
    "        y,\n",
    "        n_repeats=100,       # shuffle each feature 10 times\n",
    "        random_state=42,\n",
    "        scoring=spearman_scorer,\n",
    "        n_jobs=-1,\n",
    "    )\n",
    "\n",
    "    imp_df = pd.DataFrame({\n",
    "        'feature': current_features,\n",
    "        'importance': result.importances_mean,\n",
    "        'std': result.importances_std\n",
    "    }).sort_values('importance', ascending=False)\n",
    "\n",
    "    # spearman_score_num = cross_val_score(model, filtered[current_features], filtered['log_inhibition'], cv=cv, scoring=spearman_scorer, n_jobs=-1).mean()\n",
    "    # print(\"Spearman score everything: \", spearman_score_num)\n",
    "    print(imp_df.head(10))"
   ],
   "id": "f1d95fc523b9d0ab",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {
    "jupyter": {}
   },
   "cell_type": "code",
   "source": [
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.width', None)\n",
    "pd.set_option('display.max_colwidth', None)\n",
    "from sklearn.inspection import permutation_importance\n",
    "\n",
    "def spearman_score(y_true, y_pred):\n",
    "    return stats.spearmanr(y_true, y_pred).correlation ** 2\n",
    "spearman_scorer = make_scorer(spearman_score, greater_is_better=True)\n",
    "\n",
    "for model, X, y, current_features, test, train, in models:\n",
    "\n",
    "    result = permutation_importance(\n",
    "        model,\n",
    "        X[current_features],\n",
    "        y,\n",
    "        n_repeats=100,       # shuffle each feature 10 times\n",
    "        random_state=42,\n",
    "        scoring=spearman_scorer,\n",
    "        n_jobs=-1,\n",
    "    )\n",
    "\n",
    "    imp_df = pd.DataFrame({\n",
    "        'feature': current_features,\n",
    "        'importance': result.importances_mean,\n",
    "        'std': result.importances_std\n",
    "    }).sort_values('importance', ascending=False)\n",
    "\n",
    "    # spearman_score_num = cross_val_score(model, filtered[current_features], filtered['log_inhibition'], cv=cv, scoring=spearman_scorer, n_jobs=-1).mean()\n",
    "    # print(\"Spearman score everything: \", spearman_score_num)\n",
    "    print(imp_df.head(10))"
   ],
   "id": "3d9e4d54082da453",
   "outputs": [],
   "execution_count": null
  },
  {
   "metadata": {},
   "cell_type": "code",
   "source": "dsm_su95_rev_wGU_pos",
   "id": "705222e948acb8b6",
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "16784a84f32b5724",
   "metadata": {},
   "source": [
    "print(\"The baseline is: \", baseline_score)\n",
    "print(\"The spearman baseline is: \", spearman_score)\n",
    "cv = 5\n",
    "scorer = spearman_scorer\n",
    "jobs = 1\n",
    "feature_importances = []\n",
    "\n",
    "# for feature in features_base + on_target_energy_features + fold_features:\n",
    "for feature in features_base:\n",
    "    score_without_feature = cross_val_score(model, X.drop(columns=[feature]), y, cv=cv, scoring=scorer,\n",
    "                                            n_jobs=jobs).mean()\n",
    "    # print(f\"Dropping {feature}: performance drop = {baseline_score - score_without_feature}\")\n",
    "    feature_importances.append((feature, float(baseline_score - score_without_feature)))\n",
    "\n",
    "print(\"Done base features\")\n",
    "\n",
    "for i in range(len(one_hot_encoded_features)):\n",
    "    feature_name = ONE_HOT_FEATURE_NAMES[i]\n",
    "    feature = one_hot_encoded_features[i]\n",
    "\n",
    "    score_without_feature = cross_val_score(model, X.drop(columns=feature), y, cv=cv, scoring=scorer,\n",
    "                                            n_jobs=jobs).mean()\n",
    "    performance_diff = baseline_score - score_without_feature\n",
    "    feature_importances.append((feature_name, float(performance_diff)))\n",
    "\n",
    "#\n",
    "# score_without_on_target_energy = cross_val_score(model, X.drop(columns=on_target_energy_features), y, cv=cv, scoring=scorer, n_jobs=jobs).mean()\n",
    "# performance_drop_all_on_target_energy = baseline_score - score_without_on_target_energy\n",
    "# feature_importances.append((\"All on target energy\", float(performance_drop_all_on_target_energy)))\n",
    "\n",
    "score_without_fold_features = cross_val_score(model, X.drop(columns=fold_features), y, cv=cv, scoring=scorer,\n",
    "                                              n_jobs=jobs).mean()\n",
    "performance_drop_fold_features = baseline_score - score_without_fold_features\n",
    "feature_importances.append((\"Fold features\", float(performance_drop_fold_features)))\n",
    "\n",
    "sorted_data = sorted(feature_importances, key=lambda x: x[1], reverse=True)\n",
    "for datum in sorted_data:\n",
    "    print(datum)\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "17cd4d57b12bcc05",
   "metadata": {},
   "source": [
    "print(filtered.columns)\n",
    "\n",
    "for gene in genes_u:\n",
    "    print(\"Gene:\", gene)\n",
    "\n",
    "    try:\n",
    "        predicted_gene = model.predict(filtered[filtered[CANONICAL_GENE] == gene][features])\n",
    "        y_gene = filtered[filtered[CANONICAL_GENE] == gene]['log_inhibition']\n",
    "        corr, p_value = pearsonr(y_gene, predicted_gene)\n",
    "        print(\"Pearson Corr: \", corr, \" p_value: \", p_value)\n",
    "        corr, p_value = spearmanr(y_gene, predicted_gene)\n",
    "        print(\"Spearman Corr: \", corr, \" p_value: \", p_value)\n",
    "    except:\n",
    "        pass\n",
    "# plt.scatter(predicted, y)\n",
    "# plt.show()"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "646f69982894297b",
   "metadata": {},
   "source": [
    "for gene in genes_u:\n",
    "    print(\"Gene: \", gene)\n",
    "    print(len(set(list(\n",
    "        all_data_human_gene_premrna_no_nan[all_data_human_gene_premrna_no_nan[CANONICAL_GENE] == gene][SEQUENCE]))))"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "id": "b30d279fcc1f6752",
   "metadata": {},
   "source": [
    "subset = ['Sequence', 'Chemical_Pattern', 'Smiles', 'Linkage', 'Location', 'Modification', 'Primer_probe_set', 'ISIS',\n",
    "          'Cell_line', VOLUME, DENSITY]\n",
    "result = all_data_human_gene_premrna_no_nan.groupby(subset).filter(lambda x: len(x) > 20)\n",
    "# result = all_data_human_gene_premrna_no_nan\n",
    "result = result.sort_values(by=['Sequence'])\n",
    "result"
   ],
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
